{
 "metadata": {
  "name": "",
  "signature": "sha256:482fedad42adedadd77154b18770520100f72cdc914cb3919f3a520358e9762e"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# DEXSeq\n",
      "\n",
      "I'm going to run DEXSeq using the coverage of the splice junctions. \n",
      "I will run an analysis for each of \n",
      "\n",
      "* BRCA\n",
      "* CLL\n",
      "* UM\n",
      "* BRCA, CLL, UM\n",
      "* BRCA, CLL, LUAD, LUSC, UM\n",
      "\n",
      "DEXSeq was designed to be used with exon coverage data but can be \n",
      "used with splice junction data as well. However, we have to make some\n",
      "dummy data to satisfy the input file requirements for DEXSeq. I'll take\n",
      "the DEXSeq output and parse into a format that is useful for later on.\n",
      "\n",
      "#### Run time\n",
      "\n",
      "Note that the DEXSeq runs can take a little while (minutes to hours\n",
      "with 30 threads) and that DEXSeq is multi-threaded. Set the `threads` \n",
      "variable below depending on how many threads you have available on \n",
      "your machine. With 30 threads, I think DEXSeq finished running in less\n",
      "than 24 hours.\n",
      "\n",
      "\n",
      "#### Extra dependencies\n",
      "\n",
      "There was some sort of dependency issue for the versions of R and DEXSeq\n",
      "that I have available via `rpy2`. See  [here](http://seqanswers.com/forums/showthread.php?t=42946) and [here](https://stat.ethz.ch/pipermail/bioconductor/2014-April/059098.html) for more info.\n",
      "Since I couldn't use `rpy2` here, I just print out R scripts and run them\n",
      "using a version of R I compiled where DEXSeq works. If you want to replicate \n",
      "these results, I recommend trying to match my R version (3.0.0) and the \n",
      "version of the packages (see the `.Rout` files in this notebook's output\n",
      "folder for my `sessionInfo()`). You may be able to use newer versions of \n",
      "R etc. without using the multiple threads, but this will be pretty slow."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from copy import deepcopy\n",
      "import os\n",
      "import tempfile\n",
      "\n",
      "import ds2014 as ds\n",
      "import pandas as pd\n",
      "\n",
      "%load_ext rmagic"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "curdir = os.getcwd()\n",
      "outdir = os.path.join(os.path.split(curdir)[0], \n",
      "                      'output', \n",
      "                      'dexseq')\n",
      "ds.makedir(outdir)\n",
      "\n",
      "# Number of threads to use for DEXSeq\n",
      "threads = 30\n",
      "\n",
      "meta = pd.read_table(ds.metadata, index_col=0, header=0)\n",
      "\n",
      "gene_info = pd.read_table(ds.gene_info, header=0,\n",
      "                          index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def make_dexseq_junction_input(counts, annot, meta):\n",
      "    \"\"\"\n",
      "    Make files needed for DEXSeq\n",
      "    \n",
      "    DEXseq requires several files that are specifically ordered. That is,\n",
      "    the rows of the different files are supposed to be matched across files.\n",
      "    Since I'm not using DEXSeq in the typical way (I'm providing\n",
      "    splice junction coverage instead of exon coverage), I'll generate\n",
      "    the files myself.\n",
      "    \"\"\"\n",
      "    # File with counts per junction\n",
      "    n, jxn_counts = tempfile.mkstemp()\n",
      "    counts = deepcopy(counts)\n",
      "    counts.columns = [ x.replace('-', '.') for x in counts.columns ]\n",
      "    counts.to_csv(jxn_counts, sep='\\t')\n",
      "    \n",
      "    # File with \"exon\" identifiers\n",
      "    junc_number = []\n",
      "    junc_count = dict()\n",
      "    for gene in annot.gene_id:\n",
      "        if junc_count.has_key(gene):\n",
      "            junc_count[gene] += 1\n",
      "        else:\n",
      "            junc_count[gene] = 1 \n",
      "        junc_number.append(junc_count[gene])\n",
      "    n, exon_ids = tempfile.mkstemp()\n",
      "    f = open(exon_ids, 'w')\n",
      "    f.write('exon_id\\n')\n",
      "    f.write('\\n'.join([ str(x) for x in junc_number ]) + '\\n')\n",
      "    f.close()\n",
      "    \n",
      "    # File with gene identifiers\n",
      "    n, gene_ids = tempfile.mkstemp()\n",
      "    annot.gene_id.to_csv(gene_ids, sep='\\t',\n",
      "                         index=False, header=True)\n",
      "    \n",
      "    # File with \"exon\" intervals\n",
      "    gene_strand = dict(zip(annot.gene_id.values, annot.strand))\n",
      "    # This is just dummy data so DEXSeq will run.\n",
      "    intervals = []\n",
      "    chrom = 'chrF' # F for fake\n",
      "    start = 1 \n",
      "    end = 10\n",
      "    for gene in annot.gene_id:\n",
      "        strand = gene_strand[gene]\n",
      "        intervals.append('\\t'.join([chrom, str(start), str(end), strand]))\n",
      "        start += 15\n",
      "        end += 15\n",
      "    n, exon_intervals = tempfile.mkstemp()\n",
      "    f = open(exon_intervals, 'w')\n",
      "    f.write('chr\\tstart\\tend\\tstrand\\n')\n",
      "    f.write('\\n'.join([ str(x) for x in intervals ]) + '\\n')\n",
      "    f.close()\n",
      "    \n",
      "    # File with \"transcripts\"\n",
      "    # This is just dummy data so DEXSeq will run.\n",
      "    n, transcripts = tempfile.mkstemp()\n",
      "    f = open(transcripts, 'w')\n",
      "    f.write('transcripts\\n')\n",
      "    f.write('\\n'.join(list(annot.index)) + '\\n')\n",
      "    f.close()\n",
      "    \n",
      "    # File with study design\n",
      "    n, design = tempfile.mkstemp()\n",
      "    tdf = meta.ix[[ x.replace('.', '-') for x in counts.columns ]]\n",
      "    tdf.index = [ x.replace('-', '.') for x in tdf.index ]\n",
      "    if len(set(tdf['disease_abbr'])) > 1:\n",
      "        tdf[['disease_abbr', 'condition']].to_csv(design, sep='\\t')\n",
      "    else:\n",
      "        tdf['condition'].to_csv(design, sep='\\t', header=True)\n",
      "    \n",
      "    return jxn_counts, exon_ids, gene_ids, exon_intervals, transcripts, design"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Due to the dependencies issues I mentioned above, I \n",
      "wrote functions that print out R scripts that will \n",
      "run DEXSeq for one cancer (`make_dexseq_script_single`)\n",
      "or multiple cancers (`make_dexseq_script_multiple`).\n",
      "\n",
      "I've also made a function to parse the DEXSeq results\n",
      "into a more manageable format."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def make_dexseq_script_multiple(counts, exon_ids, gene_ids, \n",
      "                                exon_intervals, transcripts,\n",
      "                                design, script, outfile, \n",
      "                                size_factors, threads=threads):\n",
      "    lines = [\n",
      "        'library(DEXSeq)',\n",
      "        'library(multicore)',\n",
      "        '',\n",
      "        'countData = read.csv(\"{}\", header=T, row.names=1, sep=\"\\\\t\",check.names=F)'.format(counts),\n",
      "        'geneIDs = read.csv(\"{}\", header=T)'.format(gene_ids),\n",
      "        'geneIDs = as.vector(geneIDs[,])',\n",
      "        'exonIDs = read.csv(\"{}\",header=T)'.format(exon_ids),\n",
      "        'exonIDs = as.vector(exonIDs[,])',\n",
      "        'exonIntervals = read.csv(\"{}\",header=T,sep=\"\\\\t\")'.format(exon_intervals),\n",
      "        'transcripts = read.csv(\"{}\",header=T)'.format(transcripts),\n",
      "        'transcripts = as.vector(transcripts[,])',\n",
      "        'design = read.csv(\"{}\", header=T, row.names=1, sep=\"\\\\t\")'.format(design),\n",
      "        '',\n",
      "        'ECS = newExonCountSet(countData, design, geneIDs, exonIDs, exonIntervals, transcripts)',\n",
      "        'ECS = estimateSizeFactors(ECS)',\n",
      "        'sf = as.data.frame(sizeFactors(ECS))',\n",
      "        'colnames(sf) = \"size_factors\"',\n",
      "        'write.table(sf, \"{}\", quote=F, sep=\"\\\\t\", col.names=NA)'.format(size_factors),\n",
      "        '',\n",
      "        'formuladispersion = ~ sample + (exon + disease_abbr) * condition',\n",
      "        'ECS = estimateDispersionsTRT(ECS, formula=formuladispersion, nCores={})'.format(threads),\n",
      "        'ECS = fitDispersionFunction(ECS)',\n",
      "        'formula0 = ~ sample + disease_abbr * exon + condition', \n",
      "        'formula1 = ~ sample + disease_abbr * exon + condition * exon',\n",
      "        'ECS = testForDEUTRT(ECS,formula0 = formula0,formula1 = formula1, nCores={})'.format(threads),\n",
      "        'ECS = estimatelog2FoldChanges(ECS, nCores={})'.format(threads),\n",
      "        'exonRes = DEUresultTable(ECS)',\n",
      "        '',\n",
      "        'write.table(fData(ECS), \"{}\", quote=F, sep=\"\\\\t\", col.names=NA)'.format(outfile),\n",
      "        'sessionInfo()',\n",
      "    ]\n",
      "    f = open(script, 'w')\n",
      "    f.write('\\n'.join(lines) + '\\n')\n",
      "    f.close()\n",
      "    \n",
      "def make_dexseq_script_single(counts, exon_ids, gene_ids, \n",
      "                              exon_intervals, transcripts,\n",
      "                              design, script, outfile, \n",
      "                              size_factors, threads=threads):\n",
      "    lines = [\n",
      "        'library(DEXSeq)',\n",
      "        'library(multicore)',\n",
      "        '',\n",
      "        'countData = read.csv(\"{}\", header=T, row.names=1, sep=\"\\\\t\",check.names=F)'.format(counts),\n",
      "        'geneIDs = read.csv(\"{}\", header=T)'.format(gene_ids),\n",
      "        'geneIDs = as.vector(geneIDs[,])',\n",
      "        'exonIDs = read.csv(\"{}\",header=T)'.format(exon_ids),\n",
      "        'exonIDs = as.vector(exonIDs[,])',\n",
      "        'exonIntervals = read.csv(\"{}\",header=T,sep=\"\\\\t\")'.format(exon_intervals),\n",
      "        'transcripts = read.csv(\"{}\",header=T)'.format(transcripts),\n",
      "        'transcripts = as.vector(transcripts[,])',\n",
      "        'design = read.csv(\"{}\", header=T, row.names=1, sep=\"\\\\t\")'.format(design),\n",
      "        '',\n",
      "        'ECS = newExonCountSet(countData, design, geneIDs, exonIDs, exonIntervals, transcripts)',\n",
      "        'ECS = estimateSizeFactors(ECS)',\n",
      "        'sf = as.data.frame(sizeFactors(ECS))',\n",
      "        'colnames(sf) = \"size_factors\"',\n",
      "        'write.table(sf, \"{}\", quote=F, sep=\"\\\\t\", col.names=NA)'.format(size_factors),\n",
      "        '',\n",
      "        'ECS = estimateDispersionsTRT(ECS, nCores={})'.format(threads),\n",
      "        'ECS = fitDispersionFunction(ECS)',\n",
      "        'ECS = testForDEUTRT(ECS, nCores={})'.format(threads),\n",
      "        'ECS = estimatelog2FoldChanges(ECS, nCores={})'.format(threads),\n",
      "        'exonRes = DEUresultTable(ECS)',\n",
      "        '',\n",
      "        'write.table(fData(ECS), \"{}\", quote=F, sep=\"\\\\t\", col.names=NA)'.format(outfile),\n",
      "        'sessionInfo()',\n",
      "    ]\n",
      "    f = open(script, 'w')\n",
      "    f.write('\\n'.join(lines) + '\\n')\n",
      "    f.close()\n",
      "    \n",
      "def parse_results(outfile, size_factors, cancers, annot, gene_info=gene_info):\n",
      "    res = pd.read_table(outfile, header=0,\n",
      "                        index_col=0)\n",
      "\n",
      "    res.index = res.transcripts\n",
      "    res = res.drop(['exonID', 'dispBeforeSharing', 'dispFitted',\n",
      "                    'dispersion', 'chr', 'start', 'end', \n",
      "                    'strand', 'transcripts', 'geneID'], axis=1)\n",
      "    res = res.join(annot)\n",
      "    res = res.join(gene_info, on='gene_id')\n",
      "    res.index.name = 'junction'\n",
      "    if 'log2fold(WT/MUT)' in res.columns:\n",
      "        res['log2fold(MUT/WT)'] = log2(1 / 2 ** res['log2fold(WT/MUT)'])\n",
      "    elif 'log2fold(MUT/WT)' in res.columns:\n",
      "        res['log2fold(WT/MUT)'] = log2(1 / 2 ** res['log2fold(MUT/WT)'])\n",
      "        \n",
      "    s = pd.read_table(size_factors,\n",
      "                      index_col=0,\n",
      "                      header=0)\n",
      "    s.index = [ x.replace('.', '-') for x in s.index ]\n",
      "    s.to_csv(size_factors, sep='\\t')\n",
      "    return res"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The function `run_dexseq` below runs DEXSeq for \n",
      "the various combinations of cancers we are interested\n",
      "in. Note that I use relative paths (sorry) and specify\n",
      "the R path."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def run_dexseq(cancers):\n",
      "    dest = os.path.join(outdir, \n",
      "                        '{}_dexseq_results.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "    if cancers == ['BRCA']:\n",
      "        countfile = ds.brca_sj_counts\n",
      "        annotfile = ds.brca_sj_annot\n",
      "    elif cancers == ['CLL']:\n",
      "        countfile = ds.cll_sj_counts\n",
      "        annotfile = ds.cll_sj_annot\n",
      "    elif cancers == ['UM']:\n",
      "        countfile = ds.um_sj_counts\n",
      "        annotfile = ds.um_sj_annot\n",
      "    elif cancers == ['BRCA', 'CLL', 'UM']:\n",
      "        countfile = ds.brca_cll_um_sj_counts\n",
      "        annotfile = ds.brca_cll_um_sj_annot\n",
      "    elif cancers == ['BRCA', 'CLL', 'LUAD', 'LUSC', 'UM']:\n",
      "        countfile = ds.brca_cll_luad_lusc_um_sj_counts\n",
      "        annotfile = ds.brca_cll_luad_lusc_um_sj_annot\n",
      "    else:\n",
      "        countfile = None\n",
      "        annotfile = None\n",
      "    if not os.path.exists(dest):\n",
      "        counts = pd.read_table(countfile, index_col=0,\n",
      "                               header=0)\n",
      "        annot = pd.read_table(annotfile, index_col=0,\n",
      "                              header=0)\n",
      "        jxn_counts, exon_ids, gene_ids, exon_intervals, transcripts, design = make_dexseq_junction_input(counts, annot, meta)\n",
      "\n",
      "        script = os.path.join(\n",
      "            outdir, \n",
      "            '{}.R'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "        size_factors = os.path.join(\n",
      "            outdir, \n",
      "            '{}_size_factors.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "        n, outfile = tempfile.mkstemp()\n",
      "        if len(cancers) == 1:\n",
      "            make_dexseq_script_single(jxn_counts, exon_ids, gene_ids, \n",
      "                                      exon_intervals, transcripts,\n",
      "                                      design, script, outfile, \n",
      "                                      size_factors, threads=threads)\n",
      "        else:\n",
      "            make_dexseq_script_multiple(jxn_counts, exon_ids, gene_ids, \n",
      "                                        exon_intervals, transcripts,\n",
      "                                        design, script, outfile, \n",
      "                                        size_factors, threads=threads)\n",
      "    if cancers == ['BRCA']:\n",
      "        !../software/R-3.0.0/bin/R CMD BATCH ../output/dexseq/brca.R ../output/dexseq/brca.Rout\n",
      "        annotfile = ds.brca_sj_annot\n",
      "    elif cancers == ['CLL']:\n",
      "        !../software/R-3.0.0/bin/R CMD BATCH ../output/dexseq/cll.R ../output/dexseq/cll.Rout\n",
      "    elif cancers == ['UM']:\n",
      "        !../software/R-3.0.0/bin/R CMD BATCH ../output/dexseq/um.R ../output/dexseq/um.Rout\n",
      "    elif cancers == ['BRCA', 'CLL', 'UM']:\n",
      "        !../software/R-3.0.0/bin/R CMD BATCH ../output/dexseq/brca_cll_um.R ../output/dexseq/brca_cll_um.Rout\n",
      "    elif cancers == ['BRCA', 'CLL', 'LUAD', 'LUSC', 'UM']:\n",
      "        !../software/R-3.0.0/bin/R CMD BATCH ../output/dexseq/brca_cll_luad_lusc_um.R ../output/dexseq/brca_cll_luad_lusc_um.Rout\n",
      "    \n",
      "    res = parse_results(outfile, size_factors, cancers, annot)\n",
      "    res.to_csv(dest, sep='\\t')\n",
      "    for f in jxn_counts, exon_ids, gene_ids, exon_intervals, transcripts, design, outfile:\n",
      "        os.remove(f)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Now I'll run DEXSeq for the different cancer combinations."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['BRCA']\n",
      "dest = os.path.join(outdir, \n",
      "                    '{}_dexseq_results.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "if not os.path.exists(dest):\n",
      "    run_dexseq(cancers)\n",
      "else:\n",
      "    res = pd.read_table(dest, header=0,\n",
      "                        index_col=0)\n",
      "    print(dest)\n",
      "    print(res.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "WARNING: ignoring environment value of R_HOME\r\n"
       ]
      }
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['CLL']\n",
      "dest = os.path.join(outdir, \n",
      "                    '{}_dexseq_results.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "if not os.path.exists(dest):\n",
      "    run_dexseq(cancers)\n",
      "else:\n",
      "    res = pd.read_table(dest, header=0,\n",
      "                        index_col=0)\n",
      "    print(dest)\n",
      "    print(res.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/dexseq/cll_dexseq_results.tsv\n",
        "                          testable    pvalue   padjust  log2fold(WT/MUT)  \\\n",
        "junction                                                                   \n",
        "chr20:49551774-49552684:-     True  0.269337  0.701325          0.078461   \n",
        "chr20:49552800-49557401:-     True  0.455347  0.827601         -0.039536   \n",
        "chr20:49552800-49558567:-     True  0.957638  0.991132          0.192218   \n",
        "chr20:49557471-49557641:-     True  0.316528  0.741015         -0.513647   \n",
        "chr20:49557471-49557665:-     True  0.495506  0.846666         -0.479947   \n",
        "\n",
        "                           chrom     start       end strand intron_motif  \\\n",
        "junction                                                                   \n",
        "chr20:49551774-49552684:-  chr20  49551774  49552684      -        CT/AC   \n",
        "chr20:49552800-49557401:-  chr20  49552800  49557401      -        CT/AC   \n",
        "chr20:49552800-49558567:-  chr20  49552800  49558567      -        CT/AC   \n",
        "chr20:49557471-49557641:-  chr20  49557471  49557641      -        CT/AC   \n",
        "chr20:49557471-49557665:-  chr20  49557471  49557665      -        CT/AC   \n",
        "\n",
        "                          annotated    ...             acceptor novel_donor  \\\n",
        "junction                               ...                                    \n",
        "chr20:49551774-49552684:-      True    ...     chr20:49551774:-       False   \n",
        "chr20:49552800-49557401:-      True    ...     chr20:49552800:-       False   \n",
        "chr20:49552800-49558567:-      True    ...     chr20:49552800:-       False   \n",
        "chr20:49557471-49557641:-      True    ...     chr20:49557471:-       False   \n",
        "chr20:49557471-49557665:-      True    ...     chr20:49557471:-       False   \n",
        "\n",
        "                          novel_acceptor upstream_donor_dist  \\\n",
        "junction                                                       \n",
        "chr20:49551774-49552684:-          False                 NaN   \n",
        "chr20:49552800-49557401:-          False                 NaN   \n",
        "chr20:49552800-49558567:-          False                 NaN   \n",
        "chr20:49557471-49557641:-          False                 NaN   \n",
        "chr20:49557471-49557665:-          False                 NaN   \n",
        "\n",
        "                          downstream_donor_dist upstream_acceptor_dist  \\\n",
        "junction                                                                 \n",
        "chr20:49551774-49552684:-                   NaN                    NaN   \n",
        "chr20:49552800-49557401:-                   NaN                    NaN   \n",
        "chr20:49552800-49558567:-                   NaN                    NaN   \n",
        "chr20:49557471-49557641:-                   NaN                    NaN   \n",
        "chr20:49557471-49557665:-                   NaN                    NaN   \n",
        "\n",
        "                          downstream_acceptor_dist       gene_type  \\\n",
        "junction                                                             \n",
        "chr20:49551774-49552684:-                      NaN  protein_coding   \n",
        "chr20:49552800-49557401:-                      NaN  protein_coding   \n",
        "chr20:49552800-49558567:-                      NaN  protein_coding   \n",
        "chr20:49557471-49557641:-                      NaN  protein_coding   \n",
        "chr20:49557471-49557665:-                      NaN  protein_coding   \n",
        "\n",
        "                           gene_status  gene_name  \n",
        "junction                                           \n",
        "chr20:49551774-49552684:-        KNOWN       DPM1  \n",
        "chr20:49552800-49557401:-        KNOWN       DPM1  \n",
        "chr20:49552800-49558567:-        KNOWN       DPM1  \n",
        "chr20:49557471-49557641:-        KNOWN       DPM1  \n",
        "chr20:49557471-49557665:-        KNOWN       DPM1  \n",
        "\n",
        "[5 rows x 25 columns]\n"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['UM']\n",
      "dest = os.path.join(outdir, \n",
      "                    '{}_dexseq_results.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "if not os.path.exists(dest):\n",
      "    run_dexseq(cancers)\n",
      "else:\n",
      "    res = pd.read_table(dest, header=0,\n",
      "                        index_col=0)\n",
      "    print(dest)\n",
      "    print(res.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['BRCA', 'CLL', 'UM']\n",
      "dest = os.path.join(outdir, \n",
      "                    '{}_dexseq_results.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "if not os.path.exists(dest):\n",
      "    run_dexseq(cancers)\n",
      "else:\n",
      "    res = pd.read_table(dest, header=0,\n",
      "                        index_col=0)\n",
      "    print(dest)\n",
      "    print(res.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['BRCA', 'CLL', 'LUAD', 'LUSC', 'UM']\n",
      "dest = os.path.join(outdir, \n",
      "                    '{}_dexseq_results.tsv'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "if not os.path.exists(dest):\n",
      "    run_dexseq(cancers)\n",
      "else:\n",
      "    res = pd.read_table(dest, header=0,\n",
      "                        index_col=0)\n",
      "    print(dest)\n",
      "    print(res.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "R doesn't like dashes in sample names so I have to\n",
      "fix the TCGA UUIDs after R has had a go at them."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for f in [ds.brca_dexseq_size_factors, ds.cll_dexseq_size_factors,\n",
      "          ds.um_dexseq_size_factors, ds.brca_cll_luad_lusc_um_dexseq_size_factors,\n",
      "          ds.brca_cll_um_dexseq_size_factors]:\n",
      "    s = pd.read_table(f,\n",
      "                      index_col=0,\n",
      "                      header=0)\n",
      "    s.index = [ x.replace('.', '-') for x in s.index ]\n",
      "    s.to_csv(f, sep='\\t')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}