{
 "metadata": {
  "name": "",
  "signature": "sha256:6026bc4801ae533333543a86f1bb7a04c3d905a9e5d54de8ae33b581a05868b0"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Figure 2\n",
      "\n",
      "I'm going to plot the nucleotide frequency of the \n",
      "last 50 bp of the intron for TODO."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "\n",
      "import ds2014 as ds\n",
      "from scipy.stats import fisher_exact\n",
      "import matplotlib.gridspec as gridspec\n",
      "import matplotlib.pyplot as plt\n",
      "import pandas as pd\n",
      "import weblogolib as logo"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rcParams.update({'font.size': 7})\n",
      "rcParams['ps.usedistiller'] = 'xpdf'\n",
      "matplotlib.rc('font',**{'family':'sans-serif','sans-serif':['Arial']})\n",
      "#matplotlib.rc('text', usetex=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "curdir = os.getcwd()\n",
      "outdir = os.path.join(os.path.split(curdir)[0], \n",
      "                      'output', \n",
      "                      'figure02')\n",
      "ds.makedir(outdir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def plot_weblogo(fn, fout):\n",
      "    \"\"\"Make nucleotide frequency plot for fasta file fn\"\"\"\n",
      "    f = open(fn, 'r')\n",
      "    seqs = logo.read_seq_data(f)\n",
      "    data = logo.LogoData.from_seqs(seqs)\n",
      "    options = logo.LogoOptions()\n",
      "    options.fineprint = ''\n",
      "    options.unit_name = 'probability'\n",
      "    #options.logo_title = 'Novel 50 bp upstream of 3\\' SS'\n",
      "    options.logo_title = ''\n",
      "    options.first_index = -50\n",
      "    options.stacks_per_line = 50\n",
      "    options.color_scheme = logo.ColorScheme([logo.ColorGroup(\"TCU\", \"orange\"), \n",
      "                                             logo.ColorGroup(\"AG\", \"blue\")])\n",
      "    options.show_xaxis = False\n",
      "    options.show_yaxis = False\n",
      "    logo_format = logo.LogoFormat(data, options)\n",
      "    fout = open(fout, 'w')\n",
      "    fout.write(logo.png_print_formatter(data, logo_format))\n",
      "    #fout.write(logo.pdf_formatter(data, logo_format))\n",
      "    fout.close()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for thing in [ds.proximal_intron_seq, \n",
      "              ds.annot_intron_seq, \n",
      "              ds.control_intron_seq,\n",
      "              ds.distal_intron_seq]:\n",
      "\n",
      "    fn = os.path.join(ds.root, 'output',\n",
      "                      'figure02',\n",
      "                      thing)\n",
      "    fout = os.path.join(outdir, \n",
      "                        os.path.split(thing)[1][0:-3] + '.png')\n",
      "    plot_weblogo(thing, fout)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def read_fasta(fn):\n",
      "    f = open(fn, 'r')\n",
      "    lines = [ list(x.upper().strip()) for x in f.readlines() ][1::2]\n",
      "    f.close()\n",
      "    df = pd.DataFrame(lines)\n",
      "    df = df.apply(lambda x: x.value_counts()).replace(np.nan, 0).T    \n",
      "    return df\n",
      "\n",
      "def adenine_enrichment(fn, fn2):\n",
      "    \"\"\"\n",
      "    Calculate p-value for enrichment of adenines \n",
      "    at each position for two fasta files using Fisher\n",
      "    exact test\n",
      "    \"\"\"\n",
      "    df = read_fasta(fn)\n",
      "    df2 = read_fasta(fn2)\n",
      "    pvals = []\n",
      "    for i in range(df.shape[0]):\n",
      "        odds, pvalue = fisher_exact([[df.ix[i, 'A'], df.ix[i, ['T', 'C', 'G']].sum()],\n",
      "                                     [df2.ix[i, 'A'], df2.ix[i, ['T', 'C', 'G']].sum()]])\n",
      "        pvals.append(pvalue)\n",
      "    return pvals"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def plot_pvals(test, control):\n",
      "    pvals = adenine_enrichment(test, control)\n",
      "    b = ax.bar(arange(len(pvals)) + 0.1, \n",
      "               -log10(pvals),\n",
      "               width=0.80, color='grey',\n",
      "               lw=0, alpha=0.5)\n",
      "    ax.hlines(-log10(0.05), 0, 50, alpha=0.5, lw=0.5)\n",
      "    for l in ax.get_xticklines() + ax.get_yticklines():\n",
      "        l.set_markersize(0)\n",
      "    #ax.spines['top'].set_visible(False)\n",
      "    #ax.spines['bottom'].set_visible(False)\n",
      "    #xt = ax.set_xticks([])\n",
      "    #yt = ax.set_yticks([])\n",
      "    ds.clean_axis(ax)\n",
      "    return ax\n",
      "\n",
      "def plot_nums(ax):\n",
      "    ds.clean_axis(ax)\n",
      "    xlimits = ax.set_xlim(-50, 1)\n",
      "    xt = ax.set_xticks(arange(0, 50, 5) + 0.5)\n",
      "    ax.set_xticklabels(arange(-50, 0, 5))\n",
      "    ax.set_xlim(0, 50)\n",
      "    ax.tick_params(axis='x', direction='out')\n",
      "    ax.spines['bottom'].set_visible(True)\n",
      "    ax.xaxis.tick_bottom()\n",
      "    return ax\n",
      "\n",
      "fig = plt.figure(figsize=(3, 4.23), dpi=600)\n",
      "\n",
      "pw = 0.1\n",
      "bw = 0.07\n",
      "gs = gridspec.GridSpec(12, 1,\n",
      "                       height_ratios=[0.355, pw, bw, pw, pw, bw,\n",
      "                                      pw, pw, bw, pw, pw, bw],\n",
      "                       hspace=0, wspace=0)\n",
      "\n",
      "t = fig.text(0.005, 0.965, 'a', weight='bold', \n",
      "             size=ds.subfigure_label_fontsize)\n",
      "t = fig.text(0.005, 1 - 0.355 / 1.335, \n",
      "             'b', weight='bold', \n",
      "             size=ds.subfigure_label_fontsize)\n",
      "t = fig.text(0.005, 1 - (0.355 + pw + 1 * bw) / 1.335 - 0.02,\n",
      "             'c', weight='bold', \n",
      "             size=ds.subfigure_label_fontsize)\n",
      "t = fig.text(0.005, 1 - (0.355 + 2 * pw + 3 * bw) / 1.335 - 0.04,\n",
      "             'd', weight='bold', \n",
      "             size=ds.subfigure_label_fontsize)\n",
      "t = fig.text(0.005, 1 - (0.355 + 3 * pw + 5 * bw) / 1.335 - 0.05,\n",
      "             'e', weight='bold', \n",
      "             size=ds.subfigure_label_fontsize)\n",
      "\n",
      "\n",
      "# Diagram to be added in Illustrator\n",
      "ax = fig.add_subplot(gs[0, 0])\n",
      "ds.clean_axis(ax)\n",
      "\n",
      "# Control junctions intron sequence diagram\n",
      "ax = fig.add_subplot(gs[1, 0])\n",
      "ax = plot_nums(ax)\n",
      "\n",
      "# Blank space\n",
      "ax = fig.add_subplot(gs[2, 0])\n",
      "ds.clean_axis(ax)\n",
      "\n",
      "# Annotated junctions with upstream proximal\n",
      "# acceptors adenine enrichment\n",
      "ax = fig.add_subplot(gs[3, 0])\n",
      "ax = plot_pvals(ds.annot_intron_seq, ds.control_intron_seq)\n",
      "\n",
      "# Annotated junctions with upstream proximal\n",
      "# acceptors intron sequence diagram\n",
      "ax = fig.add_subplot(gs[4, 0])\n",
      "ax = plot_nums(ax)\n",
      "\n",
      "# Blank space\n",
      "ax = fig.add_subplot(gs[5, 0])\n",
      "ds.clean_axis(ax)\n",
      "\n",
      "# Proximal cryptic acceptors adenine enrichment\n",
      "ax = fig.add_subplot(gs[6, 0])\n",
      "ax = plot_pvals(ds.proximal_intron_seq, ds.control_intron_seq)\n",
      "\n",
      "# Proximal cryptic acceptors intron sequence diagram\n",
      "ax = fig.add_subplot(gs[7, 0])\n",
      "ax = plot_nums(ax)\n",
      "\n",
      "# Blank space\n",
      "ax = fig.add_subplot(gs[8, 0])\n",
      "ds.clean_axis(ax)\n",
      "\n",
      "# Distal cryptic acceptors adenine enrichment\n",
      "ax = fig.add_subplot(gs[9, 0])\n",
      "ax = plot_pvals(ds.distal_intron_seq, ds.control_intron_seq)\n",
      "\n",
      "# Distal cryptic acceptors intron sequence diagram\n",
      "ax = fig.add_subplot(gs[10, 0])\n",
      "ax = plot_nums(ax)\n",
      "\n",
      "# Blank space\n",
      "ax = fig.add_subplot(gs[11, 0])\n",
      "ds.clean_axis(ax)\n",
      "\n",
      "fig.tight_layout(h_pad=0, w_pad=0)\n",
      "fig.savefig(os.path.join(outdir, 'figure02.pdf'),\n",
      "            transparent=True, dpi=600)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "display_data",
       "png": "iVBORw0KGgoAAAANSUhEUgAAAN0AAAEzCAYAAAC1/QZcAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAACWZJREFUeJzt3E1oXIUax+E3tagVa0QRXXWlYmwsXbgoiBDiQgiC+LEU\nBTMWcS/uQly5ExQXigiahQgupOBGcBGMiAuhdarFD0oiKARFQdoYOoX2bi6lt/ZOks6Zf5IzzwNC\nkvnom6m/vmfOTDJ28eLFiwXE7NnuAWDUiA7CRAdhooMw0UGY6CBMdBAmOgjrG91ff/1V09PTtW/f\nvrr99tvrtddeS80FrdU3ut9++62OHDlSP/zwQz3++OP15ptvpuaC1trb78L9+/dXt9uto0eP1tra\nWp07dy41F7RW3033xhtv1KlTp+rtt9+uAwcO1IULF1JzQWv1je7JJ5+s8+fP15EjR2rfvn119uzZ\n+vPPP1OzQSuN+SkDyPKSAYSJDsJEB2GigzDRQZjoIEx0ECY6CBMdhIkOwkQHYaKDMNFBmOggTHQQ\nJjoIEx2EiQ7CRAdhooMw0UGY6CBMdBAmOggTHYSJDsJEB2GigzDRQZjoIEx0ECY6CBMdhIkOwkQH\nYaKDMNFBmOggTHQQJjoIEx2EiQ7CRAdhooMw0UGY6CBMdBAmOggTHYSJDsJEB2GigzDRQZjoIEx0\nECY6CBMdhIkOwkQHYaKDMNFBmOggTHQQ1je6999/v/bs2VO9Xi81D7TepjbdxYsXhz0HjIxNRdfp\ndGp8fLwee+yxOnPmzLBnglbbVHTT09O1tLRUS0tL9dZbbw17Jmi1vZu50tNPP1379++vgwcP1unT\np4c9E7TapjbdBx98UCdPnqxTp07VxMTEsGeCVusb3djYWN1yyy118uTJevjhh2tqaqpeeuml1GzQ\nSmMXnZqEKC+OQ5joIEx0ECY6CBMdhIkOwkQHYaKDMNFBmOggTHQQJjoIEx2EiQ7CRAdhooMw0UGY\n6CBMdBAmOggTHYSJDsJEB2GigyGZn5+/6tf9slkIs+kgTHQQJjoIEx2EiQ7CRAdhooOwgaObmZlp\nYg4YGQNFt76+Xqurq03NAiNhoHekrK2t1fLyck1OTjY5E7TaNW+6ubm5mp2drW632+Q80HoDbbpe\nr1edTqcWFhaanAlazRueIazv4WWv16vZ2dm67bbb6s4776zXX389NRe01t5+F7777rv18ccf15df\nflnHjx+vY8eO1Ysvvlg33XRTaj622eLi4qWPp6amtm2ONukb3ffff1933313HTp0qA4dOlTPPfdc\nai5orb6Hlw888ED99NNPdfz48frwww9rYmKi/vjjj9Rs0Ep9N12n06lvvvmmpqen68Ybb6xXXnml\n7rjjjtRs0ErOXtKX53TN84ZnCOt7eAlbZTNuzKaDMNFBmOggTHQQ5kRKwy4/kVDV/pMJG504cWLl\n32w6CLPpriL5r7NNMHpsOggbiU036Dbpd/srn8O1nc08uFZEt5v+R9juSLf7sdpNh+7DmrUV0aVt\ndzjsbo1Et7KyUisrK03c1VWdOHHi0seHDx/ue/lGt9/M/V95/Y0+H+TPThv2YzHI51f7ux3EVr/X\nrd5+I/9vOzYW3UZ/0OUGWdtTU1N9b7/R/W318q3e/yCauO+tPLZNPxZNf96kJh6Lpg43h3J4udXh\nrrz+Tj0WT9josWj6e9tpj9VWTlpt9aTWTvle/RArhHmdDsJEB2GigzDRQZjoIEx0ECY6CBMdhIkO\nwkQHYaKDMNFBmOggTHQQJjoIEx2EiQ7CRAdhooMw0UGY6CBMdDAk8/PzV/26X8EHYTYdhIkOwkQH\nYaKDMNFBmOggbODoZmZmmpgDRsZA0a2vr9fq6mpTs8BIGOjF8bW1tVpeXq7JyckmZ4JWu+ZNNzc3\nV7Ozs9XtdpucB1pvoE3X6/Wq0+nUwsJCkzNBq3nvJYR5yQDCthTdXXfdVa+++uqwZmEELS4uXvpv\nVGwpurGxsRobGxvWLDAS+kb3zz//1FNPPVXj4+P1xBNP1Llz51JzQWvt7XfhO++8U5999lktLi7W\n77//XseOHUvNRUtdfhg5NTW1bXNsp77R/fzzz3XvvffWgw8+WFVV4+PjkaGgzfoeXh48eLB+/PHH\n+vrrr+vTTz+tv//+OzUXtFbfTffCCy/U0tJSPfroo/XII4/UPffck5qLlnA4+W99o7v++uvro48+\nSs0CI6FvdJA2CpvRO1IgTHQQ5vCSXWujt441fXja1KGv6GjUKDwnG5To2LGu3GRbjXin/gMgOnaN\npn8SYbuiFB38VypCZy8hzKZjIDv1edPV7JRZbToIs+nYkp2yLXYz0fE/RDV8fgUfhHlOB2GigzDR\nQZjoIEx0ECY6CBMdhIkOwkQHYaKDMNFBmOggTHQQJjoIEx2EiQ7CRAdhooMw0UGY6CBMdBAmOhiS\n+fn5q37dr+CDMJsOwkQHYaKDMNFBmOggTHQQJjoIGzi6mZmZJuaAkTFQdOvr67W6utrULDASBnpH\nytraWi0vL9fk5GSTM0GrXfOmm5ubq9nZ2ep2u03OA6030Kbr9XrV6XRqYWGhyZmg1bzhGcK8ZABh\nooOwvtGdP3++nn/++br11lvrvvvuq88//zw1F7TW3n4Xvvfee/XJJ5/UV199VV988UU988wz9csv\nv9QNN9yQmo+WWVxcvPTx1NTUts2xnfpG1+1268yZM/XQQw/VhQsX6uzZs3X69Om6//77U/NB6/SN\nbmJiosbHx2thYaF+/fXX+u677+rAgQOp2aCV+kZ39OjR+vbbb+vZZ5+t6667rl5++eW6+eabU7NB\nK3mdjijP6bxkAHF9Dy9h2EZx84mOXePyQKt2b6QOLyHMpmOohnn4uFs3n00HYTYdbFJTW1t0NKrJ\nw8krDx/bwuElhNl07Cht3W6Xs+kgzKZjIKP4jpJB7YrodvNf7G59LWk3avqxHtb/d41Et7KyUisr\nK5u+/okTJy59fPjw4S1dfzO3H+T+r2Wey115+36zb3TbYdjqY7eV6w/j80Fs9L1s5fNrme3/hepH\neyDMiRQIEx2EiQ7CRAdhooMw0UGY6CBMdBAmOggTHYSJDsJEB2GigzDRQZjoIEx0ECY6CBMdhIkO\nwkQHYaKDMNHBkMzPz1/1634FH4TZdBAmOggTHYSJDsJEB2Gig7CBo5uZmWliDhgZA0W3vr5eq6ur\nTc0CI2GgF8fX1tZqeXm5Jicnm5wJWu2aN93c3FzNzs5Wt9ttch5ovYE2Xa/Xq06nUwsLC03OBK3m\nvZcQ5iUDCBMdhIkOwkQHYaKDMNFBmOggTHQQ9h9LVe7UAEVacwAAAABJRU5ErkJggg==\n",
       "text": [
        "<matplotlib.figure.Figure at 0x7f6601d47d90>"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}