{
 "metadata": {
  "name": "",
  "signature": "sha256:75c7274b8a8b3bb47b4f7bdcd125f6878c11e8df2a3b8829adf42d1aed8634e1"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Gencode, SJ.out, and eXpress Processing\n",
      "\n",
      "I'm going to make some useful files using the Gencode annotation \n",
      "for downstream use. I'm also going to combine the SJ.out files \n",
      "from STAR and the results files from eXpress. These files can take\n",
      "several minutes to make but we only need to do it once. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import glob\n",
      "import os\n",
      "\n",
      "import cdpybio as cpb\n",
      "import ds2014 as ds\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "curdir = os.getcwd()\n",
      "outdir = os.path.join(os.path.split(curdir)[0], \n",
      "                      'output', \n",
      "                      'gencode_sjout_express_processing')\n",
      "ds.makedir(outdir)\n",
      "\n",
      "meta = pd.read_table('../data/metadata.tsv',\n",
      "                     index_col=0, header=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Gencode\n",
      "\n",
      "We'll parse the Gencode GTF file and make three files so that we can have easier to access to some information later on.\n",
      "\n",
      "#### `gene_info.tsv`\n",
      "\n",
      "A table with the Gencode gene ID, gene type, gene status, and gene name.\n",
      "\n",
      "#### `transcripts_genes.tsv`\n",
      "\n",
      "A table with Gencode transcript IDs and the corresponding Gencode gene ID for the gene that contains that transcript.\n",
      "#### `junction_info.tsv`\n",
      "\n",
      "A table with information about the splice junctions implied by the GTF file. Has the following columns: gene (Gencode gene ID), chrom, start, end, strand, chrom:start, chrom:end, donor, acceptor, intron. Note that the coordinates in this file are one-based and that splice junctions are represented by the one-based coordinates of the first and last base pairs of the junction's intron.\n",
      "\n",
      "It can take several minutes to make these files, but we only have to do it once."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# Get gene info\n",
      "f = os.path.join(outdir, 'gene_info.tsv')\n",
      "if not os.path.exists(f):\n",
      "    print('Making '.format(f))\n",
      "    gene_info = cpb.gencode.make_gene_info_df(ds.gencode_gtf)\n",
      "    gene_info.to_csv(f, sep='\\t')\n",
      "else:\n",
      "    gene_info = pd.read_table(f, index_col=0,\n",
      "                              header=0)\n",
      "print(f)\n",
      "print(gene_info.head())\n",
      "\n",
      "# Get mapping from transcripts to genes\n",
      "f = ds.transcripts_genes\n",
      "if not os.path.exists(f):\n",
      "    print('Making '.format(f))\n",
      "    tg = cpb.gencode.make_transcript_gene_se(ds.gencode_gtf)\n",
      "    tg.to_csv(f, sep='\\t')\n",
      "else:\n",
      "    tg = pd.read_table(f, index_col=0,\n",
      "                       header=0)\n",
      "print(f)\n",
      "print(tg.head())\n",
      "    \n",
      "# Get splice junction information\n",
      "f = ds.junction_info\n",
      "if not os.path.exists(f):\n",
      "    print('Making '.format(f))\n",
      "    jxn_info = cpb.gencode.make_splice_junction_df(ds.gencode_gtf)\n",
      "    jxn_info.to_csv(f, sep='\\t')\n",
      "else:\n",
      "    jxn_info = pd.read_table(f, index_col=0,\n",
      "                             header=0)\n",
      "print(f)\n",
      "print(jxn_info.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/gene_info.tsv\n",
        "                         gene_type gene_status gene_name\n",
        "gene_id                                                 \n",
        "ENSG00000000003.10  protein_coding       KNOWN    TSPAN6\n",
        "ENSG00000000005.5   protein_coding       KNOWN      TNMD\n",
        "ENSG00000000419.8   protein_coding       KNOWN      DPM1\n",
        "ENSG00000000457.8   protein_coding       KNOWN     SCYL3\n",
        "ENSG00000000460.11  protein_coding       KNOWN  C1orf112\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/transcripts_genes.tsv"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "                   ENSG00000223972.4\n",
        "ENST00000456328.2                   \n",
        "ENST00000515242.2  ENSG00000223972.4\n",
        "ENST00000518655.2  ENSG00000223972.4\n",
        "ENST00000450305.2  ENSG00000223972.4\n",
        "ENST00000438504.2  ENSG00000227232.3\n",
        "ENST00000541675.1  ENSG00000227232.3\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/junction_info.tsv"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "                                 gene chrom  start    end strand chrom:start  \\\n",
        "chr1:12228-12612:+  ENSG00000223972.4  chr1  12228  12612      +  chr1:12228   \n",
        "chr1:12722-13220:+  ENSG00000223972.4  chr1  12722  13220      +  chr1:12722   \n",
        "chr1:12722-13224:+  ENSG00000223972.4  chr1  12722  13224      +  chr1:12722   \n",
        "chr1:12228-12594:+  ENSG00000223972.4  chr1  12228  12594      +  chr1:12228   \n",
        "chr1:12722-13402:+  ENSG00000223972.4  chr1  12722  13402      +  chr1:12722   \n",
        "\n",
        "                     chrom:end         donor      acceptor            intron  \n",
        "chr1:12228-12612:+  chr1:12612  chr1:12228:+  chr1:12612:+  chr1:12228-12612  \n",
        "chr1:12722-13220:+  chr1:13220  chr1:12722:+  chr1:13220:+  chr1:12722-13220  \n",
        "chr1:12722-13224:+  chr1:13224  chr1:12722:+  chr1:13224:+  chr1:12722-13224  \n",
        "chr1:12228-12594:+  chr1:12594  chr1:12228:+  chr1:12594:+  chr1:12228-12594  \n",
        "chr1:12722-13402:+  chr1:13402  chr1:12722:+  chr1:13402:+  chr1:12722-13402  \n"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## eXpress\n",
      "\n",
      "Now we'll combine the eXpress output for each sample. We'll make a file that represents gene expression using the mapping of transcripts to genes that we made from the Gencode GTF. The expression of a gene will be defined as the sum of the expression values of the gene's different transcripts."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def define_sample_name(x):\n",
      "    \"\"\"Convert filename into sample name\"\"\"\n",
      "    return os.path.split(x)[1].split('_')[0]\n",
      "\n",
      "f = ds.transcript_eff_counts\n",
      "g = ds.gene_eff_counts\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)):\n",
      "    print('Making {} and {}'.format(f, g))\n",
      "    fns = glob.glob(os.path.join(ds.expressdir,'*xprs'))\n",
      "    t_counts, g_counts = cpb.express.combine_express_output(fns,\n",
      "                                                            tg=ds.transcripts_genes,\n",
      "                                                            define_sample_name=define_sample_name)\n",
      "    t_counts.to_csv(f, sep='\\t')\n",
      "    g_counts.to_csv(g, sep='\\t')\n",
      "else:\n",
      "    t_counts = pd.read_table(f, index_col=0, header=0)\n",
      "    g_counts = pd.read_table(g, index_col=0, header=0)\n",
      "    \n",
      "print(f)\n",
      "print(t_counts.head().T.head().T)\n",
      "print(g)\n",
      "print(g_counts.head().T.head().T)\n",
      "\n",
      "f = ds.transcript_fpkm\n",
      "g = ds.gene_fpkm\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)):\n",
      "    print('Making {} and {}'.format(f, g))\n",
      "    fns = glob.glob(os.path.join(ds.expressdir,'*xprs'))\n",
      "    t_counts, g_counts = cpb.express.combine_express_output(fns, column='fpkm',\n",
      "                                                            tg=ds.transcripts_genes,\n",
      "                                                            define_sample_name=define_sample_name)\n",
      "    t_counts.to_csv(f, sep='\\t')\n",
      "    g_counts.to_csv(g, sep='\\t')\n",
      "else:\n",
      "    t_counts = pd.read_table(f, index_col=0, header=0)\n",
      "    g_counts = pd.read_table(g, index_col=0, header=0)\n",
      "    \n",
      "print(f)\n",
      "print(t_counts.head().T.head().T)\n",
      "print(g)\n",
      "print(g_counts.head().T.head().T)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/transcript_eff_counts.tsv\n",
        "                       SA52609       SA00242      SA65385       SA66418  \\\n",
        "transcript                                                                \n",
        "ENST00000000233.5  4863.395523   5557.168316  4266.391720  12439.941170   \n",
        "ENST00000000412.3  1239.997275  14088.976271  4501.962640   4734.836138   \n",
        "ENST00000000442.6   334.750300    167.046679   139.647167    329.622798   \n",
        "ENST00000001008.4   522.208458    933.814178  2410.700228   1250.319623   \n",
        "ENST00000001146.2     0.000000      0.000000     0.000000      0.000000   \n",
        "\n",
        "                       SA93051  \n",
        "transcript                      \n",
        "ENST00000000233.5  7463.408172  \n",
        "ENST00000000412.3  2681.298419  \n",
        "ENST00000000442.6   153.283357  \n",
        "ENST00000001008.4   340.892025  \n",
        "ENST00000001146.2     0.000000  \n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/gene_eff_counts.tsv\n",
        "                       SA52609      SA00242      SA65385      SA66418  \\\n",
        "gene                                                                    \n",
        "ENSG00000000003.10    0.000000     0.000000     0.000000     0.000000   \n",
        "ENSG00000000005.5     0.000000     0.000000     0.000000     0.000000   \n",
        "ENSG00000000419.8   931.713417  1518.278983  1382.558273  2285.320124   \n",
        "ENSG00000000457.8   175.575633  1469.358940   418.807390   212.445158   \n",
        "ENSG00000000460.11  538.603884   790.647781   296.399321   610.205752   \n",
        "\n",
        "                        SA93051  \n",
        "gene                             \n",
        "ENSG00000000003.10     2.087007  \n",
        "ENSG00000000005.5      0.000000  \n",
        "ENSG00000000419.8   3762.019699  \n",
        "ENSG00000000457.8    416.309420  \n",
        "ENSG00000000460.11   773.438493  \n",
        "Making /raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/transcript_fpkm.tsv and /raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/gene_fpkm.tsv\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/transcript_fpkm.tsv"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "                     SA52609    SA00242    SA65385     SA66418     SA93051\n",
        "transcript                                                                \n",
        "ENST00000000233.5  74.564424  86.331931  67.402668  151.343159  122.268637\n",
        "ENST00000000412.3   7.608676  87.597889  28.465220   23.053970   17.580019\n",
        "ENST00000000442.6   2.555729   1.292283   1.098627    1.996933    1.250474\n",
        "ENST00000001008.4   3.895471   7.058337  18.530347    7.400985    2.717182\n",
        "ENST00000001146.2   0.000000   0.000000   0.000000    0.000000    0.000000\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/gene_fpkm.tsv\n",
        "                      SA52609    SA00242    SA65385    SA66418    SA93051\n",
        "gene                                                                     \n",
        "ENSG00000000003.10   0.000000   0.000000   0.000000   0.000000   0.017095\n",
        "ENSG00000000005.5    0.000000   0.000000   0.000000   0.000000   0.000000\n",
        "ENSG00000000419.8   14.721466  23.515176  21.584587  28.116530  61.936714\n",
        "ENSG00000000457.8    0.950150   9.853850   2.654094   0.999382   2.515710\n",
        "ENSG00000000460.11   2.630035   4.796081   1.717578   2.438147   4.617584\n"
       ]
      }
     ],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## STAR\n",
      "\n",
      "Now we'll combine the `SJ.out.tab` files from STAR. This relies on the [cdpybio python package](https://github.com/cdeboever3/cdpybio). First we combine all of the different `SJ.out.tab` files so that we have a file with the coverage of every junction by uniquely mapped reads. Then we filter novel junctions to keep only those that share a donor or acceptor (or both) with a known Gencode junction.\n",
      "\n",
      "### `Log.final.out` files\n",
      "\n",
      "We'll combine together the `Log.final.out` files so we\n",
      "can look at some alignment statistics."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def define_sample_name(x):\n",
      "    \"\"\"Convert filename into sample name\"\"\"\n",
      "    return os.path.split(x)[1].split('.')[0]\n",
      "\n",
      "if not os.path.exists(ds.star_logs):\n",
      "    logs = cpb.star.make_logs_df(glob.glob('{}/*out'.format(ds.logfinaldir)),\n",
      "                                 define_sample_name=define_sample_name)\n",
      "    logs.to_csv(ds.star_logs, sep='\\t')\n",
      "else:\n",
      "    logs = pd.read_table(ds.star_logs, header=0, index_col=0)\n",
      "print(ds.star_logs)\n",
      "print(logs.head())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/star_logs.tsv\n",
        "            Started job on Started mapping on      Finished on  \\\n",
        "SRR628582  Sep 12 21:18:47    Sep 12 21:33:26  Sep 12 21:41:42   \n",
        "SRR628583  Sep 12 21:21:27    Sep 12 21:35:19  Sep 12 21:45:24   \n",
        "SRR628584  Sep 12 21:15:31    Sep 12 21:20:40  Sep 12 21:27:10   \n",
        "SRR628585  Sep 12 21:20:44    Sep 12 21:31:34  Sep 12 21:41:17   \n",
        "SRR628586  Sep 12 21:15:59    Sep 12 21:23:01  Sep 12 21:29:24   \n",
        "\n",
        "           Mapping speed, Million of reads per hour  Number of input reads  \\\n",
        "SRR628582                                     87.50               12055472   \n",
        "SRR628583                                     84.36               14177111   \n",
        "SRR628584                                     83.57                9053195   \n",
        "SRR628585                                     76.46               12381973   \n",
        "SRR628586                                     83.36                8868661   \n",
        "\n",
        "           Average input read length  Uniquely mapped reads number  \\\n",
        "SRR628582                        201                      11263152   \n",
        "SRR628583                        201                      12256155   \n",
        "SRR628584                        201                       8488674   \n",
        "SRR628585                        201                      10391879   \n",
        "SRR628586                        201                       8308843   \n",
        "\n",
        "           Uniquely mapped reads %  Average mapped length  \\\n",
        "SRR628582                    93.43                 198.96   \n",
        "SRR628583                    86.45                 197.53   \n",
        "SRR628584                    93.76                 199.15   \n",
        "SRR628585                    83.93                 194.66   \n",
        "SRR628586                    93.69                 199.20   \n",
        "\n",
        "           Number of splices: Total            ...             \\\n",
        "SRR628582                   8431629            ...              \n",
        "SRR628583                   6233198            ...              \n",
        "SRR628584                   5469305            ...              \n",
        "SRR628585                   6401648            ...              \n",
        "SRR628586                   5998784            ...              \n",
        "\n",
        "           Deletion average length  Insertion rate per base  \\\n",
        "SRR628582                     1.78                     0.01   \n",
        "SRR628583                     1.70                     0.01   \n",
        "SRR628584                     1.60                     0.01   \n",
        "SRR628585                     1.64                     0.01   \n",
        "SRR628586                     1.72                     0.01   \n",
        "\n",
        "           Insertion average length  Number of reads mapped to multiple loci  \\\n",
        "SRR628582                      1.35                                   789208   \n",
        "SRR628583                      1.36                                  1909825   \n",
        "SRR628584                      1.46                                   561133   \n",
        "SRR628585                      1.35                                  1985225   \n",
        "SRR628586                      1.42                                   557348   \n",
        "\n",
        "           % of reads mapped to multiple loci  \\\n",
        "SRR628582                                6.55   \n",
        "SRR628583                               13.47   \n",
        "SRR628584                                6.20   \n",
        "SRR628585                               16.03   \n",
        "SRR628586                                6.28   \n",
        "\n",
        "           Number of reads mapped to too many loci  \\\n",
        "SRR628582                                       22   \n",
        "SRR628583                                      176   \n",
        "SRR628584                                       45   \n",
        "SRR628585                                       68   \n",
        "SRR628586                                       29   \n",
        "\n",
        "           % of reads mapped to too many loci  \\\n",
        "SRR628582                                   0   \n",
        "SRR628583                                   0   \n",
        "SRR628584                                   0   \n",
        "SRR628585                                   0   \n",
        "SRR628586                                   0   \n",
        "\n",
        "           % of reads unmapped: too many mismatches  \\\n",
        "SRR628582                                         0   \n",
        "SRR628583                                         0   \n",
        "SRR628584                                         0   \n",
        "SRR628585                                         0   \n",
        "SRR628586                                         0   \n",
        "\n",
        "           % of reads unmapped: too short  % of reads unmapped: other  \n",
        "SRR628582                               0                        0.03  \n",
        "SRR628583                               0                        0.07  \n",
        "SRR628584                               0                        0.04  \n",
        "SRR628585                               0                        0.04  \n",
        "SRR628586                               0                        0.03  \n",
        "\n",
        "[5 rows x 27 columns]\n"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "### `SJ.out.tab` files\n",
      "\n",
      "#### Junction filtering\n",
      "\n",
      "##### Unique read coverage\n",
      "\n",
      "`SJ.out.tab` files contain many unannotated junctions with low coverage. Many of these are likely noise or cannot be used in downstream analyses due to the low coverage. When combining together the `SJ.out.tab` files, we will only keep junctions that have at least 20 uniquely mapped junctions-spanning reads summed over all samples. So if a junction has 20 unique junction-spanning reads in one sample, we keep it. If a junction has one unique junction-spanning read in 20 different samples, we keep it. Other junctions are discarded. Since some analyses will only focus on a particular subset of the cancers I'm including in the study, we'll combine `SJ.out.tab` for each different subset since the coverage filtering will behave differently for different groups of samples.\n",
      "\n",
      "##### Novel junctions\n",
      "\n",
      "There are many novel (potential) junctions in the `SJ.out.tab` files. To enrich for novel junctions that are likely real (as opposed to technical noise or something) we'll only keep novel junctions that share a donor acceptor with a known Gencode junction. Note that both a novel junction's donor and acceptor could be in Gencode but not annotated together as a splice site."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def define_sample_name(x):\n",
      "    \"\"\"Convert filename into sample name\"\"\"\n",
      "    return os.path.split(x)[1].split('.SJ')[0]\n",
      "\n",
      "def combine_sj_out_files(cancers, meta=meta, outdir=outdir, \n",
      "                         define_sample_name=define_sample_name):\n",
      "    \"\"\"Run cpb.star.combine_sj_out for all samples for a subset of cancers\"\"\"\n",
      "    samples = list(meta[meta.disease_abbr.apply(lambda x: x in cancers)].index)\n",
      "    fns = [ os.path.join(ds.sjoutdir, '{}.SJ.out.tab'.format(x)) for x in samples ]\n",
      "    sf = os.path.join(outdir, '{}_combine_sj_out_stats.txt'.format('_'.join([ x.lower() for x in cancers ])))\n",
      "    counts, annot, stats = cpb.star.combine_sj_out(fns, \n",
      "                                                   ds.junction_info,\n",
      "                                                   define_sample_name=define_sample_name)\n",
      "    return counts, annot, stats"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['UM']\n",
      "f = ds.um_sj_counts\n",
      "g = ds.um_sj_annot\n",
      "h = ds.um_sj_stats\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)) or (not os.path.exists(h)):\n",
      "    print('Combining junctions for {}'.format(', '.join(cancers)))\n",
      "    counts, annot, stats = combine_sj_out_files(cancers)\n",
      "    counts.to_csv(f, sep='\\t')\n",
      "    annot.to_csv(g, sep='\\t')\n",
      "    f = open(h, 'w')\n",
      "    f.write('\\n'.join(stats) + '\\n')\n",
      "    f.close()\n",
      "else:\n",
      "    counts = pd.read_table(f, index_col=0, header=0)\n",
      "    annot = pd.read_table(g, index_col=0, header=0)\n",
      "    t = open(h)\n",
      "    stats = [ x.strip() for x in t.readlines() ]\n",
      "    t.close()\n",
      "print(f)\n",
      "print(counts.head())\n",
      "print(g)\n",
      "print(annot.head())\n",
      "print(h)\n",
      "print('\\n'.join(stats))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/um_sj_counts.tsv\n",
        "                          SRR628582  SRR628583  SRR628584  SRR628585  \\\n",
        "chrX:99884984-99885755:-         75         57         31         41   \n",
        "chrX:99885864-99887481:-         55         32         26         27   \n",
        "chrX:99887566-99888401:-         49         47         23         30   \n",
        "chrX:99888537-99888927:-         68         45         20         37   \n",
        "chrX:99889027-99890174:-         72         30         20         29   \n",
        "\n",
        "                          SRR628586  SRR628587  SRR628588  SRR628589  \n",
        "chrX:99884984-99885755:-         27         76         48         43  \n",
        "chrX:99885864-99887481:-         21         41         38         31  \n",
        "chrX:99887566-99888401:-         24         48         31         20  \n",
        "chrX:99888537-99888927:-         20         52         40         25  \n",
        "chrX:99889027-99890174:-         32         50         26         17  \n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/um_sj_annot.tsv\n",
        "                         chrom     start       end strand intron_motif  \\\n",
        "chrX:99884984-99885755:-  chrX  99884984  99885755      -        CT/AC   \n",
        "chrX:99885864-99887481:-  chrX  99885864  99887481      -        CT/AC   \n",
        "chrX:99887566-99888401:-  chrX  99887566  99888401      -        CT/AC   \n",
        "chrX:99888537-99888927:-  chrX  99888537  99888927      -        CT/AC   \n",
        "chrX:99889027-99890174:-  chrX  99889027  99890174      -        CT/AC   \n",
        "\n",
        "                         annotated ext_annotated    chrom:start  \\\n",
        "chrX:99884984-99885755:-      True          True  chrX:99884984   \n",
        "chrX:99885864-99887481:-      True          True  chrX:99885864   \n",
        "chrX:99887566-99888401:-      True          True  chrX:99887566   \n",
        "chrX:99888537-99888927:-      True          True  chrX:99888537   \n",
        "chrX:99889027-99890174:-      True          True  chrX:99889027   \n",
        "\n",
        "                              chrom:end             gene_id            donor  \\\n",
        "chrX:99884984-99885755:-  chrX:99885755  ENSG00000000003.10  chrX:99885755:-   \n",
        "chrX:99885864-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99887566-99888401:-  chrX:99888401  ENSG00000000003.10  chrX:99888401:-   \n",
        "chrX:99888537-99888927:-  chrX:99888927  ENSG00000000003.10  chrX:99888927:-   \n",
        "chrX:99889027-99890174:-  chrX:99890174  ENSG00000000003.10  chrX:99890174:-   \n",
        "\n",
        "                                 acceptor novel_donor novel_acceptor  \\\n",
        "chrX:99884984-99885755:-  chrX:99884984:-       False          False   \n",
        "chrX:99885864-99887481:-  chrX:99885864:-       False          False   \n",
        "chrX:99887566-99888401:-  chrX:99887566:-       False          False   \n",
        "chrX:99888537-99888927:-  chrX:99888537:-       False          False   \n",
        "chrX:99889027-99890174:-  chrX:99889027:-       False          False   \n",
        "\n",
        "                          upstream_donor_dist  downstream_donor_dist  \\\n",
        "chrX:99884984-99885755:-                  NaN                    NaN   \n",
        "chrX:99885864-99887481:-                  NaN                    NaN   \n",
        "chrX:99887566-99888401:-                  NaN                    NaN   \n",
        "chrX:99888537-99888927:-                  NaN                    NaN   \n",
        "chrX:99889027-99890174:-                  NaN                    NaN   \n",
        "\n",
        "                          upstream_acceptor_dist  downstream_acceptor_dist  \n",
        "chrX:99884984-99885755:-                     NaN                       NaN  \n",
        "chrX:99885864-99887481:-                     NaN                       NaN  \n",
        "chrX:99887566-99888401:-                     NaN                       NaN  \n",
        "chrX:99888537-99888927:-                     NaN                       NaN  \n",
        "chrX:99889027-99890174:-                     NaN                       NaN  \n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/um_sj_stats.tsv\n",
        "Number of junctions in SJ.out file per sample\n",
        "SRR628589\t149,532\n",
        "SRR628588\t155,879\n",
        "SRR628583\t184,484\n",
        "SRR628582\t176,456\n",
        "SRR628587\t180,954\n",
        "SRR628586\t171,557\n",
        "SRR628585\t171,575\n",
        "SRR628584\t165,142\n",
        "\n",
        "SJ.out panel size\t(8, 114426, 3)\n",
        "\n",
        "External database stats\n",
        "Read external annotation\t/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/junction_info.tsv\n",
        "Total number of junctions\t346,021\n",
        "Number of junctions used in only one gene\t335,976\n",
        "\n",
        "Junction filtering stats\n",
        "Number of junctions in external annotation\t335,976\n",
        "Number observed junctions in STAR annotation\t110,393\n",
        "Number observed junctions in external annotation\t107,595\n",
        "Number of observed junctions not in external annotation\t3,447\n",
        "Number of observed junctions not in external annotation but in STAR annotation\t2,798\n",
        "Number of observed junctions not in external annotation and not in STAR annotation\t649\n",
        "Number of novel donors\t1,163\n",
        "Number of novel junctions with novel donors\t1,181\n",
        "Number of novel acceptors\t1,452\n",
        "Number of novel junctions with novel acceptors\t1,469\n",
        "Number of novel junctions with new combination of donor and acceptor\t797\n"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['CLL']\n",
      "f = ds.cll_sj_counts\n",
      "g = ds.cll_sj_annot\n",
      "h = ds.cll_sj_stats\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)) or (not os.path.exists(h)):\n",
      "    print('Combining junctions for {}'.format(', '.join(cancers)))\n",
      "    counts, annot, stats = combine_sj_out_files(cancers)\n",
      "    counts.to_csv(f, sep='\\t')\n",
      "    annot.to_csv(g, sep='\\t')\n",
      "    f = open(h, 'w')\n",
      "    f.write('\\n'.join(stats) + '\\n')\n",
      "    f.close()\n",
      "else:\n",
      "    counts = pd.read_table(f, index_col=0, header=0)\n",
      "    annot = pd.read_table(g, index_col=0, header=0)\n",
      "    t = open(h)\n",
      "    stats = [ x.strip() for x in t.readlines() ]\n",
      "    t.close()\n",
      "print(f)\n",
      "print(counts.head())\n",
      "print(g)\n",
      "print(annot.head())\n",
      "print(h)\n",
      "print('\\n'.join(stats))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/cll_sj_counts.tsv\n",
        "                           SA00242  SA03982  SA05603  SA15699  SA22083  \\\n",
        "chr20:49551774-49552684:-      158      158      262      199      339   \n",
        "chr20:49552800-49557401:-      117      108      108      104      192   \n",
        "chr20:49552800-49558567:-        4        8        1        3        4   \n",
        "chr20:49557471-49557641:-        1        9        0        9       13   \n",
        "chr20:49557471-49557665:-        8        8        0        6        8   \n",
        "\n",
        "                           SA24823  SA29677  SA30914  SA52314  SA52609  \\\n",
        "chr20:49551774-49552684:-      532      336       45       92      194   \n",
        "chr20:49552800-49557401:-      356      222       20       37       72   \n",
        "chr20:49552800-49558567:-       30        6        0        0        3   \n",
        "chr20:49557471-49557641:-        0        0        2        1        0   \n",
        "chr20:49557471-49557665:-        0        0        0        0        1   \n",
        "\n",
        "                           SA58773  SA65385  SA66418  SA83613  SA93051  \\\n",
        "chr20:49551774-49552684:-       68      118      235      268      248   \n",
        "chr20:49552800-49557401:-       33       52      114      137      162   \n",
        "chr20:49552800-49558567:-        2        4        8        2        9   \n",
        "chr20:49557471-49557641:-        0        3        0        3        0   \n",
        "chr20:49557471-49557665:-        2        4        1        3        0   \n",
        "\n",
        "                           SA95730  \n",
        "chr20:49551774-49552684:-      271  \n",
        "chr20:49552800-49557401:-      209  \n",
        "chr20:49552800-49558567:-        6  \n",
        "chr20:49557471-49557641:-        6  \n",
        "chr20:49557471-49557665:-       10  \n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/cll_sj_annot.tsv\n",
        "                           chrom     start       end strand intron_motif  \\\n",
        "chr20:49551774-49552684:-  chr20  49551774  49552684      -        CT/AC   \n",
        "chr20:49552800-49557401:-  chr20  49552800  49557401      -        CT/AC   \n",
        "chr20:49552800-49558567:-  chr20  49552800  49558567      -        CT/AC   \n",
        "chr20:49557471-49557641:-  chr20  49557471  49557641      -        CT/AC   \n",
        "chr20:49557471-49557665:-  chr20  49557471  49557665      -        CT/AC   \n",
        "\n",
        "                          annotated ext_annotated     chrom:start  \\\n",
        "chr20:49551774-49552684:-      True          True  chr20:49551774   \n",
        "chr20:49552800-49557401:-      True          True  chr20:49552800   \n",
        "chr20:49552800-49558567:-      True         False  chr20:49552800   \n",
        "chr20:49557471-49557641:-      True          True  chr20:49557471   \n",
        "chr20:49557471-49557665:-      True          True  chr20:49557471   \n",
        "\n",
        "                                chrom:end            gene_id  \\\n",
        "chr20:49551774-49552684:-  chr20:49552684  ENSG00000000419.8   \n",
        "chr20:49552800-49557401:-  chr20:49557401  ENSG00000000419.8   \n",
        "chr20:49552800-49558567:-  chr20:49558567  ENSG00000000419.8   \n",
        "chr20:49557471-49557641:-  chr20:49557641  ENSG00000000419.8   \n",
        "chr20:49557471-49557665:-  chr20:49557665  ENSG00000000419.8   \n",
        "\n",
        "                                      donor          acceptor novel_donor  \\\n",
        "chr20:49551774-49552684:-  chr20:49552684:-  chr20:49551774:-       False   \n",
        "chr20:49552800-49557401:-  chr20:49557401:-  chr20:49552800:-       False   \n",
        "chr20:49552800-49558567:-  chr20:49558567:-  chr20:49552800:-       False   \n",
        "chr20:49557471-49557641:-  chr20:49557641:-  chr20:49557471:-       False   \n",
        "chr20:49557471-49557665:-  chr20:49557665:-  chr20:49557471:-       False   \n",
        "\n",
        "                          novel_acceptor  upstream_donor_dist  \\\n",
        "chr20:49551774-49552684:-          False                  NaN   \n",
        "chr20:49552800-49557401:-          False                  NaN   \n",
        "chr20:49552800-49558567:-          False                  NaN   \n",
        "chr20:49557471-49557641:-          False                  NaN   \n",
        "chr20:49557471-49557665:-          False                  NaN   \n",
        "\n",
        "                           downstream_donor_dist  upstream_acceptor_dist  \\\n",
        "chr20:49551774-49552684:-                    NaN                     NaN   \n",
        "chr20:49552800-49557401:-                    NaN                     NaN   \n",
        "chr20:49552800-49558567:-                    NaN                     NaN   \n",
        "chr20:49557471-49557641:-                    NaN                     NaN   \n",
        "chr20:49557471-49557665:-                    NaN                     NaN   \n",
        "\n",
        "                           downstream_acceptor_dist  \n",
        "chr20:49551774-49552684:-                       NaN  \n",
        "chr20:49552800-49557401:-                       NaN  \n",
        "chr20:49552800-49558567:-                       NaN  \n",
        "chr20:49557471-49557641:-                       NaN  \n",
        "chr20:49557471-49557665:-                       NaN  "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/cll_sj_stats.tsv\n",
        "Number of junctions in SJ.out file per sample\n",
        "SA95730\t351,754\n",
        "SA24823\t303,139\n",
        "SA22083\t445,161\n",
        "SA15699\t442,822\n",
        "SA52314\t348,728\n",
        "SA05603\t444,627\n",
        "SA30914\t473,430\n",
        "SA52609\t399,459\n",
        "SA83613\t528,432\n",
        "SA65385\t432,073\n",
        "SA03982\t353,861\n",
        "SA93051\t416,274\n",
        "SA29677\t350,773\n",
        "SA00242\t340,552\n",
        "SA66418\t442,140\n",
        "SA58773\t313,381\n",
        "\n",
        "SJ.out panel size\t(16, 246436, 3)\n",
        "\n",
        "External database stats\n",
        "Read external annotation\t/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/junction_info.tsv\n",
        "Total number of junctions\t346,021\n",
        "Number of junctions used in only one gene\t335,976\n",
        "\n",
        "Junction filtering stats\n",
        "Number of junctions in external annotation\t335,976\n",
        "Number observed junctions in STAR annotation\t173,725\n",
        "Number observed junctions in external annotation\t156,755\n",
        "Number of observed junctions not in external annotation\t60,582\n",
        "Number of observed junctions not in external annotation but in STAR annotation\t16,970\n",
        "Number of observed junctions not in external annotation and not in STAR annotation\t43,612\n",
        "Number of novel donors\t17,188\n",
        "Number of novel junctions with novel donors\t18,177\n",
        "Number of novel acceptors\t23,920\n",
        "Number of novel junctions with novel acceptors\t25,229\n",
        "Number of novel junctions with new combination of donor and acceptor\t17,176\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['BRCA']\n",
      "f = ds.brca_sj_counts\n",
      "g = ds.brca_sj_annot\n",
      "h = ds.brca_sj_stats\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)) or (not os.path.exists(h)):\n",
      "    print('Combining junctions for {}'.format(', '.join(cancers)))\n",
      "    counts, annot, stats = combine_sj_out_files(cancers)\n",
      "    counts.to_csv(f, sep='\\t')\n",
      "    annot.to_csv(g, sep='\\t')\n",
      "    f = open(h, 'w')\n",
      "    f.write('\\n'.join(stats) + '\\n')\n",
      "    f.close()\n",
      "else:\n",
      "    counts = pd.read_table(f, index_col=0, header=0)\n",
      "    annot = pd.read_table(g, index_col=0, header=0)\n",
      "    t = open(h)\n",
      "    stats = [ x.strip() for x in t.readlines() ]\n",
      "    t.close()\n",
      "print(f)\n",
      "print(counts.head())\n",
      "print(g)\n",
      "print(annot.head())\n",
      "print(h)\n",
      "print('\\n'.join(stats))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_sj_counts.tsv\n",
        "                          01e63570-aa28-4169-81da-76e838845e78  \\\n",
        "chrX:99884984-99885755:-                                   468   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   388   \n",
        "chrX:99887566-99888401:-                                   402   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          04f38ef0-e3f5-452a-bc19-e5ae76a666ca  \\\n",
        "chrX:99884984-99885755:-                                   251   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   194   \n",
        "chrX:99887566-99888401:-                                   268   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          059067a2-7bf7-4360-b564-3c94319ac2e2  \\\n",
        "chrX:99884984-99885755:-                                    93   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    85   \n",
        "chrX:99887566-99888401:-                                   108   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          09004786-790a-48c9-b3ab-76e92f2e69dd  \\\n",
        "chrX:99884984-99885755:-                                   614   \n",
        "chrX:99884984-99887481:-                                     9   \n",
        "chrX:99885864-99887481:-                                   563   \n",
        "chrX:99887566-99888401:-                                   568   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          12bd8213-844d-4b11-8036-d547d4112447  \\\n",
        "chrX:99884984-99885755:-                                   306   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   211   \n",
        "chrX:99887566-99888401:-                                   203   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          26835e9c-11f9-476f-9d6b-ebc0ae3442e8  \\\n",
        "chrX:99884984-99885755:-                                   456   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   406   \n",
        "chrX:99887566-99888401:-                                   468   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          3ed7f6c6-5ab0-40cd-8da2-124a7451d287  \\\n",
        "chrX:99884984-99885755:-                                    60   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                    47   \n",
        "chrX:99887566-99888401:-                                    50   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          40dd04c6-ce5d-4c30-aaff-86ece9c78507  \\\n",
        "chrX:99884984-99885755:-                                   241   \n",
        "chrX:99884984-99887481:-                                     3   \n",
        "chrX:99885864-99887481:-                                   181   \n",
        "chrX:99887566-99888401:-                                   222   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          42857e49-6357-4fc2-966e-50c8d9ba841f  \\\n",
        "chrX:99884984-99885755:-                                    84   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    65   \n",
        "chrX:99887566-99888401:-                                    41   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          4c9289e0-378e-4293-877c-e12d207f384e  \\\n",
        "chrX:99884984-99885755:-                                   308   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   277   \n",
        "chrX:99887566-99888401:-                                   291   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                                          ...                   \\\n",
        "chrX:99884984-99885755:-                  ...                    \n",
        "chrX:99884984-99887481:-                  ...                    \n",
        "chrX:99885864-99887481:-                  ...                    \n",
        "chrX:99887566-99888401:-                  ...                    \n",
        "chrX:99887566-99888927:-                  ...                    \n",
        "\n",
        "                          b6b7fb9f-129f-4baa-961b-5715adb22a62  \\\n",
        "chrX:99884984-99885755:-                                   151   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   118   \n",
        "chrX:99887566-99888401:-                                   112   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          bb496a29-fd16-4ffe-a000-4e70ea9aa76c  \\\n",
        "chrX:99884984-99885755:-                                   153   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   132   \n",
        "chrX:99887566-99888401:-                                   112   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          c52c3859-08f8-4f50-8596-bfa497a0d24f  \\\n",
        "chrX:99884984-99885755:-                                   258   \n",
        "chrX:99884984-99887481:-                                     1   \n",
        "chrX:99885864-99887481:-                                   205   \n",
        "chrX:99887566-99888401:-                                   283   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          cd37724f-9a98-4676-b9a5-e8d6999db1f5  \\\n",
        "chrX:99884984-99885755:-                                   284   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   246   \n",
        "chrX:99887566-99888401:-                                   296   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          d217d14a-2132-433c-bcac-ccab9d296724  \\\n",
        "chrX:99884984-99885755:-                                   219   \n",
        "chrX:99884984-99887481:-                                     3   \n",
        "chrX:99885864-99887481:-                                   178   \n",
        "chrX:99887566-99888401:-                                   202   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d38df7db-3c58-4865-ae5a-e1e0b80201f1  \\\n",
        "chrX:99884984-99885755:-                                   110   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                    67   \n",
        "chrX:99887566-99888401:-                                    69   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d3e5466a-4861-4eb9-b5e9-f1b851afd72b  \\\n",
        "chrX:99884984-99885755:-                                   216   \n",
        "chrX:99884984-99887481:-                                     1   \n",
        "chrX:99885864-99887481:-                                   172   \n",
        "chrX:99887566-99888401:-                                   196   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d6b6eb3b-00bd-412d-a610-bc474e0cc319  \\\n",
        "chrX:99884984-99885755:-                                   111   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                   125   \n",
        "chrX:99887566-99888401:-                                   131   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          dfb5fd44-421f-473b-8aef-c7d275c5adcb  \\\n",
        "chrX:99884984-99885755:-                                    84   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    57   \n",
        "chrX:99887566-99888401:-                                    91   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          fde0f3c2-c9db-4a98-89f9-53685c482160  \n",
        "chrX:99884984-99885755:-                                   455  \n",
        "chrX:99884984-99887481:-                                     1  \n",
        "chrX:99885864-99887481:-                                   335  \n",
        "chrX:99887566-99888401:-                                   396  \n",
        "chrX:99887566-99888927:-                                     1  \n",
        "\n",
        "[5 rows x 32 columns]\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_sj_annot.tsv\n",
        "                         chrom     start       end strand intron_motif  \\\n",
        "chrX:99884984-99885755:-  chrX  99884984  99885755      -        CT/AC   \n",
        "chrX:99884984-99887481:-  chrX  99884984  99887481      -        CT/AC   \n",
        "chrX:99885864-99887481:-  chrX  99885864  99887481      -        CT/AC   \n",
        "chrX:99887566-99888401:-  chrX  99887566  99888401      -        CT/AC   \n",
        "chrX:99887566-99888927:-  chrX  99887566  99888927      -        CT/AC   \n",
        "\n",
        "                         annotated ext_annotated    chrom:start  \\\n",
        "chrX:99884984-99885755:-      True          True  chrX:99884984   \n",
        "chrX:99884984-99887481:-      True         False  chrX:99884984   \n",
        "chrX:99885864-99887481:-      True          True  chrX:99885864   \n",
        "chrX:99887566-99888401:-      True          True  chrX:99887566   \n",
        "chrX:99887566-99888927:-      True         False  chrX:99887566   \n",
        "\n",
        "                              chrom:end             gene_id            donor  \\\n",
        "chrX:99884984-99885755:-  chrX:99885755  ENSG00000000003.10  chrX:99885755:-   \n",
        "chrX:99884984-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99885864-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99887566-99888401:-  chrX:99888401  ENSG00000000003.10  chrX:99888401:-   \n",
        "chrX:99887566-99888927:-  chrX:99888927  ENSG00000000003.10  chrX:99888927:-   \n",
        "\n",
        "                                 acceptor novel_donor novel_acceptor  \\\n",
        "chrX:99884984-99885755:-  chrX:99884984:-       False          False   \n",
        "chrX:99884984-99887481:-  chrX:99884984:-       False          False   \n",
        "chrX:99885864-99887481:-  chrX:99885864:-       False          False   \n",
        "chrX:99887566-99888401:-  chrX:99887566:-       False          False   \n",
        "chrX:99887566-99888927:-  chrX:99887566:-       False          False   \n",
        "\n",
        "                          upstream_donor_dist  downstream_donor_dist  \\\n",
        "chrX:99884984-99885755:-                  NaN                    NaN   \n",
        "chrX:99884984-99887481:-                  NaN                    NaN   \n",
        "chrX:99885864-99887481:-                  NaN                    NaN   \n",
        "chrX:99887566-99888401:-                  NaN                    NaN   \n",
        "chrX:99887566-99888927:-                  NaN                    NaN   \n",
        "\n",
        "                          upstream_acceptor_dist  downstream_acceptor_dist  \n",
        "chrX:99884984-99885755:-                     NaN                       NaN  \n",
        "chrX:99884984-99887481:-                     NaN                       NaN  \n",
        "chrX:99885864-99887481:-                     NaN                       NaN  \n",
        "chrX:99887566-99888401:-                     NaN                       NaN  \n",
        "chrX:99887566-99888927:-                     NaN                       NaN  "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_sj_stats.tsv\n",
        "Number of junctions in SJ.out file per sample\n",
        "71263324-796a-4f08-bba6-ecb2ae483f5a\t262,712\n",
        "12bd8213-844d-4b11-8036-d547d4112447\t231,151\n",
        "09004786-790a-48c9-b3ab-76e92f2e69dd\t265,179\n",
        "3ed7f6c6-5ab0-40cd-8da2-124a7451d287\t272,056\n",
        "01e63570-aa28-4169-81da-76e838845e78\t269,335\n",
        "766bbaa9-e498-471f-9f72-63a143e6656b\t234,419\n",
        "40dd04c6-ce5d-4c30-aaff-86ece9c78507\t285,565\n",
        "d6b6eb3b-00bd-412d-a610-bc474e0cc319\t223,252\n",
        "a4097518-431c-40b9-a540-2fc0d8f76e17\t259,856\n",
        "ac27f4b6-daeb-4cf8-b868-54a70d30778b\t261,004\n",
        "fde0f3c2-c9db-4a98-89f9-53685c482160\t274,022\n",
        "6004e153-e459-44c9-b3e7-5474bac6b953\t244,289\n",
        "5726682f-cb9d-4ccb-a5c7-0664d3ad0881\t252,655\n",
        "6685b1ee-a9b3-4f67-95c5-507847dbcd63\t267,549\n",
        "545c799d-6416-40a2-86e0-a132ed4885b3\t231,074\n",
        "059067a2-7bf7-4360-b564-3c94319ac2e2\t271,313\n",
        "646cf9ce-e8e9-4637-b843-2c6908a93a8f\t317,360\n",
        "4c9289e0-378e-4293-877c-e12d207f384e\t270,185\n",
        "b6b7fb9f-129f-4baa-961b-5715adb22a62\t287,795\n",
        "cd37724f-9a98-4676-b9a5-e8d6999db1f5\t307,235\n",
        "d3e5466a-4861-4eb9-b5e9-f1b851afd72b\t306,106\n",
        "42857e49-6357-4fc2-966e-50c8d9ba841f\t199,911\n",
        "bb496a29-fd16-4ffe-a000-4e70ea9aa76c\t255,117\n",
        "26835e9c-11f9-476f-9d6b-ebc0ae3442e8\t270,073\n",
        "c52c3859-08f8-4f50-8596-bfa497a0d24f\t252,050\n",
        "a0fbd146-2f78-4b9c-bc09-4f4ccd112373\t260,124\n",
        "b48874ac-eaff-4ddb-b99b-1b96076914b1\t249,224\n",
        "04f38ef0-e3f5-452a-bc19-e5ae76a666ca\t297,439\n",
        "d217d14a-2132-433c-bcac-ccab9d296724\t258,875\n",
        "824425ea-7871-479f-9344-328cab4f26ad\t313,432\n",
        "d38df7db-3c58-4865-ae5a-e1e0b80201f1\t283,866\n",
        "dfb5fd44-421f-473b-8aef-c7d275c5adcb\t266,624\n",
        "\n",
        "SJ.out panel size\t(32, 250210, 3)\n",
        "\n",
        "External database stats\n",
        "Read external annotation\t/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/junction_info.tsv\n",
        "Total number of junctions\t346,021\n",
        "Number of junctions used in only one gene\t335,976\n",
        "\n",
        "Junction filtering stats\n",
        "Number of junctions in external annotation\t335,976\n",
        "Number observed junctions in STAR annotation\t218,194\n",
        "Number observed junctions in external annotation\t200,332\n",
        "Number of observed junctions not in external annotation\t33,375\n",
        "Number of observed junctions not in external annotation but in STAR annotation\t17,862\n",
        "Number of observed junctions not in external annotation and not in STAR annotation\t15,513\n",
        "Number of novel donors\t10,846\n",
        "Number of novel junctions with novel donors\t11,109\n",
        "Number of novel acceptors\t13,510\n",
        "Number of novel junctions with novel acceptors\t13,792\n",
        "Number of novel junctions with new combination of donor and acceptor\t8,474\n"
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['BRCA', 'CLL', 'UM']\n",
      "f = ds.brca_cll_um_sj_counts\n",
      "g = ds.brca_cll_um_sj_annot\n",
      "h = ds.brca_cll_um_sj_stats\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)) or (not os.path.exists(h)):\n",
      "    print('Combining junctions for {}'.format(', '.join(cancers)))\n",
      "    counts, annot, stats = combine_sj_out_files(cancers)\n",
      "    counts.to_csv(f, sep='\\t')\n",
      "    annot.to_csv(g, sep='\\t')\n",
      "    f = open(h, 'w')\n",
      "    f.write('\\n'.join(stats) + '\\n')\n",
      "    f.close()\n",
      "else:\n",
      "    counts = pd.read_table(f, index_col=0, header=0)\n",
      "    annot = pd.read_table(g, index_col=0, header=0)\n",
      "    t = open(h)\n",
      "    stats = [ x.strip() for x in t.readlines() ]\n",
      "    t.close()\n",
      "print(f)\n",
      "print(counts.head())\n",
      "print(g)\n",
      "print(annot.head())\n",
      "print(h)\n",
      "print('\\n'.join(stats))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_cll_um_sj_counts.tsv\n",
        "                          01e63570-aa28-4169-81da-76e838845e78  \\\n",
        "chrX:99884984-99885755:-                                   468   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   388   \n",
        "chrX:99887566-99888401:-                                   402   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          04f38ef0-e3f5-452a-bc19-e5ae76a666ca  \\\n",
        "chrX:99884984-99885755:-                                   251   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   194   \n",
        "chrX:99887566-99888401:-                                   268   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          059067a2-7bf7-4360-b564-3c94319ac2e2  \\\n",
        "chrX:99884984-99885755:-                                    93   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    85   \n",
        "chrX:99887566-99888401:-                                   108   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          09004786-790a-48c9-b3ab-76e92f2e69dd  \\\n",
        "chrX:99884984-99885755:-                                   614   \n",
        "chrX:99884984-99887481:-                                     9   \n",
        "chrX:99885864-99887481:-                                   563   \n",
        "chrX:99887566-99888401:-                                   568   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          12bd8213-844d-4b11-8036-d547d4112447  \\\n",
        "chrX:99884984-99885755:-                                   306   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   211   \n",
        "chrX:99887566-99888401:-                                   203   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          26835e9c-11f9-476f-9d6b-ebc0ae3442e8  \\\n",
        "chrX:99884984-99885755:-                                   456   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   406   \n",
        "chrX:99887566-99888401:-                                   468   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          3ed7f6c6-5ab0-40cd-8da2-124a7451d287  \\\n",
        "chrX:99884984-99885755:-                                    60   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                    47   \n",
        "chrX:99887566-99888401:-                                    50   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          40dd04c6-ce5d-4c30-aaff-86ece9c78507  \\\n",
        "chrX:99884984-99885755:-                                   241   \n",
        "chrX:99884984-99887481:-                                     3   \n",
        "chrX:99885864-99887481:-                                   181   \n",
        "chrX:99887566-99888401:-                                   222   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          42857e49-6357-4fc2-966e-50c8d9ba841f  \\\n",
        "chrX:99884984-99885755:-                                    84   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    65   \n",
        "chrX:99887566-99888401:-                                    41   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          4c9289e0-378e-4293-877c-e12d207f384e  \\\n",
        "chrX:99884984-99885755:-                                   308   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   277   \n",
        "chrX:99887566-99888401:-                                   291   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                                          ...                   \\\n",
        "chrX:99884984-99885755:-                  ...                    \n",
        "chrX:99884984-99887481:-                  ...                    \n",
        "chrX:99885864-99887481:-                  ...                    \n",
        "chrX:99887566-99888401:-                  ...                    \n",
        "chrX:99887566-99888927:-                  ...                    \n",
        "\n",
        "                          b6b7fb9f-129f-4baa-961b-5715adb22a62  \\\n",
        "chrX:99884984-99885755:-                                   151   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   118   \n",
        "chrX:99887566-99888401:-                                   112   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          bb496a29-fd16-4ffe-a000-4e70ea9aa76c  \\\n",
        "chrX:99884984-99885755:-                                   153   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   132   \n",
        "chrX:99887566-99888401:-                                   112   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          c52c3859-08f8-4f50-8596-bfa497a0d24f  \\\n",
        "chrX:99884984-99885755:-                                   258   \n",
        "chrX:99884984-99887481:-                                     1   \n",
        "chrX:99885864-99887481:-                                   205   \n",
        "chrX:99887566-99888401:-                                   283   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          cd37724f-9a98-4676-b9a5-e8d6999db1f5  \\\n",
        "chrX:99884984-99885755:-                                   284   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   246   \n",
        "chrX:99887566-99888401:-                                   296   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          d217d14a-2132-433c-bcac-ccab9d296724  \\\n",
        "chrX:99884984-99885755:-                                   219   \n",
        "chrX:99884984-99887481:-                                     3   \n",
        "chrX:99885864-99887481:-                                   178   \n",
        "chrX:99887566-99888401:-                                   202   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d38df7db-3c58-4865-ae5a-e1e0b80201f1  \\\n",
        "chrX:99884984-99885755:-                                   110   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                    67   \n",
        "chrX:99887566-99888401:-                                    69   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d3e5466a-4861-4eb9-b5e9-f1b851afd72b  \\\n",
        "chrX:99884984-99885755:-                                   216   \n",
        "chrX:99884984-99887481:-                                     1   \n",
        "chrX:99885864-99887481:-                                   172   \n",
        "chrX:99887566-99888401:-                                   196   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d6b6eb3b-00bd-412d-a610-bc474e0cc319  \\\n",
        "chrX:99884984-99885755:-                                   111   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                   125   \n",
        "chrX:99887566-99888401:-                                   131   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          dfb5fd44-421f-473b-8aef-c7d275c5adcb  \\\n",
        "chrX:99884984-99885755:-                                    84   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    57   \n",
        "chrX:99887566-99888401:-                                    91   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          fde0f3c2-c9db-4a98-89f9-53685c482160  \n",
        "chrX:99884984-99885755:-                                   455  \n",
        "chrX:99884984-99887481:-                                     1  \n",
        "chrX:99885864-99887481:-                                   335  \n",
        "chrX:99887566-99888401:-                                   396  \n",
        "chrX:99887566-99888927:-                                     1  \n",
        "\n",
        "[5 rows x 56 columns]\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_cll_um_sj_annot.tsv\n",
        "                         chrom     start       end strand intron_motif  \\\n",
        "chrX:99884984-99885755:-  chrX  99884984  99885755      -        CT/AC   \n",
        "chrX:99884984-99887481:-  chrX  99884984  99887481      -        CT/AC   \n",
        "chrX:99885864-99887481:-  chrX  99885864  99887481      -        CT/AC   \n",
        "chrX:99887566-99888401:-  chrX  99887566  99888401      -        CT/AC   \n",
        "chrX:99887566-99888927:-  chrX  99887566  99888927      -        CT/AC   \n",
        "\n",
        "                         annotated ext_annotated    chrom:start  \\\n",
        "chrX:99884984-99885755:-      True          True  chrX:99884984   \n",
        "chrX:99884984-99887481:-      True         False  chrX:99884984   \n",
        "chrX:99885864-99887481:-      True          True  chrX:99885864   \n",
        "chrX:99887566-99888401:-      True          True  chrX:99887566   \n",
        "chrX:99887566-99888927:-      True         False  chrX:99887566   \n",
        "\n",
        "                              chrom:end             gene_id            donor  \\\n",
        "chrX:99884984-99885755:-  chrX:99885755  ENSG00000000003.10  chrX:99885755:-   \n",
        "chrX:99884984-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99885864-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99887566-99888401:-  chrX:99888401  ENSG00000000003.10  chrX:99888401:-   \n",
        "chrX:99887566-99888927:-  chrX:99888927  ENSG00000000003.10  chrX:99888927:-   \n",
        "\n",
        "                                 acceptor novel_donor novel_acceptor  \\\n",
        "chrX:99884984-99885755:-  chrX:99884984:-       False          False   \n",
        "chrX:99884984-99887481:-  chrX:99884984:-       False          False   \n",
        "chrX:99885864-99887481:-  chrX:99885864:-       False          False   \n",
        "chrX:99887566-99888401:-  chrX:99887566:-       False          False   \n",
        "chrX:99887566-99888927:-  chrX:99887566:-       False          False   \n",
        "\n",
        "                          upstream_donor_dist  downstream_donor_dist  \\\n",
        "chrX:99884984-99885755:-                  NaN                    NaN   \n",
        "chrX:99884984-99887481:-                  NaN                    NaN   \n",
        "chrX:99885864-99887481:-                  NaN                    NaN   \n",
        "chrX:99887566-99888401:-                  NaN                    NaN   \n",
        "chrX:99887566-99888927:-                  NaN                    NaN   \n",
        "\n",
        "                          upstream_acceptor_dist  downstream_acceptor_dist  \n",
        "chrX:99884984-99885755:-                     NaN                       NaN  \n",
        "chrX:99884984-99887481:-                     NaN                       NaN  \n",
        "chrX:99885864-99887481:-                     NaN                       NaN  \n",
        "chrX:99887566-99888401:-                     NaN                       NaN  \n",
        "chrX:99887566-99888927:-                     NaN                       NaN  "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_cll_um_sj_stats.tsv\n",
        "Number of junctions in SJ.out file per sample\n",
        "SRR628585\t171,575\n",
        "71263324-796a-4f08-bba6-ecb2ae483f5a\t262,712\n",
        "12bd8213-844d-4b11-8036-d547d4112447\t231,151\n",
        "09004786-790a-48c9-b3ab-76e92f2e69dd\t265,179\n",
        "SA52609\t399,459\n",
        "3ed7f6c6-5ab0-40cd-8da2-124a7451d287\t272,056\n",
        "SRR628589\t149,532\n",
        "SRR628588\t155,879\n",
        "SA03982\t353,861\n",
        "01e63570-aa28-4169-81da-76e838845e78\t269,335\n",
        "766bbaa9-e498-471f-9f72-63a143e6656b\t234,419\n",
        "SA93051\t416,274\n",
        "SA29677\t350,773\n",
        "SA66418\t442,140\n",
        "SRR628586\t171,557\n",
        "40dd04c6-ce5d-4c30-aaff-86ece9c78507\t285,565\n",
        "SRR628584\t165,142\n",
        "d6b6eb3b-00bd-412d-a610-bc474e0cc319\t223,252\n",
        "SA15699\t442,822\n",
        "SA22083\t445,161\n",
        "SA52314\t348,728\n",
        "a4097518-431c-40b9-a540-2fc0d8f76e17\t259,856\n",
        "ac27f4b6-daeb-4cf8-b868-54a70d30778b\t261,004\n",
        "fde0f3c2-c9db-4a98-89f9-53685c482160\t274,022\n",
        "6004e153-e459-44c9-b3e7-5474bac6b953\t244,289\n",
        "5726682f-cb9d-4ccb-a5c7-0664d3ad0881\t252,655\n",
        "SA65385\t432,073\n",
        "6685b1ee-a9b3-4f67-95c5-507847dbcd63\t267,549\n",
        "545c799d-6416-40a2-86e0-a132ed4885b3\t231,074\n",
        "SA58773\t313,381\n",
        "SA00242\t340,552\n",
        "059067a2-7bf7-4360-b564-3c94319ac2e2\t271,313\n",
        "646cf9ce-e8e9-4637-b843-2c6908a93a8f\t317,360\n",
        "SRR628587\t180,954\n",
        "SA30914\t473,430\n",
        "4c9289e0-378e-4293-877c-e12d207f384e\t270,185\n",
        "SA83613\t528,432\n",
        "b6b7fb9f-129f-4baa-961b-5715adb22a62\t287,795\n",
        "cd37724f-9a98-4676-b9a5-e8d6999db1f5\t307,235\n",
        "d3e5466a-4861-4eb9-b5e9-f1b851afd72b\t306,106\n",
        "42857e49-6357-4fc2-966e-50c8d9ba841f\t199,911\n",
        "SRR628583\t184,484\n",
        "bb496a29-fd16-4ffe-a000-4e70ea9aa76c\t255,117\n",
        "26835e9c-11f9-476f-9d6b-ebc0ae3442e8\t270,073\n",
        "SA95730\t351,754\n",
        "SA24823\t303,139\n",
        "c52c3859-08f8-4f50-8596-bfa497a0d24f\t252,050\n",
        "a0fbd146-2f78-4b9c-bc09-4f4ccd112373\t260,124\n",
        "b48874ac-eaff-4ddb-b99b-1b96076914b1\t249,224\n",
        "SA05603\t444,627\n",
        "04f38ef0-e3f5-452a-bc19-e5ae76a666ca\t297,439\n",
        "d217d14a-2132-433c-bcac-ccab9d296724\t258,875\n",
        "824425ea-7871-479f-9344-328cab4f26ad\t313,432\n",
        "d38df7db-3c58-4865-ae5a-e1e0b80201f1\t283,866\n",
        "SRR628582\t176,456\n",
        "dfb5fd44-421f-473b-8aef-c7d275c5adcb\t266,624\n",
        "\n",
        "SJ.out panel size\t(56, 352228, 3)\n",
        "\n",
        "External database stats\n",
        "Read external annotation\t/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/junction_info.tsv\n",
        "Total number of junctions\t346,021\n",
        "Number of junctions used in only one gene\t335,976\n",
        "\n",
        "Junction filtering stats\n",
        "Number of junctions in external annotation\t335,976\n",
        "Number observed junctions in STAR annotation\t247,316\n",
        "Number observed junctions in external annotation\t221,868\n",
        "Number of observed junctions not in external annotation\t88,286\n",
        "Number of observed junctions not in external annotation but in STAR annotation\t25,448\n",
        "Number of observed junctions not in external annotation and not in STAR annotation\t62,838\n",
        "Number of novel donors\t26,460\n",
        "Number of novel junctions with novel donors\t27,793\n",
        "Number of novel acceptors\t34,842\n",
        "Number of novel junctions with novel acceptors\t36,581\n",
        "Number of novel junctions with new combination of donor and acceptor\t23,912\n"
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cancers = ['BRCA', 'CLL', 'LUAD', 'LUSC', 'UM']\n",
      "f = ds.brca_cll_luad_lusc_um_sj_counts\n",
      "g = ds.brca_cll_luad_lusc_um_sj_annot\n",
      "h = ds.brca_cll_luad_lusc_um_sj_stats\n",
      "if (not os.path.exists(f)) or (not os.path.exists(g)) or (not os.path.exists(h)):\n",
      "    print('Combining junctions for {}'.format(', '.join(cancers)))\n",
      "    counts, annot, stats = combine_sj_out_files(cancers)\n",
      "    counts.to_csv(f, sep='\\t')\n",
      "    annot.to_csv(g, sep='\\t')\n",
      "    f = open(h, 'w')\n",
      "    f.write('\\n'.join(stats) + '\\n')\n",
      "    f.close()\n",
      "else:\n",
      "    counts = pd.read_table(f, index_col=0, header=0)\n",
      "    annot = pd.read_table(g, index_col=0, header=0)\n",
      "    t = open(h)\n",
      "    stats = [ x.strip() for x in t.readlines() ]\n",
      "    t.close()\n",
      "print(f)\n",
      "print(counts.head())\n",
      "print(g)\n",
      "print(annot.head())\n",
      "print(h)\n",
      "print('\\n'.join(stats))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_cll_luad_lusc_um_sj_counts.tsv\n",
        "                          01e63570-aa28-4169-81da-76e838845e78  \\\n",
        "chrX:99884984-99885755:-                                   468   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   388   \n",
        "chrX:99887566-99888401:-                                   402   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          04f38ef0-e3f5-452a-bc19-e5ae76a666ca  \\\n",
        "chrX:99884984-99885755:-                                   251   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   194   \n",
        "chrX:99887566-99888401:-                                   268   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          059067a2-7bf7-4360-b564-3c94319ac2e2  \\\n",
        "chrX:99884984-99885755:-                                    93   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    85   \n",
        "chrX:99887566-99888401:-                                   108   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          09004786-790a-48c9-b3ab-76e92f2e69dd  \\\n",
        "chrX:99884984-99885755:-                                   614   \n",
        "chrX:99884984-99887481:-                                     9   \n",
        "chrX:99885864-99887481:-                                   563   \n",
        "chrX:99887566-99888401:-                                   568   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          12bd8213-844d-4b11-8036-d547d4112447  \\\n",
        "chrX:99884984-99885755:-                                   306   \n",
        "chrX:99884984-99887481:-                                     2   \n",
        "chrX:99885864-99887481:-                                   211   \n",
        "chrX:99887566-99888401:-                                   203   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          1a6b6ef8-736e-424e-bcb4-b1640c15bf0e  \\\n",
        "chrX:99884984-99885755:-                                  1159   \n",
        "chrX:99884984-99887481:-                                    36   \n",
        "chrX:99885864-99887481:-                                  1025   \n",
        "chrX:99887566-99888401:-                                   995   \n",
        "chrX:99887566-99888927:-                                    41   \n",
        "\n",
        "                          1c764fb7-631c-4425-8083-7cd27dcbc809  \\\n",
        "chrX:99884984-99885755:-                                   290   \n",
        "chrX:99884984-99887481:-                                    18   \n",
        "chrX:99885864-99887481:-                                   253   \n",
        "chrX:99887566-99888401:-                                   232   \n",
        "chrX:99887566-99888927:-                                     5   \n",
        "\n",
        "                          20883c55-f271-44ea-a7d2-8806e02a088f  \\\n",
        "chrX:99884984-99885755:-                                   127   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   111   \n",
        "chrX:99887566-99888401:-                                   146   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          22e1471d-75ca-4259-a147-a667914e2820  \\\n",
        "chrX:99884984-99885755:-                                   390   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   275   \n",
        "chrX:99887566-99888401:-                                   330   \n",
        "chrX:99887566-99888927:-                                     5   \n",
        "\n",
        "                          23712cb6-d37f-463d-8c24-99066067c283  \\\n",
        "chrX:99884984-99885755:-                                   290   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   280   \n",
        "chrX:99887566-99888401:-                                   259   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                                          ...                   \\\n",
        "chrX:99884984-99885755:-                  ...                    \n",
        "chrX:99884984-99887481:-                  ...                    \n",
        "chrX:99885864-99887481:-                  ...                    \n",
        "chrX:99887566-99888401:-                  ...                    \n",
        "chrX:99887566-99888927:-                  ...                    \n",
        "\n",
        "                          d3e5466a-4861-4eb9-b5e9-f1b851afd72b  \\\n",
        "chrX:99884984-99885755:-                                   216   \n",
        "chrX:99884984-99887481:-                                     1   \n",
        "chrX:99885864-99887481:-                                   172   \n",
        "chrX:99887566-99888401:-                                   196   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          d6b6eb3b-00bd-412d-a610-bc474e0cc319  \\\n",
        "chrX:99884984-99885755:-                                   111   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                   125   \n",
        "chrX:99887566-99888401:-                                   131   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          dfb5fd44-421f-473b-8aef-c7d275c5adcb  \\\n",
        "chrX:99884984-99885755:-                                    84   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    57   \n",
        "chrX:99887566-99888401:-                                    91   \n",
        "chrX:99887566-99888927:-                                     0   \n",
        "\n",
        "                          ead4b73b-6a51-4863-b24c-44e0fff1bbce  \\\n",
        "chrX:99884984-99885755:-                                   115   \n",
        "chrX:99884984-99887481:-                                     0   \n",
        "chrX:99885864-99887481:-                                    76   \n",
        "chrX:99887566-99888401:-                                    74   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          eaedb35d-ec72-45df-8790-18deafa143b8  \\\n",
        "chrX:99884984-99885755:-                                   381   \n",
        "chrX:99884984-99887481:-                                     6   \n",
        "chrX:99885864-99887481:-                                   322   \n",
        "chrX:99887566-99888401:-                                   289   \n",
        "chrX:99887566-99888927:-                                     5   \n",
        "\n",
        "                          f1d78984-f956-4ed0-9420-63a9e3de989b  \\\n",
        "chrX:99884984-99885755:-                                   292   \n",
        "chrX:99884984-99887481:-                                     5   \n",
        "chrX:99885864-99887481:-                                   238   \n",
        "chrX:99887566-99888401:-                                   250   \n",
        "chrX:99887566-99888927:-                                     2   \n",
        "\n",
        "                          faba33e9-25ba-4bd5-af3d-d266e714ffa1  \\\n",
        "chrX:99884984-99885755:-                                   201   \n",
        "chrX:99884984-99887481:-                                     4   \n",
        "chrX:99885864-99887481:-                                   148   \n",
        "chrX:99887566-99888401:-                                   170   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          fde0f3c2-c9db-4a98-89f9-53685c482160  \\\n",
        "chrX:99884984-99885755:-                                   455   \n",
        "chrX:99884984-99887481:-                                     1   \n",
        "chrX:99885864-99887481:-                                   335   \n",
        "chrX:99887566-99888401:-                                   396   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          ff780a21-8949-4bd9-8354-7f636cbfe744  \\\n",
        "chrX:99884984-99885755:-                                   161   \n",
        "chrX:99884984-99887481:-                                     3   \n",
        "chrX:99885864-99887481:-                                   142   \n",
        "chrX:99887566-99888401:-                                   202   \n",
        "chrX:99887566-99888927:-                                     1   \n",
        "\n",
        "                          fff72492-a90f-4699-a1d7-8e9186bd3cd6  \n",
        "chrX:99884984-99885755:-                                   401  \n",
        "chrX:99884984-99887481:-                                     9  \n",
        "chrX:99885864-99887481:-                                   270  \n",
        "chrX:99887566-99888401:-                                   263  \n",
        "chrX:99887566-99888927:-                                     4  \n",
        "\n",
        "[5 rows x 81 columns]\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_cll_luad_lusc_um_sj_annot.tsv\n",
        "                         chrom     start       end strand intron_motif  \\\n",
        "chrX:99884984-99885755:-  chrX  99884984  99885755      -        CT/AC   \n",
        "chrX:99884984-99887481:-  chrX  99884984  99887481      -        CT/AC   \n",
        "chrX:99885864-99887481:-  chrX  99885864  99887481      -        CT/AC   \n",
        "chrX:99887566-99888401:-  chrX  99887566  99888401      -        CT/AC   \n",
        "chrX:99887566-99888927:-  chrX  99887566  99888927      -        CT/AC   \n",
        "\n",
        "                         annotated ext_annotated    chrom:start  \\\n",
        "chrX:99884984-99885755:-      True          True  chrX:99884984   \n",
        "chrX:99884984-99887481:-      True         False  chrX:99884984   \n",
        "chrX:99885864-99887481:-      True          True  chrX:99885864   \n",
        "chrX:99887566-99888401:-      True          True  chrX:99887566   \n",
        "chrX:99887566-99888927:-      True         False  chrX:99887566   \n",
        "\n",
        "                              chrom:end             gene_id            donor  \\\n",
        "chrX:99884984-99885755:-  chrX:99885755  ENSG00000000003.10  chrX:99885755:-   \n",
        "chrX:99884984-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99885864-99887481:-  chrX:99887481  ENSG00000000003.10  chrX:99887481:-   \n",
        "chrX:99887566-99888401:-  chrX:99888401  ENSG00000000003.10  chrX:99888401:-   \n",
        "chrX:99887566-99888927:-  chrX:99888927  ENSG00000000003.10  chrX:99888927:-   \n",
        "\n",
        "                                 acceptor novel_donor novel_acceptor  \\\n",
        "chrX:99884984-99885755:-  chrX:99884984:-       False          False   \n",
        "chrX:99884984-99887481:-  chrX:99884984:-       False          False   \n",
        "chrX:99885864-99887481:-  chrX:99885864:-       False          False   \n",
        "chrX:99887566-99888401:-  chrX:99887566:-       False          False   \n",
        "chrX:99887566-99888927:-  chrX:99887566:-       False          False   \n",
        "\n",
        "                          upstream_donor_dist  downstream_donor_dist  \\\n",
        "chrX:99884984-99885755:-                  NaN                    NaN   \n",
        "chrX:99884984-99887481:-                  NaN                    NaN   \n",
        "chrX:99885864-99887481:-                  NaN                    NaN   \n",
        "chrX:99887566-99888401:-                  NaN                    NaN   \n",
        "chrX:99887566-99888927:-                  NaN                    NaN   \n",
        "\n",
        "                          upstream_acceptor_dist  downstream_acceptor_dist  \n",
        "chrX:99884984-99885755:-                     NaN                       NaN  \n",
        "chrX:99884984-99887481:-                     NaN                       NaN  \n",
        "chrX:99885864-99887481:-                     NaN                       NaN  \n",
        "chrX:99887566-99888401:-                     NaN                       NaN  \n",
        "chrX:99887566-99888927:-                     NaN                       NaN  "
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/brca_cll_luad_lusc_um_sj_stats.tsv\n",
        "Number of junctions in SJ.out file per sample\n",
        "SRR628585\t171,575\n",
        "71263324-796a-4f08-bba6-ecb2ae483f5a\t262,712\n",
        "a81337f5-e346-42f0-87a2-9aab7dc1e5b4\t251,770\n",
        "3e2f068a-8235-44c6-97cb-bbf8916a3949\t282,648\n",
        "fff72492-a90f-4699-a1d7-8e9186bd3cd6\t285,402\n",
        "SRR628582\t176,456\n",
        "12bd8213-844d-4b11-8036-d547d4112447\t231,151\n",
        "09004786-790a-48c9-b3ab-76e92f2e69dd\t265,179\n",
        "SA52609\t399,459\n",
        "48f12aef-e572-4cd8-9c5a-edda7e5385aa\t229,956\n",
        "3ed7f6c6-5ab0-40cd-8da2-124a7451d287\t272,056\n",
        "SRR628589\t149,532\n",
        "SRR628588\t155,879\n",
        "SA03982\t353,861\n",
        "01e63570-aa28-4169-81da-76e838845e78\t269,335\n",
        "766bbaa9-e498-471f-9f72-63a143e6656b\t234,419\n",
        "SA93051\t416,274\n",
        "SA29677\t350,773\n",
        "ead4b73b-6a51-4863-b24c-44e0fff1bbce\t205,404\n",
        "SA66418\t442,140\n",
        "319489b3-2148-4cec-93fb-0540d6cda6bf\t225,398\n",
        "40dd04c6-ce5d-4c30-aaff-86ece9c78507\t285,565\n",
        "SRR628584\t165,142\n",
        "23712cb6-d37f-463d-8c24-99066067c283\t255,075\n",
        "d6b6eb3b-00bd-412d-a610-bc474e0cc319\t223,252\n",
        "SRR628587\t180,954\n",
        "SA15699\t442,822\n",
        "SA22083\t445,161\n",
        "SA52314\t348,728\n",
        "a4097518-431c-40b9-a540-2fc0d8f76e17\t259,856\n",
        "ac27f4b6-daeb-4cf8-b868-54a70d30778b\t261,004\n",
        "fde0f3c2-c9db-4a98-89f9-53685c482160\t274,022\n",
        "SRR628586\t171,557\n",
        "1c764fb7-631c-4425-8083-7cd27dcbc809\t264,482\n",
        "6004e153-e459-44c9-b3e7-5474bac6b953\t244,289\n",
        "5726682f-cb9d-4ccb-a5c7-0664d3ad0881\t252,655\n",
        "SA65385\t432,073\n",
        "22e1471d-75ca-4259-a147-a667914e2820\t227,360\n",
        "6685b1ee-a9b3-4f67-95c5-507847dbcd63\t267,549\n",
        "545c799d-6416-40a2-86e0-a132ed4885b3\t231,074\n",
        "SA58773\t313,381\n",
        "SA00242\t340,552\n",
        "059067a2-7bf7-4360-b564-3c94319ac2e2\t271,313\n",
        "775f1aa3-d8aa-42a4-928f-61defc5b5134\t237,025\n",
        "646cf9ce-e8e9-4637-b843-2c6908a93a8f\t317,360\n",
        "3e82732e-8b47-4cb4-9035-22f9cdbc10c6\t178,878\n",
        "f1d78984-f956-4ed0-9420-63a9e3de989b\t311,631\n",
        "870fe86d-4cb7-46a9-8d3e-2057995e763c\t224,593\n",
        "SA30914\t473,430\n",
        "4c9289e0-378e-4293-877c-e12d207f384e\t270,185\n",
        "SA83613\t528,432\n",
        "63b2f6d2-2f86-4986-bcc1-9e1aae13f8bd\t260,288\n",
        "ff780a21-8949-4bd9-8354-7f636cbfe744\t234,476\n",
        "a3a2cddb-ba5f-4dbd-a51a-555e5d758fb3\t281,115\n",
        "b6b7fb9f-129f-4baa-961b-5715adb22a62\t287,795\n",
        "cd37724f-9a98-4676-b9a5-e8d6999db1f5\t307,235\n",
        "d3e5466a-4861-4eb9-b5e9-f1b851afd72b\t306,106\n",
        "42857e49-6357-4fc2-966e-50c8d9ba841f\t199,911\n",
        "20883c55-f271-44ea-a7d2-8806e02a088f\t168,951\n",
        "SRR628583\t184,484\n",
        "5b043b12-41f4-4bd9-89eb-57e7f27cfabf\t279,308\n",
        "bb496a29-fd16-4ffe-a000-4e70ea9aa76c\t255,117\n",
        "26835e9c-11f9-476f-9d6b-ebc0ae3442e8\t270,073\n",
        "SA95730\t351,754\n",
        "SA24823\t303,139\n",
        "c52c3859-08f8-4f50-8596-bfa497a0d24f\t252,050\n",
        "bc0af03f-aa40-4b39-8876-6f2e7397181c\t272,787\n",
        "eaedb35d-ec72-45df-8790-18deafa143b8\t302,145\n",
        "a0fbd146-2f78-4b9c-bc09-4f4ccd112373\t260,124\n",
        "b48874ac-eaff-4ddb-b99b-1b96076914b1\t249,224\n",
        "1a6b6ef8-736e-424e-bcb4-b1640c15bf0e\t299,083\n",
        "SA05603\t444,627\n",
        "04f38ef0-e3f5-452a-bc19-e5ae76a666ca\t297,439\n",
        "95350f94-dc26-45e6-8183-28e9c7250ade\t206,208\n",
        "a35a1390-32a3-4b30-a132-3aac37c1f8cb\t307,482\n",
        "d217d14a-2132-433c-bcac-ccab9d296724\t258,875\n",
        "824425ea-7871-479f-9344-328cab4f26ad\t313,432\n",
        "d38df7db-3c58-4865-ae5a-e1e0b80201f1\t283,866\n",
        "faba33e9-25ba-4bd5-af3d-d266e714ffa1\t215,430\n",
        "4c4b2906-f082-4fcc-83e0-039e163ee8cc\t225,288\n",
        "dfb5fd44-421f-473b-8aef-c7d275c5adcb\t266,624\n",
        "\n",
        "SJ.out panel size\t(81, 386833, 3)\n",
        "\n",
        "External database stats\n",
        "Read external annotation\t/raid2/projects/CLL/sf3b1/deboever-sf3b1-2014/output/gencode_sjout_express_processing/junction_info.tsv\n",
        "Total number of junctions\t346,021\n",
        "Number of junctions used in only one gene\t335,976\n",
        "\n",
        "Junction filtering stats\n",
        "Number of junctions in external annotation\t335,976\n",
        "Number observed junctions in STAR annotation\t264,120\n",
        "Number observed junctions in external annotation\t235,862\n",
        "Number of observed junctions not in external annotation\t102,678\n",
        "Number of observed junctions not in external annotation but in STAR annotation\t28,258\n",
        "Number of observed junctions not in external annotation and not in STAR annotation\t74,420\n",
        "Number of novel donors\t31,449\n",
        "Number of novel junctions with novel donors\t32,994\n",
        "Number of novel acceptors\t40,217\n",
        "Number of novel junctions with novel acceptors\t42,217\n",
        "Number of novel junctions with new combination of donor and acceptor\t27,467\n"
       ]
      }
     ],
     "prompt_number": 11
    }
   ],
   "metadata": {}
  }
 ]
}