{
 "metadata": {
  "name": "",
  "signature": "sha256:15c7fe0f92240092693480b1a980ada7f65e54c4492aca6977948662e646ad6a"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Supporting Files\n",
      "\n",
      "This notebook creates the supplementary files from the paper."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "\n",
      "import ds2014 as ds\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "curdir = os.getcwd()\n",
      "outdir = os.path.join(os.path.split(curdir)[0], \n",
      "                      'output', \n",
      "                      'supporting_files')\n",
      "ds.makedir(outdir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "meta = pd.read_table(ds.metadata, header=0,\n",
      "                     index_col=0)\n",
      "\n",
      "gene_info = pd.read_table(ds.gene_info, header=0,\n",
      "                          index_col=0)\n",
      "\n",
      "# cancer census genes from COSMIC\n",
      "cancer_gene_census = pd.read_table(ds.cancer_gene_census,\n",
      "                                   index_col=0)\n",
      "# cancer genes from NCG4.0 (http://ncg.kcl.ac.uk/)\n",
      "cancer_genes = pd.read_table(ds.cancer_genes,\n",
      "                             index_col=1, header=0)\n",
      "# tumor suppressors from TSGene (http://bioinfo.mc.vanderbilt.edu/TSGene/)\n",
      "ts_genes = pd.read_table(ds.ts_genes,\n",
      "                         index_col=1, header=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S1\n",
      "\n",
      "Samples used in this study. SF3B1 mutated samples have \n",
      "columns for frequency of SF3B1 mutation in RNA-seq data, \n",
      "mutation type, codon change and whether the mutation is \n",
      "in the HEAT 5-9 repeats. These columns are empty for \n",
      "SF3B1 wild-type tumor samples."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sf3b1_mut_info = pd.read_table(ds.sf3b1_mut_info,\n",
      "                               header=0, index_col=None)\n",
      "samples = sf3b1_mut_info[['name', 'disease_abbr',\n",
      "                           'alt_freq', 'mut_type',\n",
      "                           'codon_change', 'HEAT']]\n",
      "samples['alt_freq'] = samples.alt_freq.apply(lambda x: '{:.2f}'.format(x)).values\n",
      "samples.columns = ['Sample Name', 'Cancer', 'Alternate Allele Frequency',\n",
      "                    'Mutation Type', 'Codon Change', 'In HEAT 5-9']\n",
      "samples['SF3B1 mutated'] = True\n",
      "wt = pd.DataFrame(columns=samples.columns)\n",
      "wt['Sample Name'] = meta[meta.condition == 'WT'].index\n",
      "wt['SF3B1 mutated'] = False\n",
      "wt['Cancer'] = meta.ix[wt['Sample Name'], 'disease_abbr'].values\n",
      "samples = pd.concat([samples, wt])\n",
      "samples.to_csv(os.path.join(outdir, 'file_s01.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(samples.shape[0]))\n",
      "samples.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "82 rows\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "-c:6: SettingWithCopyWarning: A value is trying to be set on a copy of a slice from a DataFrame.\n",
        "Try using .loc[row_index,col_indexer] = value instead\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Sample Name</th>\n",
        "      <th>Cancer</th>\n",
        "      <th>Alternate Allele Frequency</th>\n",
        "      <th>Mutation Type</th>\n",
        "      <th>Codon Change</th>\n",
        "      <th>In HEAT 5-9</th>\n",
        "      <th>SF3B1 mutated</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> SA30914</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.56</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> SA95730</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.46</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> SA29677</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.52</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> SA03982</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.51</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> SA93051</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.50</td>\n",
        "      <td> missense</td>\n",
        "      <td> K666E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 4,
       "text": [
        "  Sample Name Cancer Alternate Allele Frequency Mutation Type Codon Change  \\\n",
        "0     SA30914    CLL                       0.56      missense        K700E   \n",
        "1     SA95730    CLL                       0.46      missense        K700E   \n",
        "2     SA29677    CLL                       0.52      missense        K700E   \n",
        "3     SA03982    CLL                       0.51      missense        K700E   \n",
        "4     SA93051    CLL                       0.50      missense        K666E   \n",
        "\n",
        "  In HEAT 5-9 SF3B1 mutated  \n",
        "0        True          True  \n",
        "1        True          True  \n",
        "2        True          True  \n",
        "3        True          True  \n",
        "4        True          True  "
       ]
      }
     ],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S2\n",
      "\n",
      "Summary of differential junction usage results from DEXSeq. \n",
      "DEXSeq was used to test for differential splice junction \n",
      "usage in a joint analysis of the CLL, BRCA, and UM samples \n",
      "as well as individually for each cancer type. \u201cNovel\u201d indicates \n",
      "that the junction is not annotated in Gencode. Proximal \n",
      "indicates that a novel 3\u2019SS is 10-30 bp upstream of a canonical \n",
      "Gencode 3\u2019SS. In this paper, we focused mainly on the proximal \n",
      "cryptic 3\u2019SSs used significantly more often in the SF3B1 \n",
      "mutants (last column)."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def dexseq_table_entries(fn):\n",
      "    \"\"\"\n",
      "    Return list of [number Gencode jxns tested, number novel jxns tested,\n",
      "    number jxns significnat, number novel jxns significant, number cryptic\n",
      "    3'SS significant, number proximal cryptic 3'SS significant, number \n",
      "    proximal cryptic 3'SS significant and use more highly in mutants]\n",
      "    \n",
      "    \"\"\"\n",
      "    df = pd.read_table(fn, index_col=0, header=0)\n",
      "    out = []\n",
      "    tested = df.ix[df.testable]\n",
      "    gencode_tested, novel_tested = tested.ext_annotated.value_counts()\n",
      "    out.append(gencode_tested)\n",
      "    out.append(novel_tested)\n",
      "    sig = tested[tested.padjust < ds.dexseq_p_cutoff]\n",
      "    out.append(sig.shape[0])\n",
      "    out.append(sig[sig.ext_annotated == False].shape[0])    \n",
      "    novel_acceptor = sig.novel_acceptor.value_counts()[True]\n",
      "    out.append(novel_acceptor)\n",
      "    proximal = sig[(sig.downstream_acceptor_dist >= 10) & \n",
      "                   (sig.downstream_acceptor_dist <= 30)]\n",
      "    out.append(proximal.shape[0])\n",
      "    mut_up = (proximal['log2fold(MUT/WT)'] < 0).value_counts()[False]\n",
      "    out.append(mut_up)\n",
      "    return out\n",
      "\n",
      "bcu_res = pd.DataFrame([ dexseq_table_entries(x) for x in [ds.brca_cll_um_dexseq_results, \n",
      "                                                            ds.brca_dexseq_results,\n",
      "                                                            ds.cll_dexseq_results,\n",
      "                                                            ds.um_dexseq_results] ],\n",
      "                         index=['BRCA, CLL, UM', 'BRCA', 'CLL', 'UM'],\n",
      "                         columns=['Gencode tested', 'Novel tested', \n",
      "                                  'Significant p < {}'.format(ds.dexseq_p_cutoff),\n",
      "                                  'Significant novel', 'Sig. cryptic 3\\'SS',\n",
      "                                  'Sig. proximal cryptic 3\\'SS', \n",
      "                                  'Proximal cryptic 3\\'SS, mut up'])\n",
      "bcu_res.apply(lambda x: ['{:,}'.format(y) for y in x ]).to_csv(\n",
      "    os.path.join(outdir, 'file_s02.tsv'),\n",
      "    sep='\\t')\n",
      "bcu_res"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Gencode tested</th>\n",
        "      <th>Novel tested</th>\n",
        "      <th>Significant p &lt; 0.1</th>\n",
        "      <th>Significant novel</th>\n",
        "      <th>Sig. cryptic 3'SS</th>\n",
        "      <th>Sig. proximal cryptic 3'SS</th>\n",
        "      <th>Proximal cryptic 3'SS, mut up</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>BRCA, CLL, UM</th>\n",
        "      <td> 219476</td>\n",
        "      <td> 87941</td>\n",
        "      <td>  1749</td>\n",
        "      <td> 1330</td>\n",
        "      <td> 1117</td>\n",
        "      <td> 619</td>\n",
        "      <td> 619</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>BRCA</th>\n",
        "      <td> 198210</td>\n",
        "      <td> 33104</td>\n",
        "      <td>   492</td>\n",
        "      <td>  306</td>\n",
        "      <td>  280</td>\n",
        "      <td> 192</td>\n",
        "      <td> 192</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CLL</th>\n",
        "      <td> 154980</td>\n",
        "      <td> 60296</td>\n",
        "      <td> 10662</td>\n",
        "      <td> 2635</td>\n",
        "      <td> 1476</td>\n",
        "      <td> 330</td>\n",
        "      <td> 325</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>UM</th>\n",
        "      <td> 106460</td>\n",
        "      <td>  3320</td>\n",
        "      <td>   160</td>\n",
        "      <td>   98</td>\n",
        "      <td>   86</td>\n",
        "      <td>  62</td>\n",
        "      <td>  62</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 5,
       "text": [
        "               Gencode tested  Novel tested  Significant p < 0.1  \\\n",
        "BRCA, CLL, UM          219476         87941                 1749   \n",
        "BRCA                   198210         33104                  492   \n",
        "CLL                    154980         60296                10662   \n",
        "UM                     106460          3320                  160   \n",
        "\n",
        "               Significant novel  Sig. cryptic 3'SS  \\\n",
        "BRCA, CLL, UM               1330               1117   \n",
        "BRCA                         306                280   \n",
        "CLL                         2635               1476   \n",
        "UM                            98                 86   \n",
        "\n",
        "               Sig. proximal cryptic 3'SS  Proximal cryptic 3'SS, mut up  \n",
        "BRCA, CLL, UM                         619                            619  \n",
        "BRCA                                  192                            192  \n",
        "CLL                                   330                            325  \n",
        "UM                                     62                             62  "
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S3\n",
      "\n",
      "619 cryptic 3\u2019SSs located 10-30 bp upstream of canonical \n",
      "3'SSs from joint BRCA, CLL, and UM analysis. Location of \n",
      "5' splice sites and 3'SSs are one-based coordinates that \n",
      "denote the start and end of the intron. The columns COSMIC, \n",
      "TSgene, and ncg denote whether the gene is present in COSMIC, \n",
      "TSGene, or the Network of Cancer Genes respectively. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table(ds.brca_cll_um_dexseq_results,\n",
      "                   index_col=0, header=0)\n",
      "proximal_jxns = pd.read_table(ds.proximal_jxns,\n",
      "                              squeeze=True, header=None)\n",
      "proximal = df.ix[proximal_jxns]\n",
      "\n",
      "bcu_pc = proximal[['chrom', 'strand', 'gene_name']]\n",
      "bcu_pc['5\\'SS'] = proximal.donor.apply(lambda x: x.split(':')[1])\n",
      "bcu_pc['3\\'SS'] = proximal.acceptor.apply(lambda x: x.split(':')[1])\n",
      "bcu_pc['in_frame'] = proximal.downstream_acceptor_dist % 3 == 0\n",
      "bcu_pc = bcu_pc.sort(['chrom', '5\\'SS', '3\\'SS'])\n",
      "bcu_pc = bcu_pc[['chrom', 'strand', '5\\'SS', \n",
      "                     '3\\'SS', 'in_frame', 'gene_name']]\n",
      "bcu_pc.columns = ['chrom', 'strand', '5\\'SS', '3\\'SS', 'in frame', 'gene']\n",
      "bcu_pc['COSMIC'] = bcu_pc.gene.apply(lambda x: x in cancer_gene_census.index)\n",
      "bcu_pc['TSGene'] = bcu_pc.gene.apply(lambda x: x in ts_genes.index)\n",
      "bcu_pc['NCG'] = bcu_pc.gene.apply(lambda x: x in cancer_genes.index)\n",
      "bcu_pc.to_csv(os.path.join(outdir, 'file_s03.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(bcu_pc.shape[0]))\n",
      "bcu_pc.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "619 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom</th>\n",
        "      <th>strand</th>\n",
        "      <th>5'SS</th>\n",
        "      <th>3'SS</th>\n",
        "      <th>in frame</th>\n",
        "      <th>gene</th>\n",
        "      <th>COSMIC</th>\n",
        "      <th>TSGene</th>\n",
        "      <th>NCG</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr1:100477090-100480840:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 100477090</td>\n",
        "      <td> 100480840</td>\n",
        "      <td> False</td>\n",
        "      <td> SLC35A3</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:101458311-101460665:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td> 101460665</td>\n",
        "      <td> 101458311</td>\n",
        "      <td> False</td>\n",
        "      <td>    DPH5</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109235490-109236193:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109235490</td>\n",
        "      <td> 109236193</td>\n",
        "      <td>  True</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:11272544-11272852:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td>  11272852</td>\n",
        "      <td>  11272544</td>\n",
        "      <td>  True</td>\n",
        "      <td>    MTOR</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:1233509-1233749:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td>   1233749</td>\n",
        "      <td>   1233509</td>\n",
        "      <td>  True</td>\n",
        "      <td>   ACAP3</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 6,
       "text": [
        "                           chrom strand       5'SS       3'SS in frame  \\\n",
        "chr1:100477090-100480840:+  chr1      +  100477090  100480840    False   \n",
        "chr1:101458311-101460665:-  chr1      -  101460665  101458311    False   \n",
        "chr1:109235490-109236193:+  chr1      +  109235490  109236193     True   \n",
        "chr1:11272544-11272852:-    chr1      -   11272852   11272544     True   \n",
        "chr1:1233509-1233749:-      chr1      -    1233749    1233509     True   \n",
        "\n",
        "                               gene COSMIC TSGene    NCG  \n",
        "chr1:100477090-100480840:+  SLC35A3  False  False  False  \n",
        "chr1:101458311-101460665:-     DPH5  False  False  False  \n",
        "chr1:109235490-109236193:+  PRPF38B  False  False  False  \n",
        "chr1:11272544-11272852:-       MTOR  False  False  False  \n",
        "chr1:1233509-1233749:-        ACAP3  False  False  False  "
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S4\n",
      "\n",
      "Percent spliced-in values for 324 proximal cryptic 3'SSs from \n",
      "File S3. Note that this should be 325 proximal cryptic 3'SSs \n",
      "but the canonical 3'SS associated with one of the cryptic 3'SSs\n",
      "must have been filtered out for low coverage earlier on."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bcu_psi = pd.read_table(ds.brca_cll_um_psi, index_col=0,\n",
      "                        header=0)\n",
      "bcu_psi.to_csv(os.path.join(outdir, 'file_s04.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "bcu_psi.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>01e63570-aa28-4169-81da-76e838845e78</th>\n",
        "      <th>04f38ef0-e3f5-452a-bc19-e5ae76a666ca</th>\n",
        "      <th>059067a2-7bf7-4360-b564-3c94319ac2e2</th>\n",
        "      <th>09004786-790a-48c9-b3ab-76e92f2e69dd</th>\n",
        "      <th>12bd8213-844d-4b11-8036-d547d4112447</th>\n",
        "      <th>26835e9c-11f9-476f-9d6b-ebc0ae3442e8</th>\n",
        "      <th>3ed7f6c6-5ab0-40cd-8da2-124a7451d287</th>\n",
        "      <th>40dd04c6-ce5d-4c30-aaff-86ece9c78507</th>\n",
        "      <th>42857e49-6357-4fc2-966e-50c8d9ba841f</th>\n",
        "      <th>4c9289e0-378e-4293-877c-e12d207f384e</th>\n",
        "      <th>...</th>\n",
        "      <th>b6b7fb9f-129f-4baa-961b-5715adb22a62</th>\n",
        "      <th>bb496a29-fd16-4ffe-a000-4e70ea9aa76c</th>\n",
        "      <th>c52c3859-08f8-4f50-8596-bfa497a0d24f</th>\n",
        "      <th>cd37724f-9a98-4676-b9a5-e8d6999db1f5</th>\n",
        "      <th>d217d14a-2132-433c-bcac-ccab9d296724</th>\n",
        "      <th>d38df7db-3c58-4865-ae5a-e1e0b80201f1</th>\n",
        "      <th>d3e5466a-4861-4eb9-b5e9-f1b851afd72b</th>\n",
        "      <th>d6b6eb3b-00bd-412d-a610-bc474e0cc319</th>\n",
        "      <th>dfb5fd44-421f-473b-8aef-c7d275c5adcb</th>\n",
        "      <th>fde0f3c2-c9db-4a98-89f9-53685c482160</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>junction</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr6:53381039-53385575:-</th>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.052632</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.060606</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td>...</td>\n",
        "      <td> 0.015873</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.024390</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.017857</td>\n",
        "      <td> 0.025974</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr6:41040824-41046743:+</th>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.086957</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.016949</td>\n",
        "      <td> 0.152174</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.011905</td>\n",
        "      <td>...</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.070175</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.139535</td>\n",
        "      <td> 0.192308</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr3:129284873-129285369:-</th>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.003663</td>\n",
        "      <td> 0.037879</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.006849</td>\n",
        "      <td> 0.014388</td>\n",
        "      <td> 0.132979</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.008811</td>\n",
        "      <td>...</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.041958</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.004866</td>\n",
        "      <td> 0.055215</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.097436</td>\n",
        "      <td> 0.142857</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr12:21627930-21628401:-</th>\n",
        "      <td> 0.006944</td>\n",
        "      <td> 0.007463</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.006098</td>\n",
        "      <td> 0.022727</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td>...</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.008929</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.010256</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.024691</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.040541</td>\n",
        "      <td> 0.000000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr3:50380022-50380348:-</th>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.461538</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.562500</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td>...</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.037500</td>\n",
        "      <td> 0.017778</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.142857</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.216216</td>\n",
        "      <td> 0.011111</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>5 rows \u00d7 56 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 7,
       "text": [
        "                            01e63570-aa28-4169-81da-76e838845e78  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.000000   \n",
        "chr12:21627930-21628401:-                               0.006944   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            04f38ef0-e3f5-452a-bc19-e5ae76a666ca  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.003663   \n",
        "chr12:21627930-21628401:-                               0.007463   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            059067a2-7bf7-4360-b564-3c94319ac2e2  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.052632   \n",
        "chr6:41040824-41046743:+                                0.086957   \n",
        "chr3:129284873-129285369:-                              0.037879   \n",
        "chr12:21627930-21628401:-                               0.000000   \n",
        "chr3:50380022-50380348:-                                0.461538   \n",
        "\n",
        "                            09004786-790a-48c9-b3ab-76e92f2e69dd  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                       0   \n",
        "chr6:41040824-41046743:+                                       0   \n",
        "chr3:129284873-129285369:-                                     0   \n",
        "chr12:21627930-21628401:-                                      0   \n",
        "chr3:50380022-50380348:-                                       0   \n",
        "\n",
        "                            12bd8213-844d-4b11-8036-d547d4112447  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                       0   \n",
        "chr6:41040824-41046743:+                                       0   \n",
        "chr3:129284873-129285369:-                                     0   \n",
        "chr12:21627930-21628401:-                                      0   \n",
        "chr3:50380022-50380348:-                                       0   \n",
        "\n",
        "                            26835e9c-11f9-476f-9d6b-ebc0ae3442e8  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.006849   \n",
        "chr12:21627930-21628401:-                               0.000000   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            3ed7f6c6-5ab0-40cd-8da2-124a7451d287  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.016949   \n",
        "chr3:129284873-129285369:-                              0.014388   \n",
        "chr12:21627930-21628401:-                               0.006098   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            40dd04c6-ce5d-4c30-aaff-86ece9c78507  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.060606   \n",
        "chr6:41040824-41046743:+                                0.152174   \n",
        "chr3:129284873-129285369:-                              0.132979   \n",
        "chr12:21627930-21628401:-                               0.022727   \n",
        "chr3:50380022-50380348:-                                0.562500   \n",
        "\n",
        "                            42857e49-6357-4fc2-966e-50c8d9ba841f  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                       0   \n",
        "chr6:41040824-41046743:+                                       0   \n",
        "chr3:129284873-129285369:-                                     0   \n",
        "chr12:21627930-21628401:-                                      0   \n",
        "chr3:50380022-50380348:-                                       0   \n",
        "\n",
        "                            4c9289e0-378e-4293-877c-e12d207f384e  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.011905   \n",
        "chr3:129284873-129285369:-                              0.008811   \n",
        "chr12:21627930-21628401:-                               0.000000   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                                            ...                   \\\n",
        "junction                                    ...                    \n",
        "chr6:53381039-53385575:-                    ...                    \n",
        "chr6:41040824-41046743:+                    ...                    \n",
        "chr3:129284873-129285369:-                  ...                    \n",
        "chr12:21627930-21628401:-                   ...                    \n",
        "chr3:50380022-50380348:-                    ...                    \n",
        "\n",
        "                            b6b7fb9f-129f-4baa-961b-5715adb22a62  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.015873   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.000000   \n",
        "chr12:21627930-21628401:-                               0.000000   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            bb496a29-fd16-4ffe-a000-4e70ea9aa76c  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.000000   \n",
        "chr12:21627930-21628401:-                               0.008929   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            c52c3859-08f8-4f50-8596-bfa497a0d24f  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                       0   \n",
        "chr6:41040824-41046743:+                                       0   \n",
        "chr3:129284873-129285369:-                                     0   \n",
        "chr12:21627930-21628401:-                                      0   \n",
        "chr3:50380022-50380348:-                                       0   \n",
        "\n",
        "                            cd37724f-9a98-4676-b9a5-e8d6999db1f5  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.070175   \n",
        "chr3:129284873-129285369:-                              0.041958   \n",
        "chr12:21627930-21628401:-                               0.010256   \n",
        "chr3:50380022-50380348:-                                0.037500   \n",
        "\n",
        "                            d217d14a-2132-433c-bcac-ccab9d296724  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.000000   \n",
        "chr12:21627930-21628401:-                               0.000000   \n",
        "chr3:50380022-50380348:-                                0.017778   \n",
        "\n",
        "                            d38df7db-3c58-4865-ae5a-e1e0b80201f1  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.000000   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.004866   \n",
        "chr12:21627930-21628401:-                               0.000000   \n",
        "chr3:50380022-50380348:-                                0.000000   \n",
        "\n",
        "                            d3e5466a-4861-4eb9-b5e9-f1b851afd72b  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.024390   \n",
        "chr6:41040824-41046743:+                                0.000000   \n",
        "chr3:129284873-129285369:-                              0.055215   \n",
        "chr12:21627930-21628401:-                               0.024691   \n",
        "chr3:50380022-50380348:-                                0.142857   \n",
        "\n",
        "                            d6b6eb3b-00bd-412d-a610-bc474e0cc319  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                       0   \n",
        "chr6:41040824-41046743:+                                       0   \n",
        "chr3:129284873-129285369:-                                     0   \n",
        "chr12:21627930-21628401:-                                      0   \n",
        "chr3:50380022-50380348:-                                       0   \n",
        "\n",
        "                            dfb5fd44-421f-473b-8aef-c7d275c5adcb  \\\n",
        "junction                                                           \n",
        "chr6:53381039-53385575:-                                0.017857   \n",
        "chr6:41040824-41046743:+                                0.139535   \n",
        "chr3:129284873-129285369:-                              0.097436   \n",
        "chr12:21627930-21628401:-                               0.040541   \n",
        "chr3:50380022-50380348:-                                0.216216   \n",
        "\n",
        "                            fde0f3c2-c9db-4a98-89f9-53685c482160  \n",
        "junction                                                          \n",
        "chr6:53381039-53385575:-                                0.025974  \n",
        "chr6:41040824-41046743:+                                0.192308  \n",
        "chr3:129284873-129285369:-                              0.142857  \n",
        "chr12:21627930-21628401:-                               0.000000  \n",
        "chr3:50380022-50380348:-                                0.011111  \n",
        "\n",
        "[5 rows x 56 columns]"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S5\n",
      "\n",
      "This file contains the 417 cryptic 3'SSs not located 10-30 bp \n",
      "upstream of canonical 3'SSs from joint BRCA, CLL, and UM analysis \n",
      "and used significantly more often in SF3B1 mutants. Location of \n",
      "5' splice sites and 3'SSs are one-based coordinates that denote the \n",
      "start and end of the intron. The columns COSMIC, TSgene, and ncg \n",
      "denote whether the gene is present in COSMIC, TSGene, or the \n",
      "Network of Cancer Genes respectively."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table(ds.brca_cll_um_dexseq_results,\n",
      "                   index_col=0, header=0)\n",
      "not_proximal_jxns = pd.read_table(ds.not_proximal_jxns,\n",
      "                                  squeeze=True, header=None)\n",
      "bcu_pd = df.ix[not_proximal_jxns]\n",
      "bcu_pd['5\\'SS'] = bcu_pd.donor.apply(lambda x: x.split(':')[1])\n",
      "bcu_pd['3\\'SS'] = bcu_pd.acceptor.apply(lambda x: x.split(':')[1])\n",
      "bcu_pd['in_frame'] = bcu_pd.downstream_acceptor_dist % 3 == 0\n",
      "bcu_pd = bcu_pd.sort(['chrom', '5\\'SS', '3\\'SS'])\n",
      "bcu_pd = bcu_pd[['chrom', 'strand', '5\\'SS', \n",
      "                     '3\\'SS', 'in_frame', 'gene_name']]\n",
      "bcu_pd.columns = ['chrom', 'strand', '5\\'SS', '3\\'SS', 'in frame', 'gene']\n",
      "bcu_pd['COSMIC'] = bcu_pd.gene.apply(lambda x: x in cancer_gene_census.index)\n",
      "bcu_pd['TSGene'] = bcu_pd.gene.apply(lambda x: x in ts_genes.index)\n",
      "bcu_pd['NCG'] = bcu_pd.gene.apply(lambda x: x in cancer_genes.index)\n",
      "bcu_pd.to_csv(os.path.join(outdir, 'file_s05.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(bcu_pd.shape[0]))\n",
      "bcu_pd.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "417 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom</th>\n",
        "      <th>strand</th>\n",
        "      <th>5'SS</th>\n",
        "      <th>3'SS</th>\n",
        "      <th>in frame</th>\n",
        "      <th>gene</th>\n",
        "      <th>COSMIC</th>\n",
        "      <th>TSGene</th>\n",
        "      <th>NCG</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr1:100606071-100606400:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 100606071</td>\n",
        "      <td> 100606400</td>\n",
        "      <td> False</td>\n",
        "      <td>  TRMT13</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109235490-109235804:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109235490</td>\n",
        "      <td> 109235804</td>\n",
        "      <td> False</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109238960-109240568:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109238960</td>\n",
        "      <td> 109240568</td>\n",
        "      <td>  True</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:110212194-110212385:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 110212194</td>\n",
        "      <td> 110212385</td>\n",
        "      <td> False</td>\n",
        "      <td>   GSTM2</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:113196284-113197099:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 113196284</td>\n",
        "      <td> 113197099</td>\n",
        "      <td> False</td>\n",
        "      <td>  CAPZA1</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "                           chrom strand       5'SS       3'SS in frame  \\\n",
        "chr1:100606071-100606400:+  chr1      +  100606071  100606400    False   \n",
        "chr1:109235490-109235804:+  chr1      +  109235490  109235804    False   \n",
        "chr1:109238960-109240568:+  chr1      +  109238960  109240568     True   \n",
        "chr1:110212194-110212385:+  chr1      +  110212194  110212385    False   \n",
        "chr1:113196284-113197099:+  chr1      +  113196284  113197099    False   \n",
        "\n",
        "                               gene COSMIC TSGene    NCG  \n",
        "chr1:100606071-100606400:+   TRMT13  False  False  False  \n",
        "chr1:109235490-109235804:+  PRPF38B  False  False  False  \n",
        "chr1:109238960-109240568:+  PRPF38B  False  False  False  \n",
        "chr1:110212194-110212385:+    GSTM2  False  False  False  \n",
        "chr1:113196284-113197099:+   CAPZA1  False  False  False  "
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S6\n",
      "\n",
      "GSEA results for 912 genes that contain the 619 proximal\n",
      "cryptic 3'SSs from File S4 and the 417 distal cryptic 3'SSs\n",
      "from File S5. This file was created by manually copying\n",
      "the gene IDs into the GSEA web tool and exporting the results\n",
      "as an Excel file.\n",
      "\n",
      "## File S7\n",
      "\n",
      "325 significant cryptic 3\u2019SSs located 10-30 bp upstream \n",
      "of canonical 3\u2019SSs and used more often in SF3B1 mutants \n",
      "from CLL-only DEXSeq analysis. Location of 5\u2019 splice \n",
      "sites and 3\u2019SSs are one-based coordinates that denote \n",
      "the start and end of the intron. The columns COSMIC, \n",
      "TSgene, and ncg denote whether the gene is present in \n",
      "COSMIC, TSGene, or the Network of Cancer Genes \n",
      "respectively."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table(ds.cll_dexseq_results,\n",
      "                   index_col=0, header=0)\n",
      "tested = df.ix[df.testable]\n",
      "sig = tested[tested.padjust < ds.dexseq_p_cutoff]\n",
      "proximal = sig[(sig.downstream_acceptor_dist >= 10) & \n",
      "               (sig.downstream_acceptor_dist <= 30) &\n",
      "               (sig['log2fold(MUT/WT)'] > 0)]\n",
      "\n",
      "cll_pc = proximal[['chrom', 'strand', 'gene_name']]\n",
      "cll_pc['5\\'SS'] = proximal.donor.apply(lambda x: x.split(':')[1])\n",
      "cll_pc['3\\'SS'] = proximal.acceptor.apply(lambda x: x.split(':')[1])\n",
      "cll_pc['in_frame'] = proximal.downstream_acceptor_dist % 3 == 0\n",
      "cll_pc = cll_pc.sort(['chrom', '5\\'SS', '3\\'SS'])\n",
      "cll_pc = cll_pc[['chrom', 'strand', '5\\'SS', \n",
      "                     '3\\'SS', 'in_frame', 'gene_name']]\n",
      "cll_pc.columns = ['chrom', 'strand', '5\\'SS', '3\\'SS', 'in frame', 'gene']\n",
      "cll_pc['COSMIC'] = cll_pc.gene.apply(lambda x: x in cancer_gene_census.index)\n",
      "cll_pc['TSGene'] = cll_pc.gene.apply(lambda x: x in ts_genes.index)\n",
      "cll_pc['NCG'] = cll_pc.gene.apply(lambda x: x in cancer_genes.index)\n",
      "cll_pc.to_csv(os.path.join(outdir, 'file_s07.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(cll_pc.shape[0]))\n",
      "cll_pc.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "325 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom</th>\n",
        "      <th>strand</th>\n",
        "      <th>5'SS</th>\n",
        "      <th>3'SS</th>\n",
        "      <th>in frame</th>\n",
        "      <th>gene</th>\n",
        "      <th>COSMIC</th>\n",
        "      <th>TSGene</th>\n",
        "      <th>NCG</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>junction</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr1:101458311-101460665:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td> 101460665</td>\n",
        "      <td> 101458311</td>\n",
        "      <td> False</td>\n",
        "      <td>    DPH5</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109235490-109236193:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109235490</td>\n",
        "      <td> 109236193</td>\n",
        "      <td>  True</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:145601639-145601747:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td> 145601747</td>\n",
        "      <td> 145601639</td>\n",
        "      <td> False</td>\n",
        "      <td>  POLR3C</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:153631363-153631594:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 153631363</td>\n",
        "      <td> 153631594</td>\n",
        "      <td> False</td>\n",
        "      <td>  SNAPIN</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:155282187-155287710:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 155282187</td>\n",
        "      <td> 155287710</td>\n",
        "      <td>  True</td>\n",
        "      <td>    FDPS</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "                           chrom strand       5'SS       3'SS in frame  \\\n",
        "junction                                                                 \n",
        "chr1:101458311-101460665:-  chr1      -  101460665  101458311    False   \n",
        "chr1:109235490-109236193:+  chr1      +  109235490  109236193     True   \n",
        "chr1:145601639-145601747:-  chr1      -  145601747  145601639    False   \n",
        "chr1:153631363-153631594:+  chr1      +  153631363  153631594    False   \n",
        "chr1:155282187-155287710:+  chr1      +  155282187  155287710     True   \n",
        "\n",
        "                               gene COSMIC TSGene    NCG  \n",
        "junction                                                  \n",
        "chr1:101458311-101460665:-     DPH5  False  False  False  \n",
        "chr1:109235490-109236193:+  PRPF38B  False  False  False  \n",
        "chr1:145601639-145601747:-   POLR3C  False  False  False  \n",
        "chr1:153631363-153631594:+   SNAPIN  False  False  False  \n",
        "chr1:155282187-155287710:+     FDPS  False  False  False  "
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S8\n",
      "\n",
      "Percent spliced-in values for 324 proximal cryptic 3'SSs from \n",
      "File S6. Note that this should be 325 proximal cryptic 3'SSs \n",
      "but the canonical 3'SS associated with one of the cryptic 3'SSs\n",
      "must have been filtered out for low coverage earlier on."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cll_psi = pd.read_table(ds.cll_psi, index_col=0,\n",
      "                        header=0)\n",
      "cll_psi.to_csv(os.path.join(outdir, 'file_s08.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "cll_psi.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>SA00242</th>\n",
        "      <th>SA03982</th>\n",
        "      <th>SA05603</th>\n",
        "      <th>SA15699</th>\n",
        "      <th>SA22083</th>\n",
        "      <th>SA24823</th>\n",
        "      <th>SA29677</th>\n",
        "      <th>SA30914</th>\n",
        "      <th>SA52314</th>\n",
        "      <th>SA52609</th>\n",
        "      <th>SA58773</th>\n",
        "      <th>SA65385</th>\n",
        "      <th>SA66418</th>\n",
        "      <th>SA83613</th>\n",
        "      <th>SA93051</th>\n",
        "      <th>SA95730</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>junction</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr6:41040824-41046743:+</th>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.222222</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.111732</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.185185</td>\n",
        "      <td> 0.187500</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.128571</td>\n",
        "      <td> 0.119403</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr3:129284873-129285369:-</th>\n",
        "      <td> 0.153846</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.048387</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.439024</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.181818</td>\n",
        "      <td> 0.200000</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr12:21627930-21628401:-</th>\n",
        "      <td> 0.021739</td>\n",
        "      <td> 0.070922</td>\n",
        "      <td> 0.052632</td>\n",
        "      <td> 0.016997</td>\n",
        "      <td> 0.034375</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.165138</td>\n",
        "      <td> 0.114583</td>\n",
        "      <td> 0.029412</td>\n",
        "      <td> 0.026316</td>\n",
        "      <td> 0.011111</td>\n",
        "      <td> 0.020408</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.006944</td>\n",
        "      <td> 0.227273</td>\n",
        "      <td> 0.061290</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr16:20788903-20792021:+</th>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.068966</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.060606</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.126761</td>\n",
        "      <td> 0.200000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.112360</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr7:77469618-77522908:+</th>\n",
        "      <td> 0.006270</td>\n",
        "      <td> 0.120370</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.040590</td>\n",
        "      <td> 0.004219</td>\n",
        "      <td> 0.115942</td>\n",
        "      <td> 0.147059</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.013514</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.000000</td>\n",
        "      <td> 0.091398</td>\n",
        "      <td> 0.109524</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "                             SA00242   SA03982   SA05603   SA15699   SA22083  \\\n",
        "junction                                                                       \n",
        "chr6:41040824-41046743:+    0.000000  0.222222  0.000000  0.000000  0.111732   \n",
        "chr3:129284873-129285369:-  0.153846  0.000000  0.000000  0.000000  0.048387   \n",
        "chr12:21627930-21628401:-   0.021739  0.070922  0.052632  0.016997  0.034375   \n",
        "chr16:20788903-20792021:+   0.000000  0.068966  0.000000  0.000000  0.060606   \n",
        "chr7:77469618-77522908:+    0.006270  0.120370  0.000000  0.000000  0.040590   \n",
        "\n",
        "                             SA24823   SA29677   SA30914   SA52314   SA52609  \\\n",
        "junction                                                                       \n",
        "chr6:41040824-41046743:+    0.000000  0.185185  0.187500  0.000000  0.000000   \n",
        "chr3:129284873-129285369:-  0.000000  0.439024  0.000000  0.000000  0.000000   \n",
        "chr12:21627930-21628401:-   0.000000  0.165138  0.114583  0.029412  0.026316   \n",
        "chr16:20788903-20792021:+   0.000000  0.126761  0.200000  0.000000  0.000000   \n",
        "chr7:77469618-77522908:+    0.004219  0.115942  0.147059  0.000000  0.013514   \n",
        "\n",
        "                             SA58773   SA65385  SA66418   SA83613   SA93051  \\\n",
        "junction                                                                      \n",
        "chr6:41040824-41046743:+    0.000000  0.000000        0  0.000000  0.128571   \n",
        "chr3:129284873-129285369:-  0.000000  0.000000        0  0.000000  0.181818   \n",
        "chr12:21627930-21628401:-   0.011111  0.020408        0  0.006944  0.227273   \n",
        "chr16:20788903-20792021:+   0.000000  0.000000        0  0.000000  0.000000   \n",
        "chr7:77469618-77522908:+    0.000000  0.000000        0  0.000000  0.091398   \n",
        "\n",
        "                             SA95730  \n",
        "junction                              \n",
        "chr6:41040824-41046743:+    0.119403  \n",
        "chr3:129284873-129285369:-  0.200000  \n",
        "chr12:21627930-21628401:-   0.061290  \n",
        "chr16:20788903-20792021:+   0.112360  \n",
        "chr7:77469618-77522908:+    0.109524  "
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S9\n",
      "\n",
      "272 genes that are differentially expressed between \n",
      "SF3B1 mutant and wild-type samples from joint analysis \n",
      "of CLL, BRCA, and UM using DESeq2."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bcu_de = pd.read_csv(ds.brca_cll_um_deseq2_results,\n",
      "                       index_col=0, header=0)\n",
      "bcu_de = bcu_de.dropna(subset=['padj'])\n",
      "bcu_de = bcu_de[bcu_de.padj < ds.deseq2_p_cutoff]\n",
      "bcu_de = bcu_de[['log2FoldChange', 'pvalue', 'padj']]\n",
      "bcu_de.columns = ['log2 fold change (MUT/WT)', 'p-value', \n",
      "                    'Adjusted p-value']\n",
      "bcu_de['Gene name'] = gene_info.ix[bcu_de.index, 'gene_name']\n",
      "bcu_de.index.name = 'Gene ID'\n",
      "\n",
      "bcu_de.to_csv(os.path.join(outdir, 'file_s09.tsv'),\n",
      "                sep='\\t')\n",
      "print('{} rows'.format(bcu_de.shape[0]))\n",
      "bcu_de.head()\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "272 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>log2 fold change (MUT/WT)</th>\n",
        "      <th>p-value</th>\n",
        "      <th>Adjusted p-value</th>\n",
        "      <th>Gene name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Gene ID</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>ENSG00000005156.7</th>\n",
        "      <td> 0.560285</td>\n",
        "      <td> 0.000046</td>\n",
        "      <td> 0.019017</td>\n",
        "      <td>   LIG3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000006459.6</th>\n",
        "      <td> 0.565435</td>\n",
        "      <td> 0.000841</td>\n",
        "      <td> 0.075077</td>\n",
        "      <td> JHDM1D</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008226.15</th>\n",
        "      <td> 1.028977</td>\n",
        "      <td> 0.000050</td>\n",
        "      <td> 0.019873</td>\n",
        "      <td>  DLEC1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008324.6</th>\n",
        "      <td> 0.432990</td>\n",
        "      <td> 0.001072</td>\n",
        "      <td> 0.082723</td>\n",
        "      <td> SS18L2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008988.5</th>\n",
        "      <td> 0.918372</td>\n",
        "      <td> 0.000001</td>\n",
        "      <td> 0.002189</td>\n",
        "      <td>  RPS20</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 11,
       "text": [
        "                    log2 fold change (MUT/WT)   p-value  Adjusted p-value  \\\n",
        "Gene ID                                                                     \n",
        "ENSG00000005156.7                    0.560285  0.000046          0.019017   \n",
        "ENSG00000006459.6                    0.565435  0.000841          0.075077   \n",
        "ENSG00000008226.15                   1.028977  0.000050          0.019873   \n",
        "ENSG00000008324.6                    0.432990  0.001072          0.082723   \n",
        "ENSG00000008988.5                    0.918372  0.000001          0.002189   \n",
        "\n",
        "                   Gene name  \n",
        "Gene ID                       \n",
        "ENSG00000005156.7       LIG3  \n",
        "ENSG00000006459.6     JHDM1D  \n",
        "ENSG00000008226.15     DLEC1  \n",
        "ENSG00000008324.6     SS18L2  \n",
        "ENSG00000008988.5      RPS20  "
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S10\n",
      "\n",
      "GSEA results for 272 genes differentially expressed\n",
      "between the SF3B1 mutant and wild-type samples in \n",
      "a joint analysis of all three cancers from File S9.\n",
      "This file was created by manually copying the gene IDs \n",
      "into the GSEA web tool and exporting the results\n",
      "as an Excel file.\n",
      "\n",
      "## File S11\n",
      "\n",
      "Genes whose expression differs significantly between\n",
      "*SF3B1* mutant and wild-type samples for analysis\n",
      "of CLL samples using DESeq2."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cll_de = pd.read_csv(ds.cll_deseq2_results,\n",
      "                       index_col=0, header=0)\n",
      "cll_de = cll_de.dropna(subset=['padj'])\n",
      "cll_de = cll_de[cll_de.padj < ds.deseq2_p_cutoff]\n",
      "cll_de = cll_de[['log2FoldChange', 'pvalue', 'padj']]\n",
      "cll_de.columns = ['log2 fold change (MUT/WT)', 'p-value', \n",
      "                    'Adjusted p-value']\n",
      "cll_de['Gene name'] = gene_info.ix[cll_de.index, 'gene_name']\n",
      "cll_de.index.name = 'Gene ID'\n",
      "\n",
      "cll_de.to_csv(os.path.join(outdir, 'file_s11.tsv'),\n",
      "                sep='\\t')\n",
      "print('{} rows'.format(cll_de.shape[0]))\n",
      "cll_de.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "33 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>log2 fold change (MUT/WT)</th>\n",
        "      <th>p-value</th>\n",
        "      <th>Adjusted p-value</th>\n",
        "      <th>Gene name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Gene ID</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>ENSG00000005156.7</th>\n",
        "      <td> 0.606988</td>\n",
        "      <td> 0.000080</td>\n",
        "      <td> 0.083089</td>\n",
        "      <td>    LIG3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008226.15</th>\n",
        "      <td> 1.020171</td>\n",
        "      <td> 0.000167</td>\n",
        "      <td> 0.088107</td>\n",
        "      <td>   DLEC1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000028116.12</th>\n",
        "      <td>-0.868542</td>\n",
        "      <td> 0.000042</td>\n",
        "      <td> 0.066585</td>\n",
        "      <td>    VRK2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000066027.4</th>\n",
        "      <td> 0.840471</td>\n",
        "      <td> 0.000208</td>\n",
        "      <td> 0.099852</td>\n",
        "      <td> PPP2R5A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000070785.12</th>\n",
        "      <td>-0.533570</td>\n",
        "      <td> 0.000128</td>\n",
        "      <td> 0.083089</td>\n",
        "      <td>  EIF2B3</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "                    log2 fold change (MUT/WT)   p-value  Adjusted p-value  \\\n",
        "Gene ID                                                                     \n",
        "ENSG00000005156.7                    0.606988  0.000080          0.083089   \n",
        "ENSG00000008226.15                   1.020171  0.000167          0.088107   \n",
        "ENSG00000028116.12                  -0.868542  0.000042          0.066585   \n",
        "ENSG00000066027.4                    0.840471  0.000208          0.099852   \n",
        "ENSG00000070785.12                  -0.533570  0.000128          0.083089   \n",
        "\n",
        "                   Gene name  \n",
        "Gene ID                       \n",
        "ENSG00000005156.7       LIG3  \n",
        "ENSG00000008226.15     DLEC1  \n",
        "ENSG00000028116.12      VRK2  \n",
        "ENSG00000066027.4    PPP2R5A  \n",
        "ENSG00000070785.12    EIF2B3  "
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 12
    }
   ],
   "metadata": {}
  }
 ]
}