{
 "metadata": {
  "name": "",
  "signature": "sha256:c341902e2b173a4e4fdb9750d1bd23b6d4c46d089f6ffa7d62d549c0d7e37f9b"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Supporting Files\n",
      "\n",
      "This notebook creates the supplementary files from the paper."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "\n",
      "import ds2014 as ds\n",
      "import pandas as pd"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "curdir = os.getcwd()\n",
      "outdir = os.path.join(os.path.split(curdir)[0], \n",
      "                      'output', \n",
      "                      'supporting_files')\n",
      "ds.makedir(outdir)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "meta = pd.read_table(ds.metadata, header=0,\n",
      "                     index_col=0)\n",
      "\n",
      "gene_info = pd.read_table(ds.gene_info, header=0,\n",
      "                          index_col=0)\n",
      "\n",
      "# cancer census genes from COSMIC\n",
      "cancer_gene_census = pd.read_table(ds.cancer_gene_census,\n",
      "                                   index_col=0)\n",
      "# cancer genes from NCG4.0 (http://ncg.kcl.ac.uk/)\n",
      "cancer_genes = pd.read_table(ds.cancer_genes,\n",
      "                             index_col=1, header=0)\n",
      "# tumor suppressors from TSGene (http://bioinfo.mc.vanderbilt.edu/TSGene/)\n",
      "ts_genes = pd.read_table(ds.ts_genes,\n",
      "                         index_col=1, header=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S1\n",
      "\n",
      "Samples used in this study. SF3B1 mutated samples have \n",
      "columns for frequency of SF3B1 mutation in RNA-seq data, \n",
      "mutation type, codon change and whether the mutation is \n",
      "in the HEAT 5-9 repeats. These columns are empty for \n",
      "SF3B1 wild-type tumor samples."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sf3b1_mut_info = pd.read_table(ds.sf3b1_mut_info,\n",
      "                               header=0, index_col=None)\n",
      "file_s01 = sf3b1_mut_info[['name', 'disease_abbr',\n",
      "                           'alt_freq', 'mut_type',\n",
      "                           'codon_change', 'HEAT']]\n",
      "file_s01['alt_freq'] = file_s01.alt_freq.apply(lambda x: '{:.2f}'.format(x)).values\n",
      "file_s01.columns = ['Sample Name', 'Cancer', 'Alternate Allele Frequency',\n",
      "                    'Mutation Type', 'Codon Change', 'In HEAT 5-9']\n",
      "file_s01['SF3B1 mutated'] = True\n",
      "wt = pd.DataFrame(columns=file_s01.columns)\n",
      "wt['Sample Name'] = meta[meta.condition == 'WT'].index\n",
      "wt['SF3B1 mutated'] = False\n",
      "wt['Cancer'] = meta.ix[wt['Sample Name'], 'disease_abbr'].values\n",
      "file_s01 = pd.concat([file_s01, wt])\n",
      "file_s01.to_csv(os.path.join(outdir, 'file_s01.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(file_s01.shape[0]))\n",
      "file_s01.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "82 rows\n"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "-c:6: SettingWithCopyWarning: A value is trying to be set on a copy of a slice from a DataFrame.\n",
        "Try using .loc[row_index,col_indexer] = value instead\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Sample Name</th>\n",
        "      <th>Cancer</th>\n",
        "      <th>Alternate Allele Frequency</th>\n",
        "      <th>Mutation Type</th>\n",
        "      <th>Codon Change</th>\n",
        "      <th>In HEAT 5-9</th>\n",
        "      <th>SF3B1 mutated</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0</th>\n",
        "      <td> SA30914</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.56</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1</th>\n",
        "      <td> SA95730</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.46</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2</th>\n",
        "      <td> SA29677</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.52</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3</th>\n",
        "      <td> SA03982</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.51</td>\n",
        "      <td> missense</td>\n",
        "      <td> K700E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4</th>\n",
        "      <td> SA93051</td>\n",
        "      <td> CLL</td>\n",
        "      <td> 0.50</td>\n",
        "      <td> missense</td>\n",
        "      <td> K666E</td>\n",
        "      <td> True</td>\n",
        "      <td> True</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 4,
       "text": [
        "  Sample Name Cancer Alternate Allele Frequency Mutation Type Codon Change  \\\n",
        "0     SA30914    CLL                       0.56      missense        K700E   \n",
        "1     SA95730    CLL                       0.46      missense        K700E   \n",
        "2     SA29677    CLL                       0.52      missense        K700E   \n",
        "3     SA03982    CLL                       0.51      missense        K700E   \n",
        "4     SA93051    CLL                       0.50      missense        K666E   \n",
        "\n",
        "  In HEAT 5-9 SF3B1 mutated  \n",
        "0        True          True  \n",
        "1        True          True  \n",
        "2        True          True  \n",
        "3        True          True  \n",
        "4        True          True  "
       ]
      }
     ],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S2\n",
      "\n",
      "Summary of differential junction usage results from DEXSeq. \n",
      "DEXSeq was used to test for differential splice junction \n",
      "usage in a joint analysis of the CLL, BRCA, and UM samples \n",
      "as well as individually for each cancer type. \u201cNovel\u201d indicates \n",
      "that the junction is not annotated in Gencode. Proximal \n",
      "indicates that a novel 3\u2019SS is 10-30 bp upstream of a canonical \n",
      "Gencode 3\u2019SS. In this paper, we focused mainly on the proximal \n",
      "cryptic 3\u2019SSs used significantly more often in the SF3B1 \n",
      "mutants (last column)."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def dexseq_table_entries(fn):\n",
      "    \"\"\"\n",
      "    Return list of [number Gencode jxns tested, number novel jxns tested,\n",
      "    number jxns significnat, number novel jxns significant, number cryptic\n",
      "    3'SS significant, number proximal cryptic 3'SS significant, number \n",
      "    proximal cryptic 3'SS significant and use more highly in mutants]\n",
      "    \n",
      "    \"\"\"\n",
      "    df = pd.read_table(fn, index_col=0, header=0)\n",
      "    out = []\n",
      "    tested = df.ix[df.testable]\n",
      "    gencode_tested, novel_tested = tested.ext_annotated.value_counts()\n",
      "    out.append(gencode_tested)\n",
      "    out.append(novel_tested)\n",
      "    sig = tested[tested.padjust < ds.dexseq_p_cutoff]\n",
      "    out.append(sig.shape[0])\n",
      "    out.append(sig[sig.ext_annotated == False].shape[0])    \n",
      "    novel_acceptor = sig.novel_acceptor.value_counts()[True]\n",
      "    out.append(novel_acceptor)\n",
      "    proximal = sig[(sig.downstream_acceptor_dist >= 10) & \n",
      "                   (sig.downstream_acceptor_dist <= 30)]\n",
      "    out.append(proximal.shape[0])\n",
      "    mut_up = (proximal['log2fold(MUT/WT)'] < 0).value_counts()[False]\n",
      "    out.append(mut_up)\n",
      "    return out\n",
      "\n",
      "file_s02 = pd.DataFrame([ dexseq_table_entries(x) for x in [ds.brca_cll_um_dexseq_results, \n",
      "                                                            ds.brca_dexseq_results,\n",
      "                                                            ds.cll_dexseq_results,\n",
      "                                                            ds.um_dexseq_results] ],\n",
      "                         index=['BRCA, CLL, UM', 'BRCA', 'CLL', 'UM'],\n",
      "                         columns=['Gencode tested', 'Novel tested', \n",
      "                                  'Significant p < {}'.format(ds.dexseq_p_cutoff),\n",
      "                                  'Significant novel', 'Sig. cryptic 3\\'SS',\n",
      "                                  'Sig. proximal cryptic 3\\'SS', \n",
      "                                  'Proximal cryptic 3\\'SS, mut up'])\n",
      "file_s02.apply(lambda x: ['{:,}'.format(y) for y in x ]).to_csv(\n",
      "    os.path.join(outdir, 'file_s02.tsv'),\n",
      "    sep='\\t')\n",
      "file_s02"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Gencode tested</th>\n",
        "      <th>Novel tested</th>\n",
        "      <th>Significant p &lt; 0.1</th>\n",
        "      <th>Significant novel</th>\n",
        "      <th>Sig. cryptic 3'SS</th>\n",
        "      <th>Sig. proximal cryptic 3'SS</th>\n",
        "      <th>Proximal cryptic 3'SS, mut up</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>BRCA, CLL, UM</th>\n",
        "      <td> 219476</td>\n",
        "      <td> 87941</td>\n",
        "      <td>  1749</td>\n",
        "      <td> 1330</td>\n",
        "      <td> 1117</td>\n",
        "      <td> 619</td>\n",
        "      <td> 619</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>BRCA</th>\n",
        "      <td> 198210</td>\n",
        "      <td> 33104</td>\n",
        "      <td>   492</td>\n",
        "      <td>  306</td>\n",
        "      <td>  280</td>\n",
        "      <td> 192</td>\n",
        "      <td> 192</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>CLL</th>\n",
        "      <td> 154980</td>\n",
        "      <td> 60296</td>\n",
        "      <td> 10662</td>\n",
        "      <td> 2635</td>\n",
        "      <td> 1476</td>\n",
        "      <td> 330</td>\n",
        "      <td> 325</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>UM</th>\n",
        "      <td> 106460</td>\n",
        "      <td>  3320</td>\n",
        "      <td>   160</td>\n",
        "      <td>   98</td>\n",
        "      <td>   86</td>\n",
        "      <td>  62</td>\n",
        "      <td>  62</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 5,
       "text": [
        "               Gencode tested  Novel tested  Significant p < 0.1  \\\n",
        "BRCA, CLL, UM          219476         87941                 1749   \n",
        "BRCA                   198210         33104                  492   \n",
        "CLL                    154980         60296                10662   \n",
        "UM                     106460          3320                  160   \n",
        "\n",
        "               Significant novel  Sig. cryptic 3'SS  \\\n",
        "BRCA, CLL, UM               1330               1117   \n",
        "BRCA                         306                280   \n",
        "CLL                         2635               1476   \n",
        "UM                            98                 86   \n",
        "\n",
        "               Sig. proximal cryptic 3'SS  Proximal cryptic 3'SS, mut up  \n",
        "BRCA, CLL, UM                         619                            619  \n",
        "BRCA                                  192                            192  \n",
        "CLL                                   330                            325  \n",
        "UM                                     62                             62  "
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S3\n",
      "\n",
      "619 cryptic 3\u2019SSs located 10-30 bp upstream of canonical \n",
      "3'SSs from joint BRCA, CLL, and UM analysis. Location of \n",
      "5' splice sites and 3'SSs are one-based coordinates that \n",
      "denote the start and end of the intron. The columns COSMIC, \n",
      "TSgene, and ncg denote whether the gene is present in COSMIC, \n",
      "TSGene, or the Network of Cancer Genes respectively. "
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table(ds.brca_cll_um_dexseq_results,\n",
      "                   index_col=0, header=0)\n",
      "proximal_jxns = pd.read_table(ds.proximal_jxns,\n",
      "                              squeeze=True, header=None)\n",
      "proximal = df.ix[proximal_jxns]\n",
      "\n",
      "file_s03 = proximal[['chrom', 'strand', 'gene_name']]\n",
      "file_s03['5\\'SS'] = proximal.donor.apply(lambda x: x.split(':')[1])\n",
      "file_s03['3\\'SS'] = proximal.acceptor.apply(lambda x: x.split(':')[1])\n",
      "file_s03['in_frame'] = proximal.downstream_acceptor_dist % 3 == 0\n",
      "file_s03 = file_s03.sort(['chrom', '5\\'SS', '3\\'SS'])\n",
      "file_s03 = file_s03[['chrom', 'strand', '5\\'SS', \n",
      "                     '3\\'SS', 'in_frame', 'gene_name']]\n",
      "file_s03.columns = ['chrom', 'strand', '5\\'SS', '3\\'SS', 'in frame', 'gene']\n",
      "file_s03['COSMIC'] = file_s03.gene.apply(lambda x: x in cancer_gene_census.index)\n",
      "file_s03['TSGene'] = file_s03.gene.apply(lambda x: x in ts_genes.index)\n",
      "file_s03['NCG'] = file_s03.gene.apply(lambda x: x in cancer_genes.index)\n",
      "file_s03.to_csv(os.path.join(outdir, 'file_s03.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(file_s03.shape[0]))\n",
      "file_s03.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "619 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom</th>\n",
        "      <th>strand</th>\n",
        "      <th>5'SS</th>\n",
        "      <th>3'SS</th>\n",
        "      <th>in frame</th>\n",
        "      <th>gene</th>\n",
        "      <th>COSMIC</th>\n",
        "      <th>TSGene</th>\n",
        "      <th>NCG</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr1:100477090-100480840:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 100477090</td>\n",
        "      <td> 100480840</td>\n",
        "      <td> False</td>\n",
        "      <td> SLC35A3</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:101458311-101460665:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td> 101460665</td>\n",
        "      <td> 101458311</td>\n",
        "      <td> False</td>\n",
        "      <td>    DPH5</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109235490-109236193:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109235490</td>\n",
        "      <td> 109236193</td>\n",
        "      <td>  True</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:11272544-11272852:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td>  11272852</td>\n",
        "      <td>  11272544</td>\n",
        "      <td>  True</td>\n",
        "      <td>    MTOR</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:1233509-1233749:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td>   1233749</td>\n",
        "      <td>   1233509</td>\n",
        "      <td>  True</td>\n",
        "      <td>   ACAP3</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 6,
       "text": [
        "                           chrom strand       5'SS       3'SS in frame  \\\n",
        "chr1:100477090-100480840:+  chr1      +  100477090  100480840    False   \n",
        "chr1:101458311-101460665:-  chr1      -  101460665  101458311    False   \n",
        "chr1:109235490-109236193:+  chr1      +  109235490  109236193     True   \n",
        "chr1:11272544-11272852:-    chr1      -   11272852   11272544     True   \n",
        "chr1:1233509-1233749:-      chr1      -    1233749    1233509     True   \n",
        "\n",
        "                               gene COSMIC TSGene    NCG  \n",
        "chr1:100477090-100480840:+  SLC35A3  False  False  False  \n",
        "chr1:101458311-101460665:-     DPH5  False  False  False  \n",
        "chr1:109235490-109236193:+  PRPF38B  False  False  False  \n",
        "chr1:11272544-11272852:-       MTOR  False  False  False  \n",
        "chr1:1233509-1233749:-        ACAP3  False  False  False  "
       ]
      }
     ],
     "prompt_number": 6
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S4\n",
      "\n",
      "This file contains the 417 cryptic 3'SSs not located 10-30 bp \n",
      "upstream of canonical 3'SSs from joint BRCA, CLL, and UM analysis \n",
      "and used significantly more often in SF3B1 mutants. Location of \n",
      "5' splice sites and 3'SSs are one-based coordinates that denote the \n",
      "start and end of the intron. The columns COSMIC, TSgene, and ncg \n",
      "denote whether the gene is present in COSMIC, TSGene, or the \n",
      "Network of Cancer Genes respectively."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table(ds.brca_cll_um_dexseq_results,\n",
      "                   index_col=0, header=0)\n",
      "not_proximal_jxns = pd.read_table(ds.not_proximal_jxns,\n",
      "                                  squeeze=True, header=None)\n",
      "file_s04 = df.ix[not_proximal_jxns]\n",
      "file_s04['5\\'SS'] = file_s04.donor.apply(lambda x: x.split(':')[1])\n",
      "file_s04['3\\'SS'] = file_s04.acceptor.apply(lambda x: x.split(':')[1])\n",
      "file_s04['in_frame'] = file_s04.downstream_acceptor_dist % 3 == 0\n",
      "file_s04 = file_s04.sort(['chrom', '5\\'SS', '3\\'SS'])\n",
      "file_s04 = file_s04[['chrom', 'strand', '5\\'SS', \n",
      "                     '3\\'SS', 'in_frame', 'gene_name']]\n",
      "file_s04.columns = ['chrom', 'strand', '5\\'SS', '3\\'SS', 'in frame', 'gene']\n",
      "file_s04['COSMIC'] = file_s04.gene.apply(lambda x: x in cancer_gene_census.index)\n",
      "file_s04['TSGene'] = file_s04.gene.apply(lambda x: x in ts_genes.index)\n",
      "file_s04['NCG'] = file_s04.gene.apply(lambda x: x in cancer_genes.index)\n",
      "file_s04.to_csv(os.path.join(outdir, 'file_s04.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(file_s04.shape[0]))\n",
      "file_s04.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "417 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom</th>\n",
        "      <th>strand</th>\n",
        "      <th>5'SS</th>\n",
        "      <th>3'SS</th>\n",
        "      <th>in frame</th>\n",
        "      <th>gene</th>\n",
        "      <th>COSMIC</th>\n",
        "      <th>TSGene</th>\n",
        "      <th>NCG</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr1:100606071-100606400:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 100606071</td>\n",
        "      <td> 100606400</td>\n",
        "      <td> False</td>\n",
        "      <td>  TRMT13</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109235490-109235804:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109235490</td>\n",
        "      <td> 109235804</td>\n",
        "      <td> False</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109238960-109240568:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109238960</td>\n",
        "      <td> 109240568</td>\n",
        "      <td>  True</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:110212194-110212385:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 110212194</td>\n",
        "      <td> 110212385</td>\n",
        "      <td> False</td>\n",
        "      <td>   GSTM2</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:113196284-113197099:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 113196284</td>\n",
        "      <td> 113197099</td>\n",
        "      <td> False</td>\n",
        "      <td>  CAPZA1</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 7,
       "text": [
        "                           chrom strand       5'SS       3'SS in frame  \\\n",
        "chr1:100606071-100606400:+  chr1      +  100606071  100606400    False   \n",
        "chr1:109235490-109235804:+  chr1      +  109235490  109235804    False   \n",
        "chr1:109238960-109240568:+  chr1      +  109238960  109240568     True   \n",
        "chr1:110212194-110212385:+  chr1      +  110212194  110212385    False   \n",
        "chr1:113196284-113197099:+  chr1      +  113196284  113197099    False   \n",
        "\n",
        "                               gene COSMIC TSGene    NCG  \n",
        "chr1:100606071-100606400:+   TRMT13  False  False  False  \n",
        "chr1:109235490-109235804:+  PRPF38B  False  False  False  \n",
        "chr1:109238960-109240568:+  PRPF38B  False  False  False  \n",
        "chr1:110212194-110212385:+    GSTM2  False  False  False  \n",
        "chr1:113196284-113197099:+   CAPZA1  False  False  False  "
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S5\n",
      "\n",
      "25 significant cryptic 3\u2019SSs located 10-30 bp upstream \n",
      "of canonical 3\u2019SSs and used more often in SF3B1 mutants \n",
      "from CLL-only DEXSeq analysis. Location of 5\u2019 splice \n",
      "sites and 3\u2019SSs are one-based coordinates that denote \n",
      "the start and end of the intron. The columns COSMIC, \n",
      "TSgene, and ncg denote whether the gene is present in \n",
      "COSMIC, TSGene, or the Network of Cancer Genes \n",
      "respectively."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df = pd.read_table(ds.cll_dexseq_results,\n",
      "                   index_col=0, header=0)\n",
      "tested = df.ix[df.testable]\n",
      "sig = tested[tested.padjust < ds.dexseq_p_cutoff]\n",
      "proximal = sig[(sig.downstream_acceptor_dist >= 10) & \n",
      "               (sig.downstream_acceptor_dist <= 30) &\n",
      "               (sig['log2fold(MUT/WT)'] > 0)]\n",
      "\n",
      "file_s05 = proximal[['chrom', 'strand', 'gene_name']]\n",
      "file_s05['5\\'SS'] = proximal.donor.apply(lambda x: x.split(':')[1])\n",
      "file_s05['3\\'SS'] = proximal.acceptor.apply(lambda x: x.split(':')[1])\n",
      "file_s05['in_frame'] = proximal.downstream_acceptor_dist % 3 == 0\n",
      "file_s05 = file_s05.sort(['chrom', '5\\'SS', '3\\'SS'])\n",
      "file_s05 = file_s05[['chrom', 'strand', '5\\'SS', \n",
      "                     '3\\'SS', 'in_frame', 'gene_name']]\n",
      "file_s05.columns = ['chrom', 'strand', '5\\'SS', '3\\'SS', 'in frame', 'gene']\n",
      "file_s05['COSMIC'] = file_s05.gene.apply(lambda x: x in cancer_gene_census.index)\n",
      "file_s05['TSGene'] = file_s05.gene.apply(lambda x: x in ts_genes.index)\n",
      "file_s05['NCG'] = file_s05.gene.apply(lambda x: x in cancer_genes.index)\n",
      "file_s05.to_csv(os.path.join(outdir, 'file_s05.tsv'),\n",
      "                sep='\\t', index=False)\n",
      "print('{} rows'.format(file_s05.shape[0]))\n",
      "file_s05.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "325 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>chrom</th>\n",
        "      <th>strand</th>\n",
        "      <th>5'SS</th>\n",
        "      <th>3'SS</th>\n",
        "      <th>in frame</th>\n",
        "      <th>gene</th>\n",
        "      <th>COSMIC</th>\n",
        "      <th>TSGene</th>\n",
        "      <th>NCG</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>junction</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr1:101458311-101460665:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td> 101460665</td>\n",
        "      <td> 101458311</td>\n",
        "      <td> False</td>\n",
        "      <td>    DPH5</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:109235490-109236193:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 109235490</td>\n",
        "      <td> 109236193</td>\n",
        "      <td>  True</td>\n",
        "      <td> PRPF38B</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:145601639-145601747:-</th>\n",
        "      <td> chr1</td>\n",
        "      <td> -</td>\n",
        "      <td> 145601747</td>\n",
        "      <td> 145601639</td>\n",
        "      <td> False</td>\n",
        "      <td>  POLR3C</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:153631363-153631594:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 153631363</td>\n",
        "      <td> 153631594</td>\n",
        "      <td> False</td>\n",
        "      <td>  SNAPIN</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr1:155282187-155287710:+</th>\n",
        "      <td> chr1</td>\n",
        "      <td> +</td>\n",
        "      <td> 155282187</td>\n",
        "      <td> 155287710</td>\n",
        "      <td>  True</td>\n",
        "      <td>    FDPS</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "      <td> False</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "                           chrom strand       5'SS       3'SS in frame  \\\n",
        "junction                                                                 \n",
        "chr1:101458311-101460665:-  chr1      -  101460665  101458311    False   \n",
        "chr1:109235490-109236193:+  chr1      +  109235490  109236193     True   \n",
        "chr1:145601639-145601747:-  chr1      -  145601747  145601639    False   \n",
        "chr1:153631363-153631594:+  chr1      +  153631363  153631594    False   \n",
        "chr1:155282187-155287710:+  chr1      +  155282187  155287710     True   \n",
        "\n",
        "                               gene COSMIC TSGene    NCG  \n",
        "junction                                                  \n",
        "chr1:101458311-101460665:-     DPH5  False  False  False  \n",
        "chr1:109235490-109236193:+  PRPF38B  False  False  False  \n",
        "chr1:145601639-145601747:-   POLR3C  False  False  False  \n",
        "chr1:153631363-153631594:+   SNAPIN  False  False  False  \n",
        "chr1:155282187-155287710:+     FDPS  False  False  False  "
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S6\n",
      "\n",
      "272 genes that are differentially expressed between \n",
      "SF3B1 mutant and wild-type samples from joint analysis \n",
      "of CLL, BRCA, and UM using DESeq2."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "file_s06 = pd.read_csv(ds.brca_cll_um_deseq2_results,\n",
      "                       index_col=0, header=0)\n",
      "file_s06 = file_s06.dropna(subset=['padj'])\n",
      "file_s06 = file_s06[file_s06.padj < ds.deseq2_p_cutoff]\n",
      "file_s06 = file_s06[['log2FoldChange', 'pvalue', 'padj']]\n",
      "file_s06.columns = ['log2 fold change (MUT/WT)', 'p-value', \n",
      "                    'Adjusted p-value']\n",
      "file_s06['Gene name'] = gene_info.ix[file_s06.index, 'gene_name']\n",
      "file_s06.index.name = 'Gene ID'\n",
      "\n",
      "file_s06.to_csv(os.path.join(outdir, 'file_s06.tsv'),\n",
      "                sep='\\t')\n",
      "print('{} rows'.format(file_s06.shape[0]))\n",
      "file_s06.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "272 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>log2 fold change (MUT/WT)</th>\n",
        "      <th>p-value</th>\n",
        "      <th>Adjusted p-value</th>\n",
        "      <th>Gene name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Gene ID</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>ENSG00000005156.7</th>\n",
        "      <td> 0.560285</td>\n",
        "      <td> 0.000046</td>\n",
        "      <td> 0.019017</td>\n",
        "      <td>   LIG3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000006459.6</th>\n",
        "      <td> 0.565435</td>\n",
        "      <td> 0.000841</td>\n",
        "      <td> 0.075077</td>\n",
        "      <td> JHDM1D</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008226.15</th>\n",
        "      <td> 1.028977</td>\n",
        "      <td> 0.000050</td>\n",
        "      <td> 0.019873</td>\n",
        "      <td>  DLEC1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008324.6</th>\n",
        "      <td> 0.432990</td>\n",
        "      <td> 0.001072</td>\n",
        "      <td> 0.082723</td>\n",
        "      <td> SS18L2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008988.5</th>\n",
        "      <td> 0.918372</td>\n",
        "      <td> 0.000001</td>\n",
        "      <td> 0.002189</td>\n",
        "      <td>  RPS20</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 9,
       "text": [
        "                    log2 fold change (MUT/WT)   p-value  Adjusted p-value  \\\n",
        "Gene ID                                                                     \n",
        "ENSG00000005156.7                    0.560285  0.000046          0.019017   \n",
        "ENSG00000006459.6                    0.565435  0.000841          0.075077   \n",
        "ENSG00000008226.15                   1.028977  0.000050          0.019873   \n",
        "ENSG00000008324.6                    0.432990  0.001072          0.082723   \n",
        "ENSG00000008988.5                    0.918372  0.000001          0.002189   \n",
        "\n",
        "                   Gene name  \n",
        "Gene ID                       \n",
        "ENSG00000005156.7       LIG3  \n",
        "ENSG00000006459.6     JHDM1D  \n",
        "ENSG00000008226.15     DLEC1  \n",
        "ENSG00000008324.6     SS18L2  \n",
        "ENSG00000008988.5      RPS20  "
       ]
      }
     ],
     "prompt_number": 9
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## File S7\n",
      "\n",
      "Genes whose expression differs significantly between\n",
      "*SF3B1* mutant and wild-type samples for analysis\n",
      "of CLL samples using DESeq2."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "file_s07 = pd.read_csv(ds.cll_deseq2_results,\n",
      "                       index_col=0, header=0)\n",
      "file_s07 = file_s07.dropna(subset=['padj'])\n",
      "file_s07 = file_s07[file_s07.padj < ds.deseq2_p_cutoff]\n",
      "file_s07 = file_s07[['log2FoldChange', 'pvalue', 'padj']]\n",
      "file_s07.columns = ['log2 fold change (MUT/WT)', 'p-value', \n",
      "                    'Adjusted p-value']\n",
      "file_s07['Gene name'] = gene_info.ix[file_s07.index, 'gene_name']\n",
      "file_s07.index.name = 'Gene ID'\n",
      "\n",
      "file_s07.to_csv(os.path.join(outdir, 'file_s07.tsv'),\n",
      "                sep='\\t')\n",
      "print('{} rows'.format(file_s07.shape[0]))\n",
      "file_s07.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "33 rows\n"
       ]
      },
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>log2 fold change (MUT/WT)</th>\n",
        "      <th>p-value</th>\n",
        "      <th>Adjusted p-value</th>\n",
        "      <th>Gene name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>Gene ID</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>ENSG00000005156.7</th>\n",
        "      <td> 0.606988</td>\n",
        "      <td> 0.000080</td>\n",
        "      <td> 0.083089</td>\n",
        "      <td>    LIG3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000008226.15</th>\n",
        "      <td> 1.020171</td>\n",
        "      <td> 0.000167</td>\n",
        "      <td> 0.088107</td>\n",
        "      <td>   DLEC1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000028116.12</th>\n",
        "      <td>-0.868542</td>\n",
        "      <td> 0.000042</td>\n",
        "      <td> 0.066585</td>\n",
        "      <td>    VRK2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000066027.4</th>\n",
        "      <td> 0.840471</td>\n",
        "      <td> 0.000208</td>\n",
        "      <td> 0.099852</td>\n",
        "      <td> PPP2R5A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>ENSG00000070785.12</th>\n",
        "      <td>-0.533570</td>\n",
        "      <td> 0.000128</td>\n",
        "      <td> 0.083089</td>\n",
        "      <td>  EIF2B3</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 10,
       "text": [
        "                    log2 fold change (MUT/WT)   p-value  Adjusted p-value  \\\n",
        "Gene ID                                                                     \n",
        "ENSG00000005156.7                    0.606988  0.000080          0.083089   \n",
        "ENSG00000008226.15                   1.020171  0.000167          0.088107   \n",
        "ENSG00000028116.12                  -0.868542  0.000042          0.066585   \n",
        "ENSG00000066027.4                    0.840471  0.000208          0.099852   \n",
        "ENSG00000070785.12                  -0.533570  0.000128          0.083089   \n",
        "\n",
        "                   Gene name  \n",
        "Gene ID                       \n",
        "ENSG00000005156.7       LIG3  \n",
        "ENSG00000008226.15     DLEC1  \n",
        "ENSG00000028116.12      VRK2  \n",
        "ENSG00000066027.4    PPP2R5A  \n",
        "ENSG00000070785.12    EIF2B3  "
       ]
      }
     ],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    }
   ],
   "metadata": {}
  }
 ]
}