{
 "metadata": {
  "name": "",
  "signature": "sha256:98ab0b86e9d5acc6cbd36030b1029479acadd933bcf24b0ac835298533e5da16"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "# Susceptible Introns\n",
      "\n",
      "I'd like to identify which introns my susceptible to the proximal\n",
      "cryptic splicing effect I've identified."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import glob\n",
      "import os\n",
      "import re\n",
      "import shutil\n",
      "import subprocess\n",
      "from urllib2 import urlopen\n",
      "\n",
      "from Bio.Seq import Seq\n",
      "import ds2014 as ds\n",
      "import pandas as pd\n",
      "import pybedtools as pbt"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "curdir = os.getcwd()\n",
      "outdir = os.path.join(os.path.split(curdir)[0], \n",
      "                      'output', \n",
      "                      'susceptible_introns')\n",
      "ds.makedir(outdir)\n",
      "\n",
      "meta = pd.read_table(ds.metadata, index_col=0, header=0)\n",
      "meta = meta[meta.disease_abbr.apply(lambda x: x in ['BRCA', 'CLL', 'UM'])]\n",
      "\n",
      "results = pd.read_table(ds.brca_cll_um_dexseq_results,\n",
      "                        index_col=0,\n",
      "                        header=0)\n",
      "\n",
      "counts = pd.read_table(ds.brca_cll_um_sj_counts,\n",
      "                       index_col=0,\n",
      "                       header=0)\n",
      "\n",
      "size_factors = pd.read_table(ds.brca_cll_um_dexseq_size_factors,\n",
      "                             index_col=0,\n",
      "                             header=0)\n",
      "\n",
      "annot = pd.read_table(ds.brca_cll_um_sj_annot, index_col=0,\n",
      "                      header=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Get intron sequences\n",
      "\n",
      "I'm going to get the last 50 bp of the intron for each unique\n",
      "splicing acceptor that passed the coverage cutoff. I'll go \n",
      "ahead and reverse complement the minus strand sequence."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "res = results.sort(['acceptor', 'padjust']).drop_duplicates(subset='acceptor')\n",
      "res = res.dropna(subset=['padjust'])\n",
      "not_sig = res[res.padjust > ds.dexseq_p_cutoff]\n",
      "sig = res[res.padjust < ds.dexseq_p_cutoff]\n",
      "\n",
      "if not os.path.exists(ds.susc_intron_seq):\n",
      "    plus = res[res.strand == '+']\n",
      "    minus = res[res.strand == '-']\n",
      "\n",
      "    plus = '\\n'.join(plus.chrom + '\\t' + plus.end.astype(str) + '\\t' + \n",
      "                     plus.end.astype(str) + '\\t' + plus.index + '\\t0\\t' + \n",
      "                     plus.strand) + '\\n'\n",
      "    minus = '\\n'.join(minus.chrom + '\\t' + (minus.start - 1).astype(str) + '\\t' + \n",
      "                      (minus.start - 1).astype(str) + '\\t' + minus.index + '\\t0\\t' + \n",
      "                      minus.strand) + '\\n'\n",
      "\n",
      "    plus = pbt.BedTool(plus, from_string=True)\n",
      "    minus = pbt.BedTool(minus, from_string=True)\n",
      "\n",
      "    plus = plus.slop(l=50, r=0, g=pbt.genome_registry.hg19)\n",
      "    minus = minus.slop(l=0, r=50, g=pbt.genome_registry.hg19)\n",
      "    \n",
      "    plus_seq = plus.sequence(fi=ds.hg19)\n",
      "    minus_seq = minus.sequence(fi=ds.hg19)\n",
      "\n",
      "    with open(plus_seq.seqfn) as f:\n",
      "        p_lines = [x.strip() for x in f.readlines()]\n",
      "    with open(minus_seq.seqfn) as f:\n",
      "        m_lines = [x.strip() for x in f.readlines()]\n",
      "\n",
      "    for i in range(len(m_lines))[1::2]:\n",
      "        m_lines[i] = Seq(m_lines[i]).reverse_complement().tostring()\n",
      "    \n",
      "    for i in range(len(p_lines))[0::2]:\n",
      "        p_lines[i] += ':+'\n",
      "    r = re.compile('(.+):(.+)-(.+)')    \n",
      "    for i in range(len(m_lines))[0::2]:\n",
      "        m = r.match(m_lines[i])\n",
      "        start = int(m.group(2)) + 1\n",
      "        m_lines[i] = '{}:{}-{}:-'.format(m.group(1),\n",
      "                                         start,\n",
      "                                         m.group(2))\n",
      "    \n",
      "    out = ('\\n'.join(p_lines) + '\\n' + \n",
      "           '\\n'.join(m_lines) + '\\n')\n",
      "\n",
      "    with open(ds.susc_intron_seq, 'w') as f:\n",
      "        f.write(out)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Annotate intron sequences\n",
      "\n",
      "I'll note whether each intron contains an AG dinucleotide\n",
      "10-30 bp upstream of the canonical 3'SS. I'll also predict\n",
      "branch points for each intron."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "with open(ds.susc_intron_seq) as f:\n",
      "    lines = [x.strip() for x in f.readlines()]\n",
      "seqs = pd.DataFrame(pd.Series([x.upper() for x in lines[1::2]], \n",
      "                              index=[x[1:] for x in lines[0::2]]))\n",
      "seqs.columns = ['seq']\n",
      "seqs['ag'] = seqs.seq.apply(lambda x: x[-2:] == 'AG')\n",
      "seqs['proximal_ag'] = seqs.seq.apply(lambda x: 'AG' in x[20:-10])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Let's check to make sure we obtained the correct sequences. Most \n",
      "introns should end with `AG`."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "seqs.seq.apply(lambda x: x[-2:]).value_counts()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 5,
       "text": [
        "AG    222934\n",
        "AC       260\n",
        "GG        64\n",
        "TG        61\n",
        "AT        46\n",
        "AA        41\n",
        "CG        24\n",
        "CA        22\n",
        "GC        19\n",
        "CC        19\n",
        "TT        15\n",
        "TA        12\n",
        "GA        11\n",
        "TC        11\n",
        "GT         8\n",
        "CT         7\n",
        "dtype: int64"
       ]
      }
     ],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "if not os.path.exists(ds.susc_all_bp):\n",
      "    bps = ds.predict_branch_points(ds.susc_intron_seq)\n",
      "    r = re.compile('(.+):(.+)-(.+):(.+)')\n",
      "    chrom = []\n",
      "    start = []\n",
      "    end = []\n",
      "    strand = []\n",
      "    acceptor = []\n",
      "    for ind in bps.index:\n",
      "        m = r.match(bps.ix[ind, 'seq_id'])\n",
      "        chrom.append(m.group(1))\n",
      "        start.append(int(m.group(2)))\n",
      "        end.append(int(m.group(3)))\n",
      "        strand.append(m.group(4))\n",
      "        if m.group(4) == '+':\n",
      "            acceptor.append('{}:{}:{}'.format(m.group(1),\n",
      "                                              m.group(3),\n",
      "                                              m.group(4)))\n",
      "        else:\n",
      "            acceptor.append('{}:{}:{}'.format(m.group(1),\n",
      "                                              m.group(2),\n",
      "                                              m.group(4)))\n",
      "    bps['chrom'] = chrom\n",
      "    bps['start'] = start\n",
      "    bps['end'] = end\n",
      "    bps['strand'] = strand\n",
      "    bps['acceptor'] = acceptor\n",
      "    bps.to_csv(ds.susc_all_bp, sep='\\t')\n",
      "else:\n",
      "    bps = pd.read_table(ds.susc_all_bp, index_col=0,\n",
      "                        header=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Unique branch points and susceptible introns\n",
      "\n",
      "Now I'll keep only the highest scoring branch points \n",
      "for each intron and see which introns are susceptible \n",
      "to cryptic splicing."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "uniq_bp = ds.make_uniq_bp_results(bps)\n",
      "uniq_bp.shape"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "(210944, 15)"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "uniq_bp.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>seq_id</th>\n",
        "      <th>agez</th>\n",
        "      <th>ss_dist</th>\n",
        "      <th>bp_seq</th>\n",
        "      <th>bp_scr</th>\n",
        "      <th>y_cont</th>\n",
        "      <th>ppt_off</th>\n",
        "      <th>ppt_len</th>\n",
        "      <th>ppt_scr</th>\n",
        "      <th>svm_scr</th>\n",
        "      <th>chrom</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>strand</th>\n",
        "      <th>acceptor</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chrY:7243702-7243752:+</th>\n",
        "      <td> chrY:7243702-7243752:+</td>\n",
        "      <td> 35</td>\n",
        "      <td> 37</td>\n",
        "      <td> tactgagca</td>\n",
        "      <td> 0.921587</td>\n",
        "      <td> 0.625000</td>\n",
        "      <td> 14</td>\n",
        "      <td> 18</td>\n",
        "      <td> 34</td>\n",
        "      <td> 0.395041</td>\n",
        "      <td> chrY</td>\n",
        "      <td> 7243702</td>\n",
        "      <td> 7243752</td>\n",
        "      <td> +</td>\n",
        "      <td> chrY:7243752:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chrY:7240722-7240772:+</th>\n",
        "      <td> chrY:7240722-7240772:+</td>\n",
        "      <td> 45</td>\n",
        "      <td> 12</td>\n",
        "      <td> ccgtaattt</td>\n",
        "      <td> 0.295890</td>\n",
        "      <td> 1.000000</td>\n",
        "      <td>  1</td>\n",
        "      <td>  7</td>\n",
        "      <td> 19</td>\n",
        "      <td> 0.954306</td>\n",
        "      <td> chrY</td>\n",
        "      <td> 7240722</td>\n",
        "      <td> 7240772</td>\n",
        "      <td> +</td>\n",
        "      <td> chrY:7240772:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chrY:7239733-7239783:+</th>\n",
        "      <td> chrY:7239733-7239783:+</td>\n",
        "      <td> 50</td>\n",
        "      <td> 33</td>\n",
        "      <td> ttctaacca</td>\n",
        "      <td> 3.285055</td>\n",
        "      <td> 0.714286</td>\n",
        "      <td>  4</td>\n",
        "      <td> 14</td>\n",
        "      <td> 22</td>\n",
        "      <td> 1.870481</td>\n",
        "      <td> chrY</td>\n",
        "      <td> 7239733</td>\n",
        "      <td> 7239783</td>\n",
        "      <td> +</td>\n",
        "      <td> chrY:7239783:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chrY:7235346-7235396:+</th>\n",
        "      <td> chrY:7235346-7235396:+</td>\n",
        "      <td> 50</td>\n",
        "      <td> 25</td>\n",
        "      <td> cactgacga</td>\n",
        "      <td> 3.165837</td>\n",
        "      <td> 0.700000</td>\n",
        "      <td> 10</td>\n",
        "      <td> 11</td>\n",
        "      <td> 27</td>\n",
        "      <td> 1.485977</td>\n",
        "      <td> chrY</td>\n",
        "      <td> 7235346</td>\n",
        "      <td> 7235396</td>\n",
        "      <td> +</td>\n",
        "      <td> chrY:7235396:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chrY:7224125-7224175:+</th>\n",
        "      <td> chrY:7224125-7224175:+</td>\n",
        "      <td> 50</td>\n",
        "      <td> 37</td>\n",
        "      <td> tgcttactc</td>\n",
        "      <td> 1.094126</td>\n",
        "      <td> 0.656250</td>\n",
        "      <td>  1</td>\n",
        "      <td>  9</td>\n",
        "      <td> 16</td>\n",
        "      <td> 1.127885</td>\n",
        "      <td> chrY</td>\n",
        "      <td> 7224125</td>\n",
        "      <td> 7224175</td>\n",
        "      <td> +</td>\n",
        "      <td> chrY:7224175:+</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "                                        seq_id  agez  ss_dist     bp_seq  \\\n",
        "chrY:7243702-7243752:+  chrY:7243702-7243752:+    35       37  tactgagca   \n",
        "chrY:7240722-7240772:+  chrY:7240722-7240772:+    45       12  ccgtaattt   \n",
        "chrY:7239733-7239783:+  chrY:7239733-7239783:+    50       33  ttctaacca   \n",
        "chrY:7235346-7235396:+  chrY:7235346-7235396:+    50       25  cactgacga   \n",
        "chrY:7224125-7224175:+  chrY:7224125-7224175:+    50       37  tgcttactc   \n",
        "\n",
        "                          bp_scr    y_cont  ppt_off  ppt_len  ppt_scr  \\\n",
        "chrY:7243702-7243752:+  0.921587  0.625000       14       18       34   \n",
        "chrY:7240722-7240772:+  0.295890  1.000000        1        7       19   \n",
        "chrY:7239733-7239783:+  3.285055  0.714286        4       14       22   \n",
        "chrY:7235346-7235396:+  3.165837  0.700000       10       11       27   \n",
        "chrY:7224125-7224175:+  1.094126  0.656250        1        9       16   \n",
        "\n",
        "                         svm_scr chrom    start      end strand  \\\n",
        "chrY:7243702-7243752:+  0.395041  chrY  7243702  7243752      +   \n",
        "chrY:7240722-7240772:+  0.954306  chrY  7240722  7240772      +   \n",
        "chrY:7239733-7239783:+  1.870481  chrY  7239733  7239783      +   \n",
        "chrY:7235346-7235396:+  1.485977  chrY  7235346  7235396      +   \n",
        "chrY:7224125-7224175:+  1.127885  chrY  7224125  7224175      +   \n",
        "\n",
        "                              acceptor  \n",
        "chrY:7243702-7243752:+  chrY:7243752:+  \n",
        "chrY:7240722-7240772:+  chrY:7240772:+  \n",
        "chrY:7239733-7239783:+  chrY:7239783:+  \n",
        "chrY:7235346-7235396:+  chrY:7235396:+  \n",
        "chrY:7224125-7224175:+  chrY:7224175:+  "
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "For an intron to be susceptible, it must have an AG dinucleotide 10-30 bp\n",
      "upstream of the canonical 3'SS that is also 13-17 bp downstream of the \n",
      "predicted BP.\n",
      "\n",
      "I'll want to determine some distances. I'll use\n",
      "`chr1:169799786-169799836:+` as an example. I've made\n",
      "the predicted branch point and potential cryptic \n",
      "3'SS lower case.\n",
      "    \n",
      "                bp_proximal_dist\n",
      "                |-------------|\n",
      "                              |-proximal_cryptic_dist|\n",
      "                |--------------ss_dist---------------|\n",
      "    TGTTTCTgagtaattgTATAAACATagTTGTCTTCTCTTTTATATTGCAG\n",
      "                       \n",
      "In this example:\n",
      "\n",
      "* `ss_dist = 38 (attgTATAAACATagTTGTCTTCTCTTTTATATTGCAG)`\n",
      "* `bp_proximal_dist = 15 (attgTATAAACATag)`\n",
      "* `proximal_cryptic_dist = 23 (TTGTCTTCTCTTTTATATTGCAG)`"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "susc = uniq_bp.ix[seqs[seqs.proximal_ag].index].dropna()\n",
      "print('{:,} introns have an AG 10-30 '.format(susc.shape[0]) + \n",
      "      'bp upstream of canonical 3\\'SS.')\n",
      "print('')\n",
      "\n",
      "def possible_cryptic(ind, susc=susc, seqs=seqs):\n",
      "    s = seqs.ix[ind, 'seq']\n",
      "    sub = s[-1 * susc.ix[ind, 'ss_dist'].astype(int):]\n",
      "    return 'AG' in sub[11:17]\n",
      "\n",
      "def cryptic_distance(ind, susc=susc, seqs=seqs):\n",
      "    s = seqs.ix[ind, 'seq']\n",
      "    sub = s[-1 * susc.ix[ind, 'ss_dist'].astype(int):]\n",
      "    return 13 + sub[11:17].find('AG')\n",
      "\n",
      "susc['possible_cryptic'] = [possible_cryptic(x) for x in susc.index]\n",
      "susc = susc[susc.possible_cryptic]\n",
      "print('{:,} introns have an AG 13-17 bp '.format(susc.shape[0]) + \n",
      "      'downstream of the predicted BP and ' + \n",
      "      'an AG 10-30 bp upstream of the canonical ' + \n",
      "      '3\\'SS.')\n",
      "print('')\n",
      "\n",
      "susc['bp_proximal_dist'] = [cryptic_distance(x) for x in susc.index]\n",
      "susc['proximal_cryptic_dist'] = susc.ss_dist - susc.bp_proximal_dist\n",
      "\n",
      "susc = susc[susc.proximal_cryptic_dist >= 10]\n",
      "print('{:,} introns have an AG 13-17 bp '.format(susc.shape[0]) + \n",
      "      'downstream of the predicted BP and the same ' + \n",
      "      'AG is 10-30 bp upstream of the canonical ' + \n",
      "      '3\\'SS.')\n",
      "\n",
      "def cryptic_acceptor(ind, susc=susc):\n",
      "    if susc.ix[ind, 'strand'] == '+':\n",
      "        out = '{}:{}:+'.format(susc.ix[ind, 'chrom'],\n",
      "                               int(susc.ix[ind, 'end'] - \n",
      "                                   susc.ix[ind, 'proximal_cryptic_dist']))\n",
      "    else:\n",
      "        out = '{}:{}:-'.format(susc.ix[ind, 'chrom'],\n",
      "                               int(susc.ix[ind, 'start'] + \n",
      "                                   susc.ix[ind, 'proximal_cryptic_dist']))\n",
      "    return out\n",
      "\n",
      "susc['cryptic_acceptor'] = [cryptic_acceptor(x) for x in susc.index]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "82,860 introns have an AG 10-30 bp upstream of canonical 3'SS.\n",
        "\n",
        "18,801 introns have an AG 13-17 bp downstream of the predicted BP and an AG 10-30 bp upstream of the canonical 3'SS."
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "\n",
        "11,302 introns have an AG 13-17 bp downstream of the predicted BP and the same AG is 10-30 bp upstream of the canonical 3'SS."
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "susc.head()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>seq_id</th>\n",
        "      <th>agez</th>\n",
        "      <th>ss_dist</th>\n",
        "      <th>bp_seq</th>\n",
        "      <th>bp_scr</th>\n",
        "      <th>y_cont</th>\n",
        "      <th>ppt_off</th>\n",
        "      <th>ppt_len</th>\n",
        "      <th>ppt_scr</th>\n",
        "      <th>svm_scr</th>\n",
        "      <th>chrom</th>\n",
        "      <th>start</th>\n",
        "      <th>end</th>\n",
        "      <th>strand</th>\n",
        "      <th>acceptor</th>\n",
        "      <th>possible_cryptic</th>\n",
        "      <th>bp_proximal_dist</th>\n",
        "      <th>proximal_cryptic_dist</th>\n",
        "      <th>cryptic_acceptor</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>chr10:101124123-101124173:+</th>\n",
        "      <td> chr10:101124123-101124173:+</td>\n",
        "      <td> 25</td>\n",
        "      <td> 39</td>\n",
        "      <td> tgctcatac</td>\n",
        "      <td> 1.340156</td>\n",
        "      <td> 0.764706</td>\n",
        "      <td> 16</td>\n",
        "      <td> 19</td>\n",
        "      <td> 48</td>\n",
        "      <td> 0.607877</td>\n",
        "      <td> chr10</td>\n",
        "      <td> 101124123</td>\n",
        "      <td> 101124173</td>\n",
        "      <td> +</td>\n",
        "      <td> chr10:101124173:+</td>\n",
        "      <td> True</td>\n",
        "      <td> 14</td>\n",
        "      <td> 25</td>\n",
        "      <td> chr10:101124148:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr10:101569840-101569890:+</th>\n",
        "      <td> chr10:101569840-101569890:+</td>\n",
        "      <td> 21</td>\n",
        "      <td> 36</td>\n",
        "      <td> tactgactt</td>\n",
        "      <td> 2.610401</td>\n",
        "      <td> 0.741935</td>\n",
        "      <td>  1</td>\n",
        "      <td> 11</td>\n",
        "      <td> 21</td>\n",
        "      <td> 1.795831</td>\n",
        "      <td> chr10</td>\n",
        "      <td> 101569840</td>\n",
        "      <td> 101569890</td>\n",
        "      <td> +</td>\n",
        "      <td> chr10:101569890:+</td>\n",
        "      <td> True</td>\n",
        "      <td> 15</td>\n",
        "      <td> 21</td>\n",
        "      <td> chr10:101569869:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr10:103289989-103290039:+</th>\n",
        "      <td> chr10:103289989-103290039:+</td>\n",
        "      <td> 27</td>\n",
        "      <td> 43</td>\n",
        "      <td> acttgatag</td>\n",
        "      <td>-1.146542</td>\n",
        "      <td> 0.763158</td>\n",
        "      <td> 17</td>\n",
        "      <td> 22</td>\n",
        "      <td> 55</td>\n",
        "      <td>-0.364370</td>\n",
        "      <td> chr10</td>\n",
        "      <td> 103289989</td>\n",
        "      <td> 103290039</td>\n",
        "      <td> +</td>\n",
        "      <td> chr10:103290039:+</td>\n",
        "      <td> True</td>\n",
        "      <td> 16</td>\n",
        "      <td> 27</td>\n",
        "      <td> chr10:103290012:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr10:103356719-103356769:+</th>\n",
        "      <td> chr10:103356719-103356769:+</td>\n",
        "      <td> 27</td>\n",
        "      <td> 40</td>\n",
        "      <td> tgttaataa</td>\n",
        "      <td>-0.229114</td>\n",
        "      <td> 0.714286</td>\n",
        "      <td> 13</td>\n",
        "      <td> 23</td>\n",
        "      <td> 46</td>\n",
        "      <td> 0.148413</td>\n",
        "      <td> chr10</td>\n",
        "      <td> 103356719</td>\n",
        "      <td> 103356769</td>\n",
        "      <td> +</td>\n",
        "      <td> chr10:103356769:+</td>\n",
        "      <td> True</td>\n",
        "      <td> 13</td>\n",
        "      <td> 27</td>\n",
        "      <td> chr10:103356742:+</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>chr10:103898325-103898375:+</th>\n",
        "      <td> chr10:103898325-103898375:+</td>\n",
        "      <td> 27</td>\n",
        "      <td> 44</td>\n",
        "      <td> aggtcatta</td>\n",
        "      <td>-2.701709</td>\n",
        "      <td> 0.538462</td>\n",
        "      <td> 19</td>\n",
        "      <td> 18</td>\n",
        "      <td> 35</td>\n",
        "      <td>-1.358775</td>\n",
        "      <td> chr10</td>\n",
        "      <td> 103898325</td>\n",
        "      <td> 103898375</td>\n",
        "      <td> +</td>\n",
        "      <td> chr10:103898375:+</td>\n",
        "      <td> True</td>\n",
        "      <td> 17</td>\n",
        "      <td> 27</td>\n",
        "      <td> chr10:103898348:+</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "                                                  seq_id  agez  ss_dist  \\\n",
        "chr10:101124123-101124173:+  chr10:101124123-101124173:+    25       39   \n",
        "chr10:101569840-101569890:+  chr10:101569840-101569890:+    21       36   \n",
        "chr10:103289989-103290039:+  chr10:103289989-103290039:+    27       43   \n",
        "chr10:103356719-103356769:+  chr10:103356719-103356769:+    27       40   \n",
        "chr10:103898325-103898375:+  chr10:103898325-103898375:+    27       44   \n",
        "\n",
        "                                bp_seq    bp_scr    y_cont  ppt_off  ppt_len  \\\n",
        "chr10:101124123-101124173:+  tgctcatac  1.340156  0.764706       16       19   \n",
        "chr10:101569840-101569890:+  tactgactt  2.610401  0.741935        1       11   \n",
        "chr10:103289989-103290039:+  acttgatag -1.146542  0.763158       17       22   \n",
        "chr10:103356719-103356769:+  tgttaataa -0.229114  0.714286       13       23   \n",
        "chr10:103898325-103898375:+  aggtcatta -2.701709  0.538462       19       18   \n",
        "\n",
        "                             ppt_scr   svm_scr  chrom      start        end  \\\n",
        "chr10:101124123-101124173:+       48  0.607877  chr10  101124123  101124173   \n",
        "chr10:101569840-101569890:+       21  1.795831  chr10  101569840  101569890   \n",
        "chr10:103289989-103290039:+       55 -0.364370  chr10  103289989  103290039   \n",
        "chr10:103356719-103356769:+       46  0.148413  chr10  103356719  103356769   \n",
        "chr10:103898325-103898375:+       35 -1.358775  chr10  103898325  103898375   \n",
        "\n",
        "                            strand           acceptor possible_cryptic  \\\n",
        "chr10:101124123-101124173:+      +  chr10:101124173:+             True   \n",
        "chr10:101569840-101569890:+      +  chr10:101569890:+             True   \n",
        "chr10:103289989-103290039:+      +  chr10:103290039:+             True   \n",
        "chr10:103356719-103356769:+      +  chr10:103356769:+             True   \n",
        "chr10:103898325-103898375:+      +  chr10:103898375:+             True   \n",
        "\n",
        "                             bp_proximal_dist  proximal_cryptic_dist  \\\n",
        "chr10:101124123-101124173:+                14                     25   \n",
        "chr10:101569840-101569890:+                15                     21   \n",
        "chr10:103289989-103290039:+                16                     27   \n",
        "chr10:103356719-103356769:+                13                     27   \n",
        "chr10:103898325-103898375:+                17                     27   \n",
        "\n",
        "                              cryptic_acceptor  \n",
        "chr10:101124123-101124173:+  chr10:101124148:+  \n",
        "chr10:101569840-101569890:+  chr10:101569869:+  \n",
        "chr10:103289989-103290039:+  chr10:103290012:+  \n",
        "chr10:103356719-103356769:+  chr10:103356742:+  \n",
        "chr10:103898325-103898375:+  chr10:103898348:+  "
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "There are 11,302 introns that are susceptible to cryptic\n",
      "splicing.\n",
      "\n",
      "Now that we have identified some potential crpytic junctions,\n",
      "we can check whether any of the cryptic junctions passed \n",
      "our coverage cutoff. So far, we've only checked that the \n",
      "canonical junctions passed the cutoff."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "n = len(set(annot.acceptor) & set(susc.cryptic_acceptor))\n",
      "print('{} introns have potential cryptic 3\\'SSs '.format(n) + \n",
      "      'that passed the coverage cutoff.')\n",
      "\n",
      "df = pd.DataFrame(list(set(annot.acceptor) & set(susc.cryptic_acceptor)),\n",
      "                  columns=['acceptor'])\n",
      "pot_c = annot.merge(df, on='acceptor')\n",
      "print('The 3\\'SSs for these {} introns are '.format(n) + \n",
      "      'used in {:,} splice junctions.'.format(pot_c.shape[0]))\n",
      "\n",
      "df = pd.DataFrame(list(set(sig.acceptor) & set(susc.cryptic_acceptor)),\n",
      "                  columns=['acceptor'])\n",
      "sig_pot_c = annot.merge(df, on='acceptor')\n",
      "print('{} of these {} introns are '.format(df.shape[0], n) + \n",
      "      'significant corresponding to {:,} '.format(sig_pot_c.shape[0]) + \n",
      "      'significant splice junctions.')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "900 introns have potential cryptic 3'SSs that passed the coverage cutoff.\n",
        "The 3'SSs for these 900 introns are used in 1,123 splice junctions."
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "310 of these 900 introns are significant corresponding to 333 significant splice junctions."
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "So of the 11,302 potential cryptic junctions, 900 were \n",
      "part of junctions that passed my coverage filters. \n",
      "These 900 potential cryptic 3'SSs correspond to 1,075 \n",
      "junctions. 310 of the cryptic 3'SSs were significant corresponding\n",
      "to 333 junctions."
     ]
    }
   ],
   "metadata": {}
  }
 ]
}